{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# GWAS sumstat Processing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Pre-requisites\n",
    "\n",
    "We provide a container image `docker://gaow/twas` that contains all software needed to run the pipeline. If you would like to configure it by yourself, please make sure you install tidyverse"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Input and Output\n",
    "## Input\n",
    "\n",
    "- `--GWAS sumstat` An sumstat text file at least with CHR, bp, SNP name, bhat and sbhat, chr and bp column shall be named CHR and BP.\n",
    "\n",
    "- `--region_list` The text file with 4 columns specifying the #Chr, P0 (Start position), P1(End position) and names of regions to analyze. The name of the column is not important but the order of the columns. It is also important that the column name of the first column starts with a #. The region_list can can be generated by using another sos pipeline SOS_ROSMAP_gene_exp_processing.ipynb.\n",
    "\n",
    "## Output\n",
    "\n",
    "- `uni_weight.RDS` a RDS file that served as the input for the mixture pipeline.\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Command interface "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: sos run GWAS_Gene_annotation.ipynb\n",
      "               [workflow_name | -t targets] [options] [workflow_options]\n",
      "  workflow_name:        Single or combined workflows defined in this script\n",
      "  targets:              One or more targets to generate\n",
      "  options:              Single-hyphen sos parameters (see \"sos run -h\" for details)\n",
      "  workflow_options:     Double-hyphen workflow-specific parameters\n",
      "\n",
      "Workflows:\n",
      "  complete_data_analysis\n",
      "  hsq\n",
      "  susie\n",
      "  susie_cv\n",
      "  susie_get_info\n",
      "\n",
      "Global Workflow Options:\n",
      "  --analysis-units VAL (as path, required)\n",
      "                        single column file each line is the data filename\n",
      "  --data-dir VAL (as path, required)\n",
      "                        Path to data directory\n",
      "  --data-suffix VAL (as str, required)\n",
      "                        data file suffix\n",
      "  --wd output (as path)\n",
      "                        Path to work directory where output locates\n",
      "  --prior . (as path)\n",
      "                        Path to prior data file: an RDS file with `U` and `w`\n",
      "                        for prior matrices and weights\n",
      "  --resid-cor . (as path)\n",
      "                        Path to residual cor/cov data file\n",
      "  --name VAL (as str, required)\n",
      "                        An identifier for your run of analysis\n",
      "  --cis NULL\n",
      "                        Only analyze `cis` variants -- cis = N means using N\n",
      "                        variants around the center column of X matrix\n",
      "  --container 'gaow/twas'\n",
      "  --job-size 2 (as int)\n",
      "\n",
      "Sections\n",
      "  complete_data_analysis_1:\n",
      "    Workflow Options:\n",
      "      --x-table X\n",
      "      --y-table 'y_res'\n",
      "      --imiss 0.1 (as float)\n",
      "                        remove a variant if it has more than imiss missing\n",
      "                        individual data\n",
      "      --maf 0.05 (as float)\n",
      "      --max-L 5 (as int)\n",
      "  hsq_2, susie_2, susie_cv_2:\n",
      "    Workflow Options:\n",
      "      --extract-snp  f'{geno_file:an}.bim'\n",
      "\n",
      "      --exclude-snp './.'\n",
      "      --keep-sample  f'{_input[1]}'\n",
      "\n",
      "  hsq_3:\n",
      "  susie_3, susie_cv_3:  Susie test\n",
      "  susie_cv_4:           CV with univariate susie\n",
      "  susie_get_info:\n",
      "  susie_get_info_2:\n",
      "  susie_4:              Create all_hsq for susie and saved all the susie object\n",
      "                        into one RData\n",
      "  susie_cv_5:           Create all_hsq for susie and saved all the susie object\n",
      "                        into one RData\n"
     ]
    }
   ],
   "source": [
    "!sos run GWAS_Gene_annotation.ipynb -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Working example\n",
    "The MWE file is availble at :\n",
    "\"https://www.synapse.org/#!Synapse:syn24179064/files/\"\n",
    "\n",
    "The time it take to run this MWE shall be around 2 minutes. Pay extra attention to the gene_start and gene_end position  when using following command on gene_exp file that are not this MWE. Also, when there is too few or too many genes that passed the heritability check, consider increasing or decreasing the --window options. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[91mERROR\u001b[0m: \u001b[91mFailed to locate twas_fusion.ipynb.sos\u001b[0m\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## Test pipeline with test data\n",
    "## Switch back to abosolute path, otherwise there will be file not found error in step 5\n",
    "nohup sos run ~/GIT/neuro-twas/workflow/GWAS_Gene_annotation.ipynb \\\n",
    "  --wd \"/mnt/mfs/statgen/neuro-twas/TWAS_sumstat/\" \\\n",
    "  --GWAS_sumstat \"/mnt/mfs/statgen/neuro-twas/TWAS_sumstat/Data/GCST90012877_buildGRCh37_colrenamed.txt\"\\\n",
    "  --region_list  \"/mnt/mfs/statgen/neuro-twas/mv_wg/wg_rds_list_final\" \\\n",
    "  --window 500000 \\\n",
    "  --container /mnt/mfs/statgen/containers/twas_latest.sif &\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Global parameter settings\n",
    "The section outlined the parameters that can be set in the command interface."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "# An index text file with 4 columns specifying the chr, start, end and names of regions to analyze\n",
    "parameter: region_list = path\n",
    "# Path to the work directory of the weight computation: output weights and cache will be saved to this directory.\n",
    "parameter: wd = path('./')# Specify the scanning window for the up and downstream radius to analyze around the region of interest, in units of Kb\n",
    "parameter: window = 500000\n",
    "# Prefix for the output file\n",
    "parameter: prefix = \"geneTpmResidualsAgeGenderAdj_rename\"\n",
    "\n",
    "# Specify the number of jobs per run.\n",
    "parameter: job_size = 2\n",
    "\n",
    "# Container option for software to run the analysis: docker or singularity\n",
    "parameter: container = 'gaow/twas'\n",
    "\n",
    "# Get regions of interest to focus on.\n",
    "regions = [x.strip().split() for x in open(region_list).readlines() if x.strip() and not x.strip().startswith('#')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Partition of the GWAS sumstat for each genes\n",
    "This step extracts the molecular phenotype for each gene and transposes them into the formats needed in the follow-up analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[GWAS_Annotation]\n",
    "# GWAS file, chr and bp column shall be named CHR and BP\n",
    "parameter: GWAS_sumstat = path\n",
    "# Number of column for beta, SE_beta, and SNP names  \n",
    "parameter: beta_col = 5\n",
    "parameter: se_beta_col = 8\n",
    "parameter: snp_col = 9\n",
    "input: GWAS_sumstat, for_each = \"regions\"\n",
    "output: f'{wd}/gwas_sumstat/{prefix}.{_regions[0]}.uni_weight.rds'\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '12h',  mem = '10G', tags = f'{step_name}_{_output:bn}'\n",
    "R: expand= \"$[ ]\", stderr = f'{_output}.stderr', stdout = f'{_output}.stdout',container = container\n",
    "    library(\"dplyr\")\n",
    "    library(\"tibble\")\n",
    "    library(\"readr\")\n",
    "    library(\"modelr\")\n",
    "    library(\"purrr\")\n",
    "    library(\"tidyr\")\n",
    "    sumstat = read_delim(\"$[_input]\",\"\\t\")\n",
    "    sumstat_ftr = sumstat%>%mutate(Z = sumstat[,$[beta_col]]/sumstat[,$[sbeta_col]])%>%\n",
    "    filter(CHR == $[_regions[1]], BP >=  $[_regions[2]] - 500000, BP <= $[_regions[2]] + 500000)%>%\n",
    "    ## remove all the NA,nan,Inf sumstat\n",
    "    filter(!is.na(Z) && !is.nan(Z) && is.finite(Z))\n",
    "    output = list()\n",
    "    output$bhat = as.matrix(sumstat_ftr[,$[beta_col]])\n",
    "    rownames(output$bhat) = sumstat_ftr[,$[snp_col]]%>%unlist%>%as.character\n",
    "    output$sbhat = as.matrix(sumstat_ftr[,$[se_beta_col]])\n",
    "    rownames(output$sbhat) = sumstat_ftr[,$[snp_col]]%>%unlist%>%as.character\n",
    "    ## remove all the NA,nan,Inf sumstat\n",
    "    \n",
    "  \n",
    "    output%>%saveRDS(\"$[_output]\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "Bash",
     "calysto_bash",
     "Bash",
     "#E6EEFF",
     ""
    ],
    [
     "SoS",
     "sos",
     "",
     "",
     "sos"
    ]
   ],
   "version": "0.22.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
