{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# TWAS FUSION analysis\n",
    "\n",
    "This notebook implements a TWAS analysis workflow using the FUSION software.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Aim\n",
    "\n",
    "The aim of this pipeline is to build a prediction model for gene expression using cis-SNPs by obtaining \"weights\" from multiple regression of gene expression levels on SNP genotypes. The weights can then be combined with the GWAS data to test for association between predicted gene expression and GWAS phenotypes. Either GWAS summary statistics or GWAS genotype + phenotype data can be used for the association testing. \n",
    "## Overview\n",
    "\n",
    "__Objective__: \n",
    "    To Compute the association between expression and SNP for TWAS analysis.\n",
    "\n",
    "__Background__:\n",
    "    SNP can modulate the functional phenotypes both directly and by modulating the expression levels of genes. \n",
    "Therefore, the integration of expression measurements and a larger scale GWAS summary association statistics will help identify the genes associated with the targeted complex traits. \n",
    "\n",
    "__Significance__:\n",
    "    By applying this method, new candidate genes whose expression level is significantly associated with complex traits can be used in prediction without actually going through the expensive gene expression measurement process. As a relatively small set of gene expression and genotyping, data can be used to impute the expression for a much larger set of phenotyped individuals from their SNP genotype data. \n",
    "\n",
    "__Method__:\n",
    "    The imputed expression can then be viewed as a linear model of genotypes with _weights based on the correlation between SNPs and gene expression__ in the training data while accounting for linkage disequilibrium (LD) SNPs. We then correlated the imputed gene expression to the trait to perform a transcriptome-wide association study (TWAS) and identify significant expression-trait associations. \n",
    " \n",
    "The weights are computed via various models: blup, bslmm, lasso,top1, and enet. BLUP(best linear unbiased predictors)/bslmm(Bayesian linear mixed model) are conducted using gemma, lasso using plink, and enet (elastic net) using cv.glmnet function in R.\n",
    "\n",
    "Before the weight calculation, the heritability of each gene are computed using GCTA; genes with insignificant heritability were screened out.\n",
    "    \n",
    "## Pre-requisites\n",
    "\n",
    "We provide a container image `docker://gaow/twas` that contains all software needed to run the pipeline. If you would like to configure it by yourself, please make sure you install the following software before running this notebook:\n",
    "- GCTA\n",
    "- PLINK\n",
    "- GEMMA\n",
    "- Modified `Fusion.compute_weights.R` scripts can be found [in this github repo](https://github.com/cumc/neuro-twas/blob/master/Workflow/FUSION.compute_weights.R).\n",
    "- The original `FUSION.assoc_test.R` script can be found [in the author's github repo](https://github.com/gusevlab/fusion_twas).\n",
    "\n",
    "You need to make both `Fusion.compute_weights.R` and `FUSION.assoc_test.R` executable with `Rscript` command. See [this line]() for an example. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Input and Output\n",
    "## Input\n",
    "- `--gwas_sumstat` The GWAS sum stat text file documenting the associations between the SNP and the disease. It shall contain at least four column: the SNP rsID, the effect allele, the other allele, and the Z-score describing the relationship between the SNP and the disease. \n",
    "- `--genotype_list` An index text file with two columns of chromosome and the corresponding PLINK bed file.\n",
    "- `--molecular-pheno`, The text file containing the table describing the molecular phenotype. It shall have regions(genes) as rows and samples as columnes\n",
    "- `--region_list` The text file with 4 columns specifying the #Chr, P0 (Start position), P1(End position) and names of regions to analyze. The name of the column is not important but the order of the columns. It is also important that the column name of the first column starts with a #. The region_list can can be generated by using another sos pipeline SOS_ROSMAP_gene_exp_processing.ipynb.\n",
    "- `--window` the region span from the specify start and end site for the cis-gene. If the gene expression only have one position column, set the window to a large number like 5E5.\n",
    "- `--window` the region span from the specify start and end site for the cis-gene. If the gene expression only have one position column, set the window to a large number like 5E5.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "**FIXME: here are also some text i dug up later in your workflow codes. Please explain them up-front. We would not expect users to read this notebook beyong the \"Working example \" section**\n",
    "This probably is served as a self reminder of what each file do and therefore not expected the user to see, so maybe move them back?\n",
    "\n",
    "```\n",
    "1. The gene expression pheno type, a three column table for each genes, with the first two columns specifing the family ID and within family ID of the samples. In the current case where all samples are unrelated, the first two columns are simply sample ID. The third column is the actual gene expression value.\n",
    "2. The plink trio file for each specific genes, containing only the snps corresponding th the regions whose expression are recorded. In particular, the snp are filtered according to the genetics regions outlined by Position+/-windows.\n",
    "```\n",
    "\n",
    "\n",
    "\n",
    "## Output\n",
    "\n",
    "- `.wgt.Rdat` The actual weight data that are computed.\n",
    "- `.weight_list.txt` The index text file recording information about each of the wgt.Rdat data, including the filename, corresponding region ID, Chromosomes, start and end position, heritability, and the SE and P-value of the heritability.\n",
    "- `.dat` the actual TWAS association of the genes, a detailed description of this output are outlined here:http://gusevlab.org/projects/fusion/\n",
    "\n",
    "\n",
    "# Command interface "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No help information is available for script run: Failed to locate twas_fusion.ipynb.sos\n"
     ]
    }
   ],
   "source": [
    "!sos run twas_fusion.ipynb -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Working example\n",
    "A minimal working example (MWE) dataset that can be downloaded from the private repo:\n",
    "https://github.com/cumc/neuro-twas/blob/master/TWAS_pipeline_MWE%202.zip\n",
    "the genotypes file can be downloaded from the following link:\n",
    "https://data.broadinstitute.org/alkesgroup/FUSION/LDREF.tar.bz2\n",
    "\n",
    "**FIXME: please upload the data to synapse.org and take it out of github. On github we don't store large datasets. Please ask me about account information for synapse.org**\n",
    "\n",
    "The time it take to run this MWE shall be around 2 minutes. Pay extra attention to the gene_start and gene_end position  when using following command on gene_exp file that are not this MWE. Also, when there is too few or too many genes that passed the heritability check, consider increasing or decreasing the --window options. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "## Test pipeline with test data\n",
    "## Switch back to abosolute path, otherwise there will be file not found error in step 5\n",
    "sos run twas_fusion.ipynb association_test \\\n",
    "  --gwas_sumstat ./data/sum_stat.sumstats \\\n",
    "  --molecular-pheno ./data/gene_exp_file.txt \\\n",
    "  --wd ./ \\\n",
    "  --genotype_list ./data/ld_ref_abs \\\n",
    "  --region_list ./data/region_list.txt \\\n",
    "  --region_name 1 \\\n",
    "  --data_start 5 \\\n",
    "  --window 500000 \\\n",
    "  --model blup lasso top1 enet \\\n",
    "  --container gaow/twas \n",
    "\n",
    "## Association test only \n",
    "sos run twas_fusion.ipynb association_test \\\n",
    "  --gwas_sumstat /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/Project/Updated/data/sum_stat.sumstats \\\n",
    "  --molecular-pheno /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/Project/Updated/data/gene_exp_file.txt \\\n",
    "  --wd /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/Project/Updated/Working_at \\\n",
    "  --genotype_list /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/Project/Updated/data/ld_ref \\\n",
    "  --region_list /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/Project/Updated/data/region_list.txt \\\n",
    "  --region_name 1 \\\n",
    "  --data_start 5 \\\n",
    "  --output_path /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/Project/Updated/At_output \\\n",
    "  --window 500000 \\\n",
    "  --model blup lasso top1 enet \\\n",
    "  --container gaow/twas"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Global parameter settings\n",
    "The section outlined the parameters that can be set in the command interface."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "# Path to the input molecular phenotype data.\n",
    "parameter: molecular_pheno = path\n",
    "# An index text file with two columns of chromosome and the corresponding PLINK bed file.\n",
    "parameter: genotype_list = path\n",
    "# An index text file with 4 columns specifying the chr, start, end and names of regions to analyze\n",
    "parameter: region_list = path\n",
    "# Path to the work directory of the weight computation: output weights and cache will be saved to this directory.\n",
    "parameter: wd = path('./')\n",
    "# Specify the directory to save fitted weights\n",
    "parameter: weights_path = f'{wd:a}/WEIGHTS'\n",
    "# Path to list of weights\n",
    "parameter: weights_list = f'{weights_path}/{molecular_pheno:bn}.weights_list.txt'\n",
    "# Path to store the output folder\n",
    "parameter: output_path = f'{wd:a}/result'\n",
    "# Specify the column in the molecular_pheno file that contains the name of the region of interest\n",
    "parameter: region_name = int\n",
    "# Specify the column in the molecular_pheno file where the actual data start\n",
    "parameter: data_start = int\n",
    "# Specify the scanning window for the up and downstream radius to analyze around the region of interest, in units of Kb\n",
    "parameter: window = 50000\n",
    "# Specify the number of jobs per run.\n",
    "parameter: job_size = 2\n",
    "# Specify the minumum p value for heratibility for the study to continue\n",
    "parameter: hsq_p = 0.01\n",
    "\n",
    "\n",
    "\n",
    "# Specify what model are used to compute weights.\n",
    "# Notice that `blsmm` can be very resource consuming.\n",
    "parameter: model = ['bslmm', 'blup' , 'lasso', 'top1', 'enet']\n",
    "# Container option for software to run the analysis: docker or singularity\n",
    "parameter: container = 'gaow/twas'\n",
    "# Get regions of interest to focus on.\n",
    "regions = [x.strip().split() for x in open(region_list).readlines() if x.strip() and not x.strip().startswith('#')]\n",
    "\n",
    "geno_inventory = dict([x.strip().split() for x in open(genotype_list).readlines() if x.strip() and not x.strip().startswith('#')])\n",
    "\n",
    "import os\n",
    "def get_genotype_file(chrom, genotype_list, geno_inventory):\n",
    "    chrom = f'{chrom}'\n",
    "    if chrom.startswith('chr'):\n",
    "        chrom = chrom[3:]\n",
    "    if chrom not in geno_inventory:\n",
    "        geno_file = f'{chrom}'\n",
    "    else:\n",
    "        geno_file = geno_inventory[chrom]\n",
    "    if not os.path.isfile(geno_file):\n",
    "        # relative path\n",
    "        if not os.path.isfile(f'{genotype_list:ad}/' + geno_file):\n",
    "            raise ValueError(f\"Cannot find genotype file {geno_file}\")\n",
    "        else:\n",
    "            geno_file = f'{genotype_list:ad}/' + geno_file\n",
    "    return path(geno_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Partition of the molecular phenotype for each genes\n",
    "This step extracts the molecular phenotype for each gene and transposes them into the formats needed in the follow-up analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[twas_fusion_1,compute_wgt_1,susie_1]\n",
    "\n",
    "input: molecular_pheno, for_each = \"regions\"\n",
    "output: f'{wd:a}/cache/{_input:bn}.{_regions[3]}.exp',\n",
    "        f'{wd:a}/cache/{_input:bn}.{_regions[3]}.pheno'\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '4h',  mem = '6G', tags = f'{step_name}_{_output[0]:bn}'\n",
    "R: expand = \"$[ ]\", stderr = f'{_output[0]}.stderr', stdout = f'{_output[0]}.stdout'\n",
    "    # Get the line number for the region in the file\n",
    "    line_num = system(\"awk '($$[region_name]==\\\"$[_regions[3]]\\\") {print NR}' $[_input]\", intern=T)\n",
    "    if (length(line_num) == 0){\n",
    "      stop( \"Cannot find $[_regions[3]] in column $[region_name]  $[_input]\")}\n",
    "    yi <- data.table::fread(file = $[_input:r], skip = as.integer(line_num) - 1, nrows = 1)\n",
    "    samplenames_yi <- data.table::fread(file = $[_input:r], skip = 0, nrows = 1)\n",
    "    colnames(yi) <- colnames(samplenames_yi)\n",
    "    readr::write_tsv(yi, path = \"$[_output[0]]\", na = \"NA\", append = FALSE, col_names = TRUE, quote_escape = \"double\")\n",
    "    yi <- as.data.frame(yi[, $[data_start]:ncol(yi), drop = FALSE])\n",
    "    yj <- rbind(colnames(yi),colnames(yi),yi)\n",
    "    readr::write_tsv(as.data.frame(t(yj)), path = \"$[_output[1]]\", na = \"NA\", append = FALSE, col_names = TRUE, quote_escape = \"double\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Construction of Plink trio for each gene\n",
    "This step constructs the plink file for each gene based on the output of previous steps. Specifically it:\n",
    "\n",
    "1. Selects only the SNPs within the start and end position of the corresponding region (gene)\n",
    "2. Replaces the Phenotype value (last column) of the .fam based on the input\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[twas_fusion_2,compute_wgt_2,susie_2]\n",
    "input: group_by = 2, group_with = 'regions'\n",
    "output: f'{_input[0]:n}.bed',\n",
    "        f'{_input[0]:n}.bim',\n",
    "        f'{_input[0]:n}.fam'\n",
    "\n",
    "# look up for genotype file\n",
    "geno_file = get_genotype_file(_regions[0],genotype_list,geno_inventory)\n",
    "\n",
    "parameter: extract_snp = f'{geno_file:an}.bim'\n",
    "parameter: exclude_snp = \"./.\"\n",
    "parameter: keep_sample = f'{_input[1]}'\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '12h',  mem = '6G', tags = f'{step_name}_{_output[0]:bn}'\n",
    "bash: expand= \"$[ ]\", stderr = f'{_output[0]}.stderr', stdout = f'{_output[0]}.stdout', container = container, volumes = [f'{geno_file:ad}:{geno_file:ad}']\n",
    "    ##### Get the locus genotypes for $[_regions[3]]\n",
    "    plink --bfile $[geno_file:an] \\\n",
    "    --pheno $[_input[1]] \\\n",
    "    --make-bed \\\n",
    "    --out $[_output[0]:n] \\\n",
    "    --chr $[_regions[0]] \\\n",
    "    --from-bp $[int(_regions[1]) - window ] \\\n",
    "    --to-bp $[int(_regions[2]) + window ] \\\n",
    "    --keep $[keep_sample] \\\n",
    "    --extract $[extract_snp]\n",
    "    --exclude $[exclude_snp]\n",
    "    --allow-no-sex || true\n",
    "    touch $[_output]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Computation of Weight\n",
    "This step computes the association between the phenotype and each of the SNPs. Specifically it:\n",
    "1. Estimates and documents the heritability (Ratio of genetics variance and phenotypic variance) of each region (gene)\n",
    "\n",
    "2. for those regions with significant heritability\n",
    "    1. Compute the association with the model specified for each SNPs\n",
    "    2. Stored the output in a wgt.RData for each region.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "#Actual weight computation \n",
    "[twas_fusion_3,compute_wgt_3]\n",
    "input: group_with = 'regions'\n",
    "output: f'{weights_path}/{_input[0]:bn}.wgt.RDat'\n",
    "import os\n",
    "skip_if(os.path.getsize(f'{_input[0]}') == 0)\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '4h',  mem = '6G', tags = f'{step_name}_{_output[0]:bn}'\n",
    "bash: expand= \"$[ ]\",stderr = f'{_output[0]:nn}.stderr', stdout = f'{_output[0]:nn}.stdout', container = container\n",
    "    FUSION.compute_weights.R \\\n",
    "    --bfile $[_input[0]:n] \\\n",
    "    --tmp $[_input[0]:n].tmp \\\n",
    "    --out $[_output[0]:nn] \\\n",
    "    --verbose 0 \\\n",
    "    --save_hsq \\\n",
    "    --PATH_gcta `which gcta64` \\\n",
    "    --PATH_gemma `which gemma` \\\n",
    "    --models \"$[\",\".join(model)]\"  \\\n",
    "    --hsq_p $[hsq_p]\n",
    "    ## Creat dummy output file that will be deleted next step\n",
    "    touch $[_output]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Creation of weight index\n",
    "This step creates an index documenting the information about each of the weight data, including the filename, corresponding region ID, Chromosomes, start and end position, heritability, and the SE and P-value of the heritability."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[twas_fusion_4,compute_wgt_4]\n",
    "input: group_by = \"all\"\n",
    "output: weights_list\n",
    "import os\n",
    "weight_files = [x for x in _input if not os.path.getsize(x) == 0]\n",
    "regions_dict = dict([(x[3], (x[0], x[1], x[2])) for x in regions])\n",
    "res = [[\"WGT\", \"ID\",\"CHR\",\"P0\",\"P1\",\"Heritability\",\"Heritability_SE\",\"Heritability_LRT_P_val\"]]\n",
    "for item in weight_files:\n",
    "    name = f'{item:bnn}'.lstrip(f'.{molecular_pheno:bn}')\n",
    "    hsq = open(f\"{item:nn}.hsq\").read().strip().split()\n",
    "    res.append([str(item),name,regions_dict[name][0], regions_dict[name][1], regions_dict[name][2], hsq[1], hsq[2], hsq[3]])\n",
    "with open(_output, 'w') as f:\n",
    "    f.write('\\n'.join(['\\t'.join(x) for x in res]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Conducting of association test\n",
    "This step conduct the association test by modifying SNPs' association and traits of interests with the association between SNPs and the molecular phenotypes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Association test\n",
    "[twas_fusion_5, association_test]\n",
    "depends: weights_list\n",
    "chrom_list = list(set(get_output(f'cut -f 3 {weights_list} | tail -n+2').strip().split(\"\\n\")))\n",
    "# Path to GWAS summary statistics data (association results between SNP and disease in a GWAS)\n",
    "parameter: gwas_sumstat = path\n",
    "parameter: asso_mem = \"60G\"\n",
    "input: gwas_sumstat, for_each = \"chrom_list\"\n",
    "output:f'{output_path}/{gwas_sumstat:bn}_chr{_chrom_list}.twas.txt'\n",
    "\n",
    "geno_file = get_genotype_file(_chrom_list, genotype_list, geno_inventory)\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '4h',  mem = asso_mem, tags = f'{step_name}_{_output[0]:bn}'\n",
    "bash: expand= \"$[ ]\", stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout', container = container, volumes = [f'{geno_file:ad}:{geno_file:ad}']\n",
    "    FUSION.assoc_test.R \\\n",
    "    --sumstats $[_input] \\\n",
    "    --weights $[weights_list] \\\n",
    "    --weights_dir / \\\n",
    "    --ref_ld_chr $[geno_file:an] \\\n",
    "    --chr $[_chrom_list] \\\n",
    "    --out $[_output]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Documentation of dummy file\n",
    "For the pipeline to run smoothly, dummy files were created in previous steps. This step finds all the dummy files and records them into a separate file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "# Clean up dummy file\n",
    "[twas_fusion_6,compute_wgt_5]\n",
    "input: group_by = \"all\"\n",
    "output: f'{wd:a}/error_no_plink.log',\n",
    "        f'{wd:a}/error_no_wgt_computed.log',\n",
    "        f'{weights_path}/all_hsq.txt',\n",
    "        f'{output_path}/all_result.txt'\n",
    "bash: expand= \"$[ ]\"\n",
    "    find $[wd:a]/cache/*.bim -size 0 -print > $[_output[0]]\n",
    "    find $[weights_path]/*.wgt.RDat -size 0 -print > $[_output[1]]\n",
    "    echo filename heritability h_SE pval >  $[_output[2]]\n",
    "    cat $[weights_path]/*hsq >>  $[_output[2]]\n",
    "    cat $[output_path]/*twas.txt > $[_output[3]]\n",
    "    touch $[_output[3]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Restarting kernel \"Bash\"\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/GIT/freshcopy/neuro-twas/Workflow\n",
      "\n"
     ]
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Conducting of susieR test\n",
    "This step use the susieR packages to recompute the heritability and the weight of the result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "#Susie tesst\n",
    "[susie_3]\n",
    "input: group_with = 'regions'\n",
    "output: f'{wd:a}/susie/{_input[0]:bn}.susie.hsq',\n",
    "        f'{wd:a}/susie/{_input[0]:bn}.wgt.txt'\n",
    "parameter: causal_variables_L = 10\n",
    "parameter: scaled_prior_variance = 0.1\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '4h',  mem = \"6G\" , tags = f'{step_name}_{_output[0]:bn}'\n",
    "R: expand= \"$[ ]\", stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout', container = container, volumes = [f'{wd:a}:{wd:a}']\n",
    "  library(\"dplyr\")\n",
    "  library(\"tibble\")\n",
    "  library(\"susieR\")\n",
    "  library(\"plink2R\")\n",
    "  genos = read_plink(_input[0]:n)\n",
    "  \n",
    "  # hsq for removing all the na snps\n",
    "  X = genos$bed%>%t()%>%na.omit()%>%t()%>%scale()\n",
    "  Y = geno$fam%>%as.tibble()%>%mutate(name = paste(V1,\":\",V2,sep = \"\"))\n",
    "  fitted1 = susie(X, Y$V6,\n",
    "                L = $[causal_variables_L],\n",
    "                estimate_residual_variance = TRUE, \n",
    "                estimate_prior_variance = FALSE,\n",
    "                scaled_prior_variance = $[scaled_prior_variance])\n",
    "  H1 = var(predict(fitted1)) / var(Y$V6)\n",
    "  # hsq for removing all the na samples\n",
    "  X2 = genos$bed%>%na.omit()%>%scale()\n",
    "  Y2 = Y%>%filter(name%in%rownames(X2))\n",
    "  fitted2 = susie(X2, Y2$V6,\n",
    "                L = $[causal_variables_L],\n",
    "                estimate_residual_variance = TRUE, \n",
    "                estimate_prior_variance = FALSE,\n",
    "                scaled_prior_variance = $[scaled_prior_variance])\n",
    "  H2 = var(predict(fitted2)) / var(Y$V6)\n",
    "  # hsq for removing all the na samples and na SNPs\n",
    "  X3 = X[rownames(X) == Y$name]\n",
    "  fitted3 = susie(X3, Y2$V6,\n",
    "                L = $[causal_variables_L],\n",
    "                estimate_residual_variance = TRUE, \n",
    "                estimate_prior_variance = FALSE,\n",
    "                scaled_prior_variance = $[scaled_prior_variance])\n",
    "  H3 = var(predict(fitted3)) / var(Y$V6)\n",
    "  # hsq for the best performing snp using naive way, excluding only the na SNP\n",
    "  sumstats <- univariate_regression(X, Y$V6)\n",
    "  z_scores <- sumstats$betahat / sumstats$sebetahat\n",
    "  top_snp = which(abs(z_scores) == max(abs(z_scores)))\n",
    "  HH = var(X[,top_snp]*sumstats$betahat[top_snp])/var(Y$V6)\n",
    "  # storing all the HSQ\n",
    "  hsq_tbl = tibble(hsq_full_sample = H1, hsq_full_snp = H2, hsq_neither = H3, hsq_naive = H4)\n",
    "  hsq_tbl%>%write_delim($[_output[0]])\n",
    "  # storing the wgt computed from susie, excluding only the na SNP\n",
    "  susie_wgt = tibble(susie = coef(fitted)[2:length(coef(fitted))])\n",
    "  susie_wgt%>%write_delim($[_output[1]])\n",
    "  \n",
    "#Create all_hsq for susie\n",
    "[susie_4]\n",
    "input: group_by = \"all\"\n",
    "output:f'{wd:a}/susie/all_hsq.txt'\n",
    "bash: expand= \"$[ ]\"\n",
    "  head -1 $[_input[0]] > $[_output]\n",
    "  cat $[wd:a]/susie/*.hsq | grep -v hsq_full_sample >> $[_output]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "SoS",
     "sos",
     "",
     ""
    ]
   ],
   "version": "0.20.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
