{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Apex Prototyping\n",
    "This notebook Document a prototype workflow for using the Apex to conduct analysis on a list of genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Pre-requisites\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Input and Output\n",
    "## Input\n",
    "\n",
    "- `--genotype_list` An index text file with two columns of chromosome and the corresponding PLINK bed file.\n",
    "- `--molecular-pheno`, The text file containing the table describing the molecular phenotype. It shall have regions(genes) as rows and samples as columnes\n",
    "- `--region_list` The text file with 4 columns specifying the #Chr, P0 (Start position), P1(End position) and names of regions to analyze. The name of the column is not important but the order of the columns. It is also important that the column name of the first column starts with a #. The region_list can can be generated by using another sos pipeline SOS_ROSMAP_gene_exp_processing.ipynb.\n",
    "\n",
    "## Output\n",
    "\n",
    "- `uni_weight.RDS` a RDS file that served as the input for the mixture pipeline.\n",
    "- `susie.RData` a R object containing all the susie output for each of the regions\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Command interface "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Working example\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[91mERROR\u001b[0m: \u001b[91mFailed to locate twas_fusion.ipynb.sos\u001b[0m\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## Test pipeline with test data\n",
    "## Switch back to abosolute path, otherwise there will be file not found error in step 5\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Global parameter settings\n",
    "The section outlined the parameters that can be set in the command interface."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "# Path to the input molecular phenotype data.\n",
    "parameter: molecular_pheno = path\n",
    "# Covariate file, in similar format as the molecular_pheno\n",
    "parameter: cov = path\n",
    "# Genotype file in vcf/vcf.gz/bcf format\n",
    "parameter: genotype = path\n",
    "# An index text file with 4 columns specifying the chr, start, end and names of regions to analyze\n",
    "parameter: region_list = path\n",
    "# Path to the work directory of the weight computation: output weights and cache will be saved to this directory.\n",
    "parameter: wd = path('./')\n",
    "# Path to store the output folder\n",
    "parameter: output_path = f'{wd:a}/result'\n",
    "# Specify the number of jobs per run.\n",
    "parameter: job_size = 2\n",
    "# Container option for software to run the analysis: docker or singularity\n",
    "parameter: container = 'gaow/twas'\n",
    "# Prefix for the analysis output\n",
    "parameter: Prefix = \"\"\n",
    "\n",
    "# Get a list of all regions\n",
    "regions = [x.strip().split() for x in open(region_list).readlines() if x.strip() and not x.strip().startswith('#')]\n",
    "# Get a list of all chrs\n",
    "chrm = set(regions[0])\n",
    "chrm = list(chrm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Fitered for the genes that are to be analyzed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[Expression_Filtering]\n",
    "input: molecular_pheno,region_list\n",
    "output: f'{wd:a}/cache/{_input:bn}.bed'\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '4h',  mem = '20G', tags = f'{step_name}_{_output[0]:bn}'\n",
    "R: expand = \"$[ ]\", stderr = f'{_output[0]}.stderr', stdout = f'{_output[0]}.stdout',container = container\n",
    "    library(\"dplyr\")\n",
    "    library(\"tibble\")\n",
    "    library(\"readr\")\n",
    "    library(\"modelr\")\n",
    "    library(\"purrr\")\n",
    "    pheno = read_delim(\"$[molecular_pheno]\",delim = \"\\t\")\n",
    "    region = read_delim(\"$[region_list]\",delim = \"\\t\")\n",
    "    output = inner_join(pheno, region, by = \"gene_ID\")\n",
    "    output%>%write_delim(\"$[_output[0]]\",delim = \"\\t\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Factor Analysis\n",
    "This step infer hidden covariates and produce residuel expression for downstream analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[cis_qtl_1, Factor_Analysis]\n",
    "output: f'{wd:a}/cache/factor/{_input:bn}.bed'\n",
    "\n",
    "# Number of latent common factors\n",
    "parameter: n_of_factor = 5\n",
    "# Number of factor analysis iterations (0 for PCA).                                         \n",
    "parameter: iteration = 0\n",
    "# Factor analysis prior\n",
    "parameter: p = 0.05\n",
    "parameter: tau = 0.05\n",
    "# Parameter for running\n",
    "parameter: thread = 3\n",
    "        \n",
    "        \n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = '12h',  mem = '20G', tags = f'{step_name}_{_output[0]:bn}'\n",
    "bash: expand= \"$[ ]\", stderr = f'{_output[0]}.stderr', stdout = f'{_output[0]}.stdout', container = container, volumes = [f'{geno_file:ad}:{geno_file:ad}']\n",
    "    cd $[wd]\n",
    "    mkdir factor\n",
    "    cd factor\n",
    "    ##### Get the locus genotypes for $[_regions[3]]\n",
    "    Apex factor --vcf $[genotype:an] \\\n",
    "    --bed $[_input] \\\n",
    "    --cov $[cov] \\\n",
    "    --out $[Prefix] \\\n",
    "    --threads $[threads] \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## LMM Regression  \n",
    "This step are done to precompute and store a) LMM null models and trait residuals and b) spline terms for LMM genotypic variances to speed up downstream analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[cis_qtl_2,LMM]\n",
    "output: f'{wd:a}/cache/lmm/{_input:bn}.bed'\n",
    "parameter: window = '[100000]'\n",
    "bash: expand= \"$[ ]\", stderr = f'{_output[0]}.stderr', stdout = f'{_output[0]}.stdout', container = container, volumes = [f'{geno_file:ad}:{geno_file:ad}']\n",
    "    cd $[wd]/cache/\n",
    "    mkdir lmm\n",
    "    cd lmm\n",
    "    Apex lmm --vcf $[genotype:an] \\\n",
    "    --bed $[_input] \\\n",
    "    --cov $[cov] \\\n",
    "    --out $[Prefix] \\\n",
    "    --threads $[threads] \\\n",
    "    --fit-null \\\n",
    "    --save-resid \\\n",
    "    --write-gvar \\\n",
    "    --window=$[window]\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## QTL Sumstat generation  \n",
    "This step generate the cis-QTL summary statistics and vcov (covariate-adjusted LD) files for downstream analysis from summary statistics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[cis_qtl_3,cis_sumstat]\n",
    "input: ,for_each = \"chrm\"\n",
    "output: f'{wd:a}/cache/sumstat/{_input:bn}.bed'\n",
    "parameter: window = '[100000]'\n",
    "bash: expand= \"$[ ]\", stderr = f'{_output[0]}.stderr', stdout = f'{_output[0]}.stdout', container = container, volumes = [f'{geno_file:ad}:{geno_file:ad}']\n",
    "    cd $[wd]/cache/\n",
    "    mkdir lmm\n",
    "    cd lmm\n",
    "    Apex cis --vcf $[genotype:an] \\\n",
    "    --bed $[_input] \\\n",
    "    --cov $[cov] \\\n",
    "    --out $[Prefix] \\\n",
    "    --threads $[threads] \\\n",
    "    --fit-null \\\n",
    "    --save-resid \\\n",
    "    --write-gvar \\\n",
    "    --window=$[window]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Conducting meta analysis for multi-variant(TBD) \n",
    "Taking the input from "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "Bash",
     "calysto_bash",
     "Bash",
     "#E6EEFF",
     ""
    ],
    [
     "SoS",
     "sos",
     "",
     "",
     "sos"
    ]
   ],
   "version": "0.20.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
