{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Interface with exterior"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ==== TODO\n",
    "# * Make sure BLUP/BSLMM weights are being scaled properly based on MAF\n",
    "\n",
    "suppressMessages(library(\"optparse\"))\n",
    "suppressMessages(library('plink2R'))\n",
    "suppressMessages(library('glmnet'))\n",
    "suppressMessages(library('methods'))\n",
    "\n",
    "option_list = list(\n",
    "  make_option(\"--bfile\", action=\"store\", default=NA, type='character',\n",
    "              help=\"Path to PLINK binary input file prefix (minus bed/bim/fam) [required]\"),\n",
    "  make_option(\"--out\", action=\"store\", default=NA, type='character',\n",
    "              help=\"Path to output files [required]\"),\n",
    "  make_option(\"--tmp\", action=\"store\", default=NA, type='character',\n",
    "              help=\"Path to temporary files [required]\"),\n",
    "  make_option(\"--pheno\", action=\"store\", default=NA, type='character',\n",
    "              help=\"Path to molecular phenotype file (PLINK format) [optional, taken from bfile otherwise]\"),\n",
    "  make_option(\"--PATH_plink\", action=\"store\", default=\"plink\", type='character',\n",
    "              help=\"Path to plink executable [%default]\"),\n",
    "  make_option(\"--PATH_gcta\", action=\"store\", default=\"gcta_nr_robust\", type='character',\n",
    "              help=\"Path to plink executable [%default]\"),\n",
    "  make_option(\"--PATH_gemma\", action=\"store\", default=\"gemma\", type='character',\n",
    "              help=\"Path to plink executable [%default]\"),\n",
    "  make_option(\"--covar\", action=\"store\", default=NA, type='character',\n",
    "              help=\"Path to quantitative covariates (PLINK format) [optional]\"),\n",
    "  make_option(\"--resid\", action=\"store_true\", default=FALSE,\n",
    "              help=\"Also regress the covariates out of the genotypes [default: %default]\"),              \n",
    "  make_option(\"--hsq_p\", action=\"store\", default=0.01, type='double',\n",
    "              help=\"Minimum heritability p-value for which to compute weights [default: %default]\"),\n",
    "  make_option(\"--hsq_set\", action=\"store\", default=NA, type='double',\n",
    "              help=\"Skip heritability estimation and set hsq estimate to this value [optional]\"),\n",
    "  make_option(\"--crossval\", action=\"store\", default=5, type='double',\n",
    "              help=\"How many folds of cross-validation, 0 to skip [default: %default]\"),\n",
    "  make_option(\"--verbose\", action=\"store\", default=1, type=\"integer\",\n",
    "              help=\"How much chatter to print: 0=nothing; 1=minimal; 2=all [default: %default]\"),\n",
    "  make_option(\"--noclean\", action=\"store_true\", default=FALSE,\n",
    "              help=\"Do not delete any temporary files (for debugging) [default: %default]\"),\n",
    "  make_option(\"--rn\", action=\"store_true\", default=FALSE,\n",
    "              help=\"Rank-normalize the phenotype after all QC: [default: %default]\"),\n",
    "  make_option(\"--save_hsq\", action=\"store_true\", default=FALSE,\n",
    "              help=\"Save heritability results even if weights are not computed [default: %default]\"),\t\t\t  \n",
    "  make_option(\"--models\", action=\"store\", default=\"blup,lasso,top1,enet\", type='character',\n",
    "              help=\"Comma-separated list of prediction models [default: %default]\\n\n",
    "\t\t\t\t\tAvailable models:\\n\n",
    "\t\t\t\t\ttop1:\\tTop eQTL (standard marginal eQTL Z-scores always computed and stored)\\n\n",
    "\t\t\t\t\tblup:\\t Best Unbiased Linear Predictor (dual of ridge regression)\\n\n",
    "\t\t\t\t\tbslmm:\\t Bayesian Sparse Linear Model (spike/slab MCMC)\\n\n",
    "\t\t\t\t\tlasso:\\t LASSO regression (with heritability used as lambda)\\n\n",
    "\t\t\t\t\tenet:\\t Elastic-net regression (with mixing parameter of 0.5)\\n\")\t\t\t  \n",
    "\t\t  \n",
    ")\n",
    "\n",
    "opt = parse_args(OptionParser(option_list=option_list))\n",
    "models = unique( c(unlist(strsplit(opt$models,',')),\"top1\") )\n",
    "M = length(models)\n",
    "\n",
    "if ( opt$verbose == 2 ) {\n",
    "  SYS_PRINT = F\n",
    "} else {\n",
    "  SYS_PRINT = T\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Function Definition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- PREDICTION MODELS\n",
    "\n",
    "# BSLMM\n",
    "weights.bslmm = function( input , bv_type , snp , out=NA ) {\n",
    "\tif ( is.na(out) ) out = paste(input,\".BSLMM\",sep='')\n",
    "\n",
    "\targ = paste( opt$PATH_gemma , \" -miss 1 -maf 0 -r2 1 -rpace 1000 -wpace 1000 -bfile \" , input , \" -bslmm \" , bv_type , \" -o \" , out , sep='' )\n",
    "\tsystem( arg , ignore.stdout=SYS_PRINT,ignore.stderr=SYS_PRINT)\n",
    "\teff = read.table( paste(out,\".param.txt\",sep=''),head=T,as.is=T)\n",
    "\teff.wgt = rep(NA,length(snp))\n",
    "\tm = match( snp , eff$rs )\n",
    "\tm.keep = !is.na(m)\n",
    "\tm = m[m.keep]\n",
    "\teff.wgt[m.keep] = (eff$alpha + eff$beta * eff$gamma)[m]\n",
    "\treturn( eff.wgt )\n",
    "}\n",
    "\n",
    "# PLINK: LASSO\n",
    "weights.lasso = function( input , hsq , snp , out=NA ) {\n",
    "\tif ( is.na(out) ) out = paste(input,\".LASSO\",sep='')\n",
    "\n",
    "\targ = paste( opt$PATH_plink , \" --allow-no-sex --bfile \" , input , \" --lasso \" , hsq , \" --out \" , out , sep='' )\n",
    "\tsystem( arg , ignore.stdout=SYS_PRINT,ignore.stderr=SYS_PRINT )\n",
    "\tif ( !file.exists(paste(out,\".lasso\",sep='')) ) {\n",
    "\tcat( paste(out,\".lasso\",sep='') , \" LASSO output did not exist\\n\" )\n",
    "\teff.wgt = rep(NA,length(snp))\n",
    "\t} else {\n",
    "\teff = read.table( paste(out,\".lasso\",sep=''),head=T,as.is=T)\n",
    "\teff.wgt = rep(0,length(snp))\n",
    "\tm = match( snp , eff$SNP )\n",
    "\tm.keep = !is.na(m)\t\n",
    "\tm = m[m.keep]\n",
    "\teff.wgt[m.keep] = eff$EFFECT[m]\n",
    "\t}\n",
    "\treturn( eff.wgt )\n",
    "}\n",
    "\n",
    "# Marginal Z-scores (used for top1)\n",
    "weights.marginal = function( genos , pheno , beta=F ) {\n",
    "\tif ( beta ) eff.wgt = t( genos ) %*% (pheno) / ( nrow(pheno) - 1)\n",
    "\telse eff.wgt = t( genos ) %*% (pheno) / sqrt( nrow(pheno) - 1 )\n",
    "\treturn( eff.wgt )\n",
    "}\n",
    "\n",
    "# Elastic Net\n",
    "weights.enet = function( genos , pheno , alpha=0.5 ) {\n",
    "\teff.wgt = matrix( 0 , ncol=1 , nrow=ncol(genos) )\n",
    "\t# remove monomorphics\n",
    "\tsds = apply( genos  , 2 , sd )\n",
    "\tkeep = sds != 0 & !is.na(sds)\n",
    "\tenet = cv.glmnet( x=genos[,keep] , y=pheno , alpha=alpha , nfold=5 , intercept=T , standardize=F )\n",
    "\teff.wgt[ keep ] = coef( enet , s = \"lambda.min\")[2:(sum(keep)+1)]\n",
    "\treturn( eff.wgt )\n",
    "}\n",
    "\n",
    "# --- CLEANUP\n",
    "cleanup = function() {\n",
    "\tif ( ! opt$noclean ) {\n",
    "\t\targ = paste(\"rm -f \" , opt$tmp , \"*\", sep='')\n",
    "\t\tsystem(arg)\n",
    "\t}\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check file and function intergrity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform i/o checks here:\n",
    "files = paste(opt$bfile,c(\".bed\",\".bim\",\".fam\"),sep='')\n",
    "if ( !is.na(opt$pheno) ) files = c(files,opt$pheno)\n",
    "if ( !is.na(opt$covar) ) files = c(files,opt$covar)\n",
    "\n",
    "for ( f in files ) {\n",
    "\tif ( !file.exists(f) ){\n",
    "\t\tcat( \"ERROR: \", f , \" input file does not exist\\n\" , sep='', file=stderr() )\n",
    "\t\tcleanup()\n",
    "\t\tq()\n",
    "\t}\n",
    "}\n",
    "\n",
    "if ( system( paste(opt$PATH_plink,\"--help\") , ignore.stdout=T,ignore.stderr=T ) != 0 ) {\n",
    "\tcat( \"ERROR: plink could not be executed, set with --PATH_plink\\n\" , sep='', file=stderr() )\n",
    "\tcleanup()\n",
    "\tq()\n",
    "}\n",
    "\n",
    "if ( !is.na(opt$hsq_set) && system( opt$PATH_gcta , ignore.stdout=T,ignore.stderr=T ) != 0 ){\n",
    "\tcat( \"ERROR: gcta could not be executed, set with --PATH_gcta\\n\" , sep='', file=stderr() )\n",
    "\tcleanup()\n",
    "\tq()\n",
    "}\n",
    "\n",
    "if ( sum(models==\"bslmm\" | models==\"blup\") != 0 && system( paste(opt$PATH_gemma,\"-h\") , ignore.stdout=T,ignore.stderr=T ) != 0 ){\n",
    "\tcat( \"ERROR: gemma could not be executed, set with --PATH_gemma or remove 'bslmm' and 'blup' from models\\n\" , sep='', file=stderr() )\n",
    "\tcleanup()\n",
    "\tq()\n",
    "}\n",
    "\n",
    "# ---\n",
    "\n",
    "fam = read.table(paste(opt$bfile,\".fam\",sep=''),as.is=T)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Intermidiate steps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make/fetch the phenotype file\n",
    "if ( !is.na(opt$pheno) ) {\n",
    "\tpheno.file = opt$pheno\n",
    "\tpheno = read.table(pheno.file,as.is=T)\n",
    "\t# Match up data\n",
    "\tm = match( paste(fam[,1],fam[,2]) , paste(pheno[,1],pheno[,2]) )\n",
    "\tm.keep = !is.na(m)\n",
    "\tfam = fam[m.keep,]\n",
    "\tm = m[m.keep]\n",
    "\tpheno = pheno[m,]\n",
    "} else {\n",
    "\tpheno.file = paste(opt$tmp,\".pheno\",sep='')\n",
    "\tpheno = fam[,c(1,2,6)]\n",
    "\twrite.table(pheno,quote=F,row.names=F,col.names=F,file=pheno.file)\n",
    "}\n",
    "\n",
    "if ( opt$rn ) {\n",
    "\tlibrary('GenABEL')\n",
    "\tlibrary(preprocessCore)\n",
    "\tpheno[,3] = rntransform( pheno[,3] )\n",
    "\twrite.table(pheno,quote=F,row.names=F,col.names=F,file=pheno.file)\n",
    "}\n",
    "\n",
    "# Load in the covariates if needed\n",
    "if ( !is.na(opt$covar) ) {\n",
    "\tcovar = ( read.table(opt$covar,as.is=T,head=T) )\n",
    "\tif ( opt$verbose >= 1 ) cat( \"Loaded\",ncol(covar)-2,\"covariates\\n\")\n",
    "\t# Match up data\n",
    "\tm = match( paste(fam[,1],fam[,2]) , paste(covar[,1],covar[,2]) )\n",
    "\tm.keep = !is.na(m)\n",
    "\tfam = fam[m.keep,]\n",
    "\tpheno = pheno[m.keep,]\n",
    "\tm = m[m.keep]\n",
    "\tcovar = covar[m,]\n",
    "\treg = summary(lm( pheno[,3] ~ as.matrix(covar[,3:ncol(covar)]) ))\n",
    "\tif ( opt$verbose >= 1 ) cat( reg$r.sq , \"variance in phenotype explained by covariates\\n\" )\n",
    "\tpheno[,3] = scale(reg$resid)\n",
    "\traw.pheno.file = pheno.file\n",
    "\tpheno.file = paste(pheno.file,\".resid\",sep='')\n",
    "\twrite.table(pheno,quote=F,row.names=F,col.names=F,file=pheno.file)\n",
    "}\n",
    "\n",
    "geno.file = opt$tmp\n",
    "\n",
    "# recode to the intersection of samples and new phenotype\n",
    "arg = paste( opt$PATH_plink ,\" --allow-no-sex --bfile \",opt$bfile,\" --pheno \",pheno.file,\" --keep \",pheno.file,\" --make-bed --out \",geno.file,sep='')\n",
    "system(arg , ignore.stdout=SYS_PRINT,ignore.stderr=SYS_PRINT)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# --- HERITABILITY ANALYSIS\n",
    "if ( is.na(opt$hsq_set) ) {\n",
    "if ( opt$verbose >= 1 ) cat(\"### Estimating heritability\\n\")\n",
    "\n",
    "# 1. generate GRM\n",
    "arg = paste( opt$PATH_plink,\" --allow-no-sex --bfile \",opt$tmp,\" --make-grm-bin --out \",opt$tmp,sep='')\n",
    "system(arg , ignore.stdout=SYS_PRINT,ignore.stderr=SYS_PRINT)\n",
    "\n",
    "# 2. estimate heritability\n",
    "if ( !is.na(opt$covar) ) {\n",
    "arg = paste( opt$PATH_gcta ,\" --grm \",opt$tmp,\" --pheno \",raw.pheno.file,\" --qcovar \",opt$covar,\" --out \",opt$tmp,\" --reml --reml-no-constrain --reml-lrt 1\",sep='')\n",
    "} else {\n",
    "arg = paste( opt$PATH_gcta ,\" --grm \",opt$tmp,\" --pheno \",pheno.file,\" --out \",opt$tmp,\" --reml --reml-no-constrain --reml-lrt 1\",sep='')\n",
    "}\n",
    "system(arg , ignore.stdout=SYS_PRINT,ignore.stderr=SYS_PRINT)\n",
    "\n",
    "# 3. evaluate LRT and V(G)/Vp\n",
    "if ( !file.exists( paste(opt$tmp,\".hsq\",sep='') ) ) {\n",
    "\tcat(opt$tmp,\"does not exist, likely GCTA could not converge, skipping gene\\n\",file=stderr())\n",
    "\tcleanup()\n",
    "\tq()\n",
    "}\n",
    "\n",
    "hsq.file = read.table(file=paste(opt$tmp,\".hsq\",sep=''),as.is=T,fill=T)\n",
    "hsq = as.numeric(unlist(hsq.file[hsq.file[,1] == \"V(G)/Vp\",2:3]))\n",
    "hsq.pv = as.numeric(unlist(hsq.file[hsq.file[,1] == \"Pval\",2]))\n",
    "\n",
    "if ( opt$verbose >= 1 ) cat(\"Heritability (se):\",hsq,\"LRT P-value:\",hsq.pv,'\\n')\n",
    "if ( opt$save_hsq ) cat( opt$out , hsq , hsq.pv , '\\n' , file=paste(opt$out,\".hsq\",sep='') )\n",
    "\n",
    "# 4. stop if insufficient (Stop at here for most genes, ~60% out of 1000 genes)\n",
    "if ( hsq[1] < 0 || hsq.pv > opt$hsq_p ) {\n",
    "\tcat(opt$tmp,\" : heritability \",hsq[1],\"; LRT P-value \",hsq.pv,\" : skipping gene\\n\",sep='',file=stderr())\n",
    "\tcleanup()\n",
    "\tq()\n",
    "}\n",
    "\n",
    "} else {\n",
    "if ( opt$verbose >= 1 ) cat(\"### Skipping heritability estimate\\n\")\n",
    "hsq = opt$hsq_set\n",
    "hsq.pv = NA\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Haven't tested yet\n",
    "There is some problem for the following codes as the few genes that get passed the aforementioned steps didn't get passed the following. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in genotypes\n",
    "genos = read_plink(geno.file,impute=\"avg\")\n",
    "mafs = apply(genos$bed,2,mean)/2\n",
    "sds = apply(genos$bed,2,sd)\n",
    "# important : genotypes are standardized and scaled here:\n",
    "genos$bed = scale(genos$bed)\n",
    "pheno = genos$fam[,c(1,2,6)]\n",
    "pheno[,3] = scale(pheno[,3])\n",
    "\n",
    "# check if any genotypes are NA\n",
    "nasnps = apply( is.na(genos$bed) , 2 , sum )\n",
    "if ( sum(nasnps) != 0 ) {\n",
    "\tcat( \"WARNING :\",sum(nasnps != 0),\"SNPs could not be scaled and were zeroed out, make sure all SNPs are polymorphic\\n\" , file=stderr())\n",
    "\tgenos$bed[,nasnps != 0] = 0\n",
    "}\n",
    "\n",
    "# regress covariates out of the genotypes as well (this is more accurate but slower)\n",
    "if ( !is.na(opt$covar) && opt$resid ) {\n",
    "\tif ( opt$verbose >= 1 ) cat(\"regressing covariates out of the genotypes\\n\")\n",
    "\tfor ( i in 1:ncol(genos$bed) ) {\n",
    "\t\tgenos$bed[,i] = summary(lm( genos$bed[,i] ~ as.matrix(covar[,3:ncol(covar)]) ))$resid\n",
    "\t}\n",
    "\tgenos$bed = scale(genos$bed)\n",
    "}\n",
    "\n",
    "N.tot = nrow(genos$bed)\n",
    "if ( opt$verbose >= 1 ) cat(nrow(pheno),\"phenotyped samples, \",nrow(genos$bed),\"genotyped samples, \",ncol(genos$bed),\" markers\\n\")\n",
    "\n",
    "# --- CROSSVALIDATION ANALYSES\n",
    "set.seed(1)\n",
    "cv.performance = matrix(NA,nrow=2,ncol=M)\n",
    "rownames(cv.performance) = c(\"rsq\",\"pval\")\n",
    "colnames(cv.performance) = models\n",
    "\n",
    "if ( opt$crossval <= 1 ) {\n",
    "if ( opt$verbose >= 1 ) cat(\"### Skipping cross-validation\\n\")\n",
    "} else {\n",
    "if ( opt$verbose >= 1 ) cat(\"### Performing\",opt$crossval,\"fold cross-validation\\n\")\n",
    "cv.all = pheno\n",
    "N = nrow(cv.all)\n",
    "cv.sample = sample(N)\n",
    "cv.all = cv.all[ cv.sample , ]\n",
    "folds = cut(seq(1,N),breaks=opt$crossval,labels=FALSE)\n",
    "\n",
    "cv.calls = matrix(NA,nrow=N,ncol=M)\n",
    "\n",
    "for ( i in 1:opt$crossval ) {\n",
    "\tif ( opt$verbose >= 1 ) cat(\"- Crossval fold\",i,\"\\n\")\n",
    "\tindx = which(folds==i,arr.ind=TRUE)\n",
    "\tcv.train = cv.all[-indx,]\n",
    "\t# store intercept\n",
    "\tintercept = mean( cv.train[,3] )\n",
    "\tcv.train[,3] = scale(cv.train[,3])\n",
    "\t\n",
    "\t# hide current fold\n",
    "\tcv.file = paste(opt$tmp,\".cv\",sep='')\n",
    "\twrite.table( cv.train , quote=F , row.names=F , col.names=F , file=paste(cv.file,\".keep\",sep=''))\t\n",
    "\targ = paste( opt$PATH_plink ,\" --allow-no-sex --bfile \",opt$tmp,\" --keep \",cv.file,\".keep --out \",cv.file,\" --make-bed\",sep='')\n",
    "\tsystem(arg , ignore.stdout=SYS_PRINT,ignore.stderr=SYS_PRINT)\n",
    "\n",
    "\tfor ( mod in 1:M ) {\n",
    "\t\tif ( models[mod] == \"blup\" ) {\n",
    "\t\t\tpred.wgt = weights.bslmm( cv.file , bv_type=2 , snp=genos$bim[,2] )\n",
    "\t\t}\n",
    "\t\telse if ( models[mod] == \"bslmm\" ) {\n",
    "\t\t\tpred.wgt = weights.bslmm( cv.file , bv_type=1 , snp=genos$bim[,2] )\n",
    "\t\t}\t\t\n",
    "\t\telse if ( models[mod] == \"lasso\" ) {\n",
    "\t\t\tpred.wgt = weights.lasso( cv.file , hsq[1] , snp=genos$bim[,2] )\n",
    "\t\t}\n",
    "\t\telse if ( models[mod] == \"enet\" ) {\n",
    "\t\t\tpred.wgt = weights.enet( genos$bed[ cv.sample[ -indx ],] , as.matrix(cv.train[,3]) , alpha=0.5 )\n",
    "\t\t}\t\t\n",
    "\t\telse if ( models[mod] == \"top1\" ) {\n",
    "\t\t\tpred.wgt = weights.marginal( genos$bed[ cv.sample[ -indx ],] , as.matrix(cv.train[,3,drop=F]) , beta=T )\n",
    "\t\t\tpred.wgt[ - which.max( pred.wgt^2 ) ] = 0\n",
    "\t\t}\n",
    "\n",
    "\t\t# predict from weights into sample\n",
    "\t\tpred.wgt[ is.na(pred.wgt) ] = 0\n",
    "\t\tcv.calls[ indx , mod ] = genos$bed[ cv.sample[ indx ] , ] %*% pred.wgt\n",
    "\t}\n",
    "}\n",
    "\n",
    "# compute rsq + P-value for each model\n",
    "for ( mod in 1:M ) {\n",
    "\tif ( !is.na(sd(cv.calls[,mod])) && sd(cv.calls[,mod]) != 0 ) {\n",
    "\t\treg = summary(lm( cv.all[,3] ~ cv.calls[,mod] ))\n",
    "\t\tcv.performance[ 1, mod ] = reg$adj.r.sq\n",
    "\t\tcv.performance[ 2, mod ] = reg$coef[2,4]\n",
    "\t} else {\n",
    "\t\tcv.performance[ 1, mod ] = NA\n",
    "\t\tcv.performance[ 2, mod ] = NA\n",
    "\t}\n",
    "}\n",
    "if ( opt$verbose >= 1 ) write.table(cv.performance,quote=F,sep='\\t')\n",
    "}\n",
    "\n",
    "# --- FULL ANALYSES\n",
    "if ( opt$verbose >= 1 ) cat(\"Computing full-sample weights\\n\")\n",
    "\n",
    "# call models to get weights\n",
    "wgt.matrix = matrix(0,nrow=nrow(genos$bim),ncol=M)\n",
    "colnames(wgt.matrix) = models\n",
    "rownames(wgt.matrix) = genos$bim[,2]\n",
    "for ( mod in 1:M ) {\n",
    "\tif ( models[mod] == \"blup\" ) {\n",
    "\t\twgt.matrix[,mod] = weights.bslmm( geno.file , bv_type=2 , snp=genos$bim[,2] , out=opt$tmp )\n",
    "\t}\n",
    "\telse if ( models[mod] == \"bslmm\" ) {\n",
    "\t\twgt.matrix[,mod] = weights.bslmm( geno.file , bv_type=1 , snp=genos$bim[,2] , out=opt$tmp )\n",
    "\t}\t\t\n",
    "\telse if ( models[mod] == \"lasso\" ) {\n",
    "\t\twgt.matrix[,mod] = weights.lasso( geno.file , hsq[1] , snp=genos$bim[,2] , out=opt$tmp )\n",
    "\t}\n",
    "\telse if ( models[mod] == \"enet\" ) {\n",
    "\t\twgt.matrix[,mod] = weights.enet( genos$bed , as.matrix(pheno[,3]) , alpha=0.5 )\n",
    "\t}\t\n",
    "\telse if ( models[mod] == \"top1\" ) {\n",
    "\t\twgt.matrix[,mod] = weights.marginal( genos$bed , as.matrix(pheno[,3]) , beta=F )\n",
    "\t}\n",
    "}\n",
    "\n",
    "# save weights, rsq, p-value for each model, and hsq to output\n",
    "snps = genos$bim\n",
    "save( wgt.matrix , snps , cv.performance , hsq, hsq.pv, N.tot , file = paste( opt$out , \".wgt.RDat\" , sep='' ) )\n",
    "# --- CLEAN-UP\n",
    "if ( opt$verbose >= 1 ) cat(\"Cleaning up\\n\")\n",
    "cleanup()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
