<p><small>Exported from <a href="http://github.com/hsun3163/jnbinder/blob/93314c88684ab4704a5db1f03c434d85b038a8f8/analysis/Data_description.ipynb"><code>analysis/Data_description.ipynb</code></a> committed by Gao Wang on Wed May 5 20:26:26 2021 <a href="http://github.com/hsun3163/jnbinder/commit/93314c88684ab4704a5db1f03c434d85b038a8f8">revision 1, 93314c8</a> <a href="https://stephenslab.github.io/ipynb-website/notes.html#Note-about-commit-ids"><span class="fa fa-question-circle"></span></a></small></p>
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="ipynb_website:version" content="0.9.7" />
<meta name="viewport" content="width=device-width, initial-scale=1" />

<link rel="stylesheet" type="text/css" href="../css/jt.css">

<link rel="stylesheet" type="text/css" href="../css/toc2.css">

<link href="../site_libs/jqueryui-1.11.4/jquery-ui.css">
<link rel="stylesheet" href="../site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" href="../site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.9.1/jquery-ui.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>

<link rel="stylesheet"
      href="../site_libs/highlightjs/null.min.css"
      type="text/css" />

<script src="../site_libs/highlightjs/highlight.pack.js"></script>
<script>hljs.initHighlightingOnLoad();</script>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>

<script src="../js/doc_toc.js"></script>
<script src="../js/docs.js"></script>

<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-MML-AM_CHTML"></script>
<script>
    MathJax.Hub.Config({
        extensions: ["tex2jax.js"],
        jax: ["input/TeX", "output/HTML-CSS"],
        tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true
        },
        "HTML-CSS": {
            preferredFont: "TeX",
            availableFonts: ["TeX"],
            styles: {
                scale: 110,
                ".MathJax_Display": {
                    "font-size": "110%",
                }
            }
        }
    });
</script>
<script>
function filterDataFrame(id) {
    var input = document.getElementById("search_" + id);
    var filter = input.value.toUpperCase();
    var table = document.getElementById("dataframe_" + id);
    var tr = table.getElementsByTagName("tr");
    // Loop through all table rows, and hide those who don't match the search query
    for (var i = 1; i < tr.length; i++) {
        for (var j = 0; j < tr[i].cells.length; ++j) {
            var matched = false;
            if (tr[i].cells[j].innerHTML.toUpperCase().indexOf(filter) != -1) {
                tr[i].style.display = "";
                matched = true
                break;
            }
            if (!matched)
                tr[i].style.display = "none";
        }
    }
}
function sortDataFrame(id, n, dtype) {
    var table = document.getElementById("dataframe_" + id);
    var tb = table.tBodies[0]; // use `<tbody>` to ignore `<thead>` and `<tfoot>` rows
    var tr = Array.prototype.slice.call(tb.rows, 0); // put rows into array
    if (dtype === 'numeric') {
        var fn = function(a, b) { 
            return parseFloat(a.cells[n].textContent) <= parseFloat(b.cells[n].textContent) ? -1 : 1;
        }
    } else {
        var fn = function(a, b) {
            var c = a.cells[n].textContent.trim().localeCompare(b.cells[n].textContent.trim()); 
            return c > 0 ? 1 : (c < 0 ? -1 : 0) }
    }
    var isSorted = function(array, fn) {
        if (array.length < 2)
            return 1;
        var direction = fn(array[0], array[1]); 
        for (var i = 1; i < array.length - 1; ++i) {
            var d = fn(array[i], array[i+1]);
            if (d == 0)
                continue;
            else if (direction == 0)
                direction = d;
            else if (direction != d)
                return 0;
            }
        return direction;
    }
    var sorted = isSorted(tr, fn);
    if (sorted == 1 || sorted == -1) {
        // if sorted already, reverse it
        for(var i = tr.length - 1; i >= 0; --i)
            tb.appendChild(tr[i]); // append each row in order
    } else {
        tr = tr.sort(fn);
        for(var i = 0; i < tr.length; ++i)
            tb.appendChild(tr[i]); // append each row in order
    }
}
</script>

<script>
$( document ).ready(function(){
            var cfg={'threshold':3,     // depth of toc (number of levels)
             'number_sections': false,
             'toc_cell': false,          // useless here
             'toc_window_display': true, // display the toc window
             "toc_section_display": "block", // display toc contents in the window
             'sideBar':true,       // sidebar or floating window
             'navigate_menu':false       // navigation menu (only in liveNotebook -- do not change)
            }
            var st={};                  // some variables used in the script
            st.rendering_toc_cell = false;
            st.config_loaded = false;
            st.extension_initialized=false;
            st.nbcontainer_marginleft = $('#notebook-container').css('margin-left')
            st.nbcontainer_marginright = $('#notebook-container').css('margin-right')
            st.nbcontainer_width = $('#notebook-container').css('width')
            st.oldTocHeight = undefined
            st.cell_toc = undefined;
            st.toc_index=0;
            // fire the main function with these parameters
            table_of_contents(cfg, st);
            var file=analysisDict[$("h1:first").attr("id")];
            $("#toc-level0 a").css("color","#126dce");
            $('a[href="#'+$("h1:first").attr("id")+'"]').hide()
            var docs=analysisArray;
            var docs_map=analysisArrayMap;
            var pos=analysisArray.indexOf(file);
            for (var a=pos;a>=0;a--){
                  $('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>').insertBefore("#toc-level0 li:eq(0)");
            }
            $('a[href="'+file+'.html'+'"]').css("color","#126dce");
            for (var a=pos+1;a<docs.length;a++){
                  $(".toc #toc-level0").append('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>');
            }
            // $("#toc-header").hide(); // comment out because it prevents search bar from displaying
    });
</script>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');
  // mark it active
  menuAnchor.parent().addClass('active');
  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>
<div class="container-fluid main-container">
<!-- tabsets -->
<script src="../site_libs/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>



<title>Multi-tissue brain TWAS</title>

<style type = "text/css">
body {
  font-family: "Droid Sans";
  padding-top: 66px;
  padding-bottom: 40px;
}
</style>
</head>

<body>
<div tabindex="-1" id="notebook" class="border-box-sizing">
<div class="container" id="notebook-container">

<!-- code folding -->

<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">Multi-tissue brain TWAS</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
<li>
  <a href="../index.html">Overview</a>
</li>
        
<li>
  <a href="../writeup.html">Writeup</a>
</li>
        
<li>
  <a href="../analysis.html">Analysis</a>
</li>
        
<li>
  <a href="../workflow.html">Workflow</a>
</li>
        
      </ul>
        
<ul class="nav navbar-nav navbar-right">
<li>
   <a href="http://github.com/hsun3163/jnbinder"> <span class="fa fa-github"></span> </a>
</li>
</ul>
        
      </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="ROSMAP-Data-Description">ROSMAP Data Description<a class="anchor-link" href="#ROSMAP-Data-Description">&#182;</a></h1><h2 id="Background">Background<a class="anchor-link" href="#Background">&#182;</a></h2><p>The data used is the ROSEMAP data set, a combination of two studies: The Religious order Study (ROS) and the Memory and Aging Project (MAP)-that recruit older individuals without known dementia and include detailed cognitive, neuroimaging and other ante-mortem phenotyping and (2) an autopsy at the time of death that includes a structured neuropathologic examination. A detailed description of the study's background can be found at</p>
<p>De Jager, P. L. et al. A multi-omic atlas of the human frontal cortex for aging and Alzheimer’s disease research. Sci. Data 5:180142 doi: 10.1038/sdata.2018.142 (2018).
Publisher’s</p>
<h2 id="Data-set">Data set<a class="anchor-link" href="#Data-set">&#182;</a></h2><p>The dataset of we used in the analysis is located on the csg cluster <code>/mnt/mfs/ctcn/datasets/rosmap/</code>. Currently there are two types of data that are being used: gene expression, and Whole genome sequencing. The path of the file is documented at each sections.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Whole-Genome-Sequencing">Whole Genome Sequencing<a class="anchor-link" href="#Whole-Genome-Sequencing">&#182;</a></h3><p>The WGS data is stored at <code>/mnt/mfs/ctcn/datasets/rosmap/wgs/ampad/variants/snvCombinedPlink/</code>
There are 7176 samples being sequenced over 9536310 snps. These information is acquired by the following commands on the cluster</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[1]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="n">cd</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">ctcn</span><span class="o">/</span><span class="n">datasets</span><span class="o">/</span><span class="n">rosmap</span><span class="o">/</span><span class="n">wgs</span><span class="o">/</span><span class="n">ampad</span><span class="o">/</span><span class="n">variants</span><span class="o">/</span><span class="n">snvCombinedPlink</span>
<span class="c1"># For number of samples</span>
<span class="n">wc</span> <span class="o">*.</span><span class="n">fam</span>
<span class="c1"># For number of SNPs</span>
<span class="n">wc</span> <span class="o">*.</span><span class="n">bim</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>


<div class="output_subarea output_stream output_stderr output_text">
<pre>File contains parsing errors: 
	[line  2]: cd /mnt/mfs/ctcn/datasets/rosmap/wgs/ampad/variants/snvCombinedPlink
wc *.fam
# For number of SNPs
wc *.bim

Invalid statements: SyntaxError(&#39;invalid syntax&#39;, (&#39;&lt;string&gt;&#39;, 2, 5, &#39;wc *.fam\n&#39;))
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Genome-type-list">Genome type list<a class="anchor-link" href="#Genome-type-list">&#182;</a></h3><p>The genome type list file is created with the following code. This file provide a diretory for the genotype file that are used for the analysis workflow to run.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
<span class="n">library</span><span class="p">(</span><span class="n">tibble</span><span class="p">)</span>
<span class="n">a</span> <span class="o">=</span> <span class="n">tibble</span><span class="p">(</span>
  <span class="nb">chr</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="n">seq</span><span class="p">(</span><span class="mi">1</span><span class="p">:</span><span class="mi">23</span><span class="p">))</span>
<span class="p">)</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span>
  <span class="nb">dir</span> <span class="o">=</span> <span class="n">paste</span><span class="p">(</span><span class="s2">&quot;/mnt/mfs/ctcn/datasets/rosmap/wgs/ampad/variants/snvCombinedPlink/chr&quot;</span><span class="p">,</span><span class="nb">chr</span><span class="p">,</span><span class="s2">&quot;.bed&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">))</span><span class="o">%&gt;%</span>
    <span class="n">select</span><span class="p">(</span><span class="s2">&quot;#chr&quot;</span> <span class="o">=</span> <span class="nb">chr</span><span class="p">,</span><span class="nb">dir</span><span class="p">)</span><span class="o">%&gt;%</span>
    <span class="n">readr</span><span class="p">::</span><span class="n">write_tsv</span><span class="p">(</span> <span class="n">path</span> <span class="o">=</span> <span class="s2">&quot;/home/hs3163/Project/Rosmap/data/Rosmap_wgs_genotype_list.txt&quot;</span><span class="p">,</span> <span class="n">na</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span><span class="p">,</span> <span class="n">append</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span> <span class="n">col_names</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">quote_escape</span> <span class="o">=</span> <span class="s2">&quot;double&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Gene-expression">Gene expression<a class="anchor-link" href="#Gene-expression">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>The gene expression data for each tissues is stored at the following folders.</p>
<p>For dorsolateral prefrontal cortex (DLPFC), 18629 genes (filtered from 58302) from 1092 samples were sequenced.
<code>/mnt/mfs/ctcn/datasets/rosmap/rnaseq/dlpfcTissue/batch0-m5/values/</code></p>
<p>For anterior commissure (AC), 19147 genes from 731 samples were sequenced
<code>/mnt/mfs/ctcn/datasets/rosmap/rnaseq/acTissue/mbatch1-4/values/</code></p>
<p>For posterior cingulate cortex (PCC), 19017 genes (filtered from 58302) from samples were sequenced
<code>/mnt/mfs/ctcn/datasets/rosmap/rnaseq/pccTissue/mbatch1-5/values/</code>
The amount of samples sequenced is acquired by the following commands on the cluster</p>
<p>In it there is various folder as outlined below:</p>
<ul>
<li>/values<ul>
<li>/raw: raw expected counts and TPM data</li>
<li>/filtered: filtered data after removing outlier samples and lowly expressed genes</li>
<li>/residuals: residuals after regressing out significant covariates associated with RNA-seq data.</li>
<li>geneTpmResidualsAgeGenderUnadj.txt</li>
<li>geneTpmResidualsAgeGenderAdj.txt</li>
<li>residuals of TPM data.
Adjusted for batch and technical covariates: AgeGenderUnadj
Adjusted for age, sex, batch, and technical covariates: AgeGenderAdj</li>
<li>/scripts: Rscripts to produce these results.
For the purpose of our analysis, we shall use the residuals value taht are ajusted for both the technical covariate and the background such as age and sex.</li>
</ul>
</li>
</ul>
<p>The information above is acquired by the following commands on cluster.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[1]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># To get the number of samples</span>
<span class="n">head</span> <span class="o">-</span><span class="mi">1</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">ctcn</span><span class="o">/</span><span class="n">datasets</span><span class="o">/</span><span class="n">rosmap</span><span class="o">/</span><span class="n">rnaseq</span><span class="o">/</span><span class="n">dlpfcTissue</span><span class="o">/</span><span class="n">batch0</span><span class="o">-</span><span class="n">m5</span><span class="o">/</span><span class="n">values</span><span class="o">/</span><span class="n">residuals</span><span class="o">/</span><span class="n">geneTpmResidualsPlusBaselineAgeGenderAdj</span><span class="o">.</span><span class="n">txt</span> <span class="o">|</span> <span class="n">wc</span>
<span class="n">head</span> <span class="o">-</span><span class="mi">1</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">ctcn</span><span class="o">/</span><span class="n">datasets</span><span class="o">/</span><span class="n">rosmap</span><span class="o">/</span><span class="n">rnaseq</span><span class="o">/</span><span class="n">acTissue</span><span class="o">/</span><span class="n">mbatch1</span><span class="o">-</span><span class="mi">4</span><span class="o">/</span><span class="n">values</span><span class="o">/</span><span class="n">residuals</span><span class="o">/</span><span class="n">geneTpmResidualsAgeGenderAdj</span><span class="o">.</span><span class="n">txt</span> <span class="o">|</span> <span class="n">wc</span>
<span class="n">head</span> <span class="o">-</span><span class="mi">1</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">ctcn</span><span class="o">/</span><span class="n">datasets</span><span class="o">/</span><span class="n">rosmap</span><span class="o">/</span><span class="n">rnaseq</span><span class="o">/</span><span class="n">pccTissue</span><span class="o">/</span><span class="n">mbatch1</span><span class="o">-</span><span class="mi">5</span><span class="o">/</span><span class="n">values</span><span class="o">/</span><span class="n">residuals</span><span class="o">/</span><span class="n">geneTpmResidualsAgeGenderAdj</span><span class="o">.</span><span class="n">txt</span>  <span class="o">|</span> <span class="n">wc</span>
<span class="c1"># To get the number of genes</span>
<span class="n">head</span> <span class="o">-</span><span class="mi">1</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">ctcn</span><span class="o">/</span><span class="n">datasets</span><span class="o">/</span><span class="n">rosmap</span><span class="o">/</span><span class="n">rnaseq</span><span class="o">/</span><span class="n">dlpfcTissue</span><span class="o">/</span><span class="n">batch0</span><span class="o">-</span><span class="n">m5</span><span class="o">/</span><span class="n">values</span><span class="o">/</span><span class="n">residuals</span><span class="o">/</span><span class="n">geneTpmResidualsPlusBaselineAgeGenderAdj</span><span class="o">.</span><span class="n">txt</span> <span class="o">|</span> <span class="n">wc</span>
<span class="n">head</span> <span class="o">-</span><span class="mi">1</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">ctcn</span><span class="o">/</span><span class="n">datasets</span><span class="o">/</span><span class="n">rosmap</span><span class="o">/</span><span class="n">rnaseq</span><span class="o">/</span><span class="n">acTissue</span><span class="o">/</span><span class="n">mbatch1</span><span class="o">-</span><span class="mi">4</span><span class="o">/</span><span class="n">values</span><span class="o">/</span><span class="n">residuals</span><span class="o">/</span><span class="n">geneTpmResidualsAgeGenderAdj</span><span class="o">.</span><span class="n">txt</span> <span class="o">|</span> <span class="n">wc</span>
<span class="n">head</span> <span class="o">-</span><span class="mi">1</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">ctcn</span><span class="o">/</span><span class="n">datasets</span><span class="o">/</span><span class="n">rosmap</span><span class="o">/</span><span class="n">rnaseq</span><span class="o">/</span><span class="n">pccTissue</span><span class="o">/</span><span class="n">mbatch1</span><span class="o">-</span><span class="mi">5</span><span class="o">/</span><span class="n">values</span><span class="o">/</span><span class="n">residuals</span><span class="o">/</span><span class="n">geneTpmResidualsAgeGenderAdj</span><span class="o">.</span><span class="n">txt</span>  <span class="o">|</span> <span class="n">wc</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>


<div class="output_subarea output_text output_error">
<pre>
<span class="ansi-red-intense-fg">[0]: 
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
script_8805321491612483450 in &lt;module&gt;
----&gt; head -1 /mnt/mfs/ctcn/datasets/rosmap/rnaseq/dlpfcTissue/batch0-m5/values/residuals/geneTpmResidualsPlusBaselineAgeGenderAdj.txt | wc
      head -1 /mnt/mfs/ctcn/datasets/rosmap/rnaseq/acTissue/mbatch1-4/values/residuals/geneTpmResidualsAgeGenderAdj.txt | wc
      head -1 /mnt/mfs/ctcn/datasets/rosmap/rnaseq/pccTissue/mbatch1-5/values/residuals/geneTpmResidualsAgeGenderAdj.txt  | wc
      head -1 /mnt/mfs/ctcn/datasets/rosmap/rnaseq/dlpfcTissue/batch0-m5/values/residuals/geneTpmResidualsPlusBaselineAgeGenderAdj.txt | wc

NameError: name &#39;head&#39; is not defined</span></pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Problemetic-SNPs">Problemetic SNPs<a class="anchor-link" href="#Problemetic-SNPs">&#182;</a></h1><p>When using the WGS data to perform TWAS fusion or any other computation that required the bed matrix to be scale() in R. 3 SNPs,7:100549543_A_G" "7:100614961_A_G" "7:100615281_A_G", on chromosome 7 have universial 1, which produce NaN in R and hence casuing error that illustrated in issue #23. The code used to identified the error is shown below</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[3]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="n">library</span><span class="p">(</span><span class="n">tibble</span><span class="p">)</span>
<span class="n">library</span><span class="p">(</span><span class="n">plink2R</span><span class="p">)</span>
<span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
<span class="n">geno</span><span class="o">=</span><span class="n">read_plink</span><span class="p">(</span><span class="s2">&quot;/Users/haosun/Documents/WG_Reasearch_Assisstant/Remote_Proj/Alz/Alz_PCC_SNP/result/chr7&quot;</span><span class="p">,</span><span class="n">impute</span><span class="o">=</span><span class="s2">&quot;avg&quot;</span><span class="p">)</span>
<span class="n">cur</span><span class="o">.</span><span class="n">genos</span> <span class="o">=</span> <span class="n">scale</span><span class="p">(</span><span class="n">geno</span><span class="err">$</span><span class="n">bed</span><span class="p">[,</span><span class="n">m</span><span class="p">[</span><span class="n">m</span><span class="o">.</span><span class="n">keep</span><span class="p">]])</span>
<span class="kp">name</span> <span class="o">=</span> <span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">cur</span><span class="o">.</span><span class="n">genos</span><span class="p">))[</span><span class="mi">1</span><span class="p">])</span>
<span class="n">t</span><span class="p">(</span><span class="n">cur</span><span class="o">.</span><span class="n">genos</span><span class="p">)</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="o">%&gt;%</span><span class="nb">filter</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="o">.</span><span class="p">[[</span><span class="kp">name</span><span class="p">]]))</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>


<div class="output_subarea output_stream output_stderr output_text">
<pre>Loading required package: Rcpp

Loading required package: RcppEigen


Attaching package: ‘dplyr’


The following objects are masked from ‘package:stats’:

    filter, lag


The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union


</pre>
</div>
</div>

<div class="output_area">

    <div class="prompt"></div>


<div class="output_subarea output_text output_error">
<pre>
Error in scale(geno$bed[, m[m.keep]]): 找不到对象&#39;m&#39;
Traceback:

1. scale(geno$bed[, m[m.keep]])</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Data-Pre-Processing">Data Pre Processing<a class="anchor-link" href="#Data-Pre-Processing">&#182;</a></h3><p>Currently the sample names used in the gene expression file and the wgs file is not the same. A sos workflow is adopted to</p>
<ul>
<li>Match the sample names of the gene expression file with that of the genotypes file</li>
<li>Create a region list file for all the regions that are contained in the gene expression file. This list allow us to flexibily control what regions are included in the analysis</li>
</ul>
<p>The Index file that were used to matched the sample name can be found here:
 <code>/mnt/mfs/ctcn/datasets/rosmap/wgs/ampad/qualityControl/sampleSheetAfterQc.csv</code></p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">neuro</span><span class="o">-</span><span class="n">twas</span><span class="o">/</span><span class="n">Workflow</span><span class="o">/</span><span class="n">SOS_ROSMAP_gene_exp_processing</span><span class="o">.</span><span class="n">ipynb</span> \
    <span class="o">--</span><span class="n">ref</span> <span class="s2">&quot;/home/hs3163/Project/Rosmap/data/patient_key_WGS.txt&quot;</span> \
    <span class="o">--</span><span class="n">name_col</span> <span class="mi">2</span> \
    <span class="o">--</span><span class="n">real_name_col</span> <span class="mi">1</span> \
    <span class="o">--</span><span class="n">gene_exp</span> <span class="s2">&quot;/mnt/mfs/ctcn/datasets/rosmap/rnaseq/dlpfcTissue/batch0-m5/values/residuals/geneTpmResidualsAgeGenderAdj.txt&quot;</span> \
    <span class="o">--</span><span class="n">start_at</span> <span class="mi">2</span> \
    <span class="o">--</span><span class="n">output</span> <span class="s2">&quot;/home/hs3163/Project/Rosmap/data/gene_exp/DLPFC&quot;</span> \
    <span class="o">-</span><span class="n">J</span> <span class="mi">6</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="o">~/</span><span class="n">system_file</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span>

<span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">neuro</span><span class="o">-</span><span class="n">twas</span><span class="o">/</span><span class="n">Workflow</span><span class="o">/</span><span class="n">SOS_ROSMAP_gene_exp_processing</span><span class="o">.</span><span class="n">ipynb</span> \
    <span class="o">--</span><span class="n">ref</span> <span class="s2">&quot;/home/hs3163/Project/Rosmap/data/patient_key_WGS.txt&quot;</span> \
    <span class="o">--</span><span class="n">name_col</span> <span class="mi">2</span> \
    <span class="o">--</span><span class="n">real_name_col</span> <span class="mi">1</span> \
    <span class="o">--</span><span class="n">gene_exp</span> <span class="s2">&quot;/mnt/mfs/ctcn/datasets/rosmap/rnaseq/pccTissue/mbatch1-5/values/residuals/geneTpmResidualsAgeGenderAdj.txt&quot;</span> \
    <span class="o">--</span><span class="n">start_at</span> <span class="mi">2</span> \
    <span class="o">--</span><span class="n">output</span> <span class="s2">&quot;/home/hs3163/Project/Rosmap/data/gene_exp/PCC&quot;</span> \
    <span class="o">-</span><span class="n">J</span> <span class="mi">6</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="o">~/</span><span class="n">system_file</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span>

<span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">neuro</span><span class="o">-</span><span class="n">twas</span><span class="o">/</span><span class="n">Workflow</span><span class="o">/</span><span class="n">SOS_ROSMAP_gene_exp_processing</span><span class="o">.</span><span class="n">ipynb</span> \
    <span class="o">--</span><span class="n">ref</span> <span class="s2">&quot;/home/hs3163/Project/Rosmap/data/patient_key_WGS.txt&quot;</span> \
    <span class="o">--</span><span class="n">name_col</span> <span class="mi">2</span> \
    <span class="o">--</span><span class="n">real_name_col</span> <span class="mi">1</span> \
    <span class="o">--</span><span class="n">gene_exp</span> <span class="s2">&quot;/mnt/mfs/ctcn/datasets/rosmap/rnaseq/acTissue/mbatch1-4/values/residuals/geneTpmResidualsAgeGenderAdj.txt&quot;</span> \
    <span class="o">--</span><span class="n">start_at</span> <span class="mi">2</span> \
    <span class="o">--</span><span class="n">output</span> <span class="s2">&quot;/home/hs3163/Project/Rosmap/data/gene_exp/AC&quot;</span> <span class="o">&amp;</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>After the processing of the data, for DLPFC, there are 852 samples, with 17891 genes.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>To estimate how many of the genes passed can pass the heritability check, following command is ran.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="kn">R:</span>
<span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
<span class="n">region_list</span><span class="o">=</span><span class="n">read_delim</span><span class="p">(</span><span class="s2">&quot;/home/hs3163/Project/Rosmap/data/gene_exp/DLPFC/geneTpmResidualsAgeGenderAdj_rename_region_list.txt&quot;</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="c1"># Select 2000 genes that have not been ran.</span>
<span class="n">index</span> <span class="o">=</span> <span class="n">sample</span><span class="p">(</span><span class="mi">201</span><span class="p">:</span><span class="n">nrow</span><span class="p">(</span><span class="n">region_list</span><span class="p">),</span><span class="mi">2000</span><span class="p">)</span>
<span class="n">region_list_selected</span> <span class="o">=</span> <span class="n">region_list</span><span class="p">[</span><span class="n">index</span><span class="p">,]</span>
<span class="n">region_list_selected</span><span class="o">%&gt;%</span><span class="n">readr</span><span class="p">::</span><span class="n">write_delim</span><span class="p">(</span><span class="s2">&quot;/home/hs3163/Project/est_hsq/est_hsq_region_list.txt&quot;</span><span class="p">,</span><span class="n">delim</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="kn">bash:</span>
<span class="c1"># To save time, compute weight without the bslmm module</span>
<span class="n">nohup</span> <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">neuro</span><span class="o">-</span><span class="n">twas</span><span class="o">/</span><span class="n">Workflow</span><span class="o">/</span><span class="n">twas_fusion</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">compute_wgt</span>  \
  <span class="o">--</span><span class="n">molecular</span><span class="o">-</span><span class="n">pheno</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">Project</span><span class="o">/</span><span class="n">Rosmap</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">gene_exp</span><span class="o">/</span><span class="n">DLPFC</span><span class="o">/</span><span class="n">geneTpmResidualsAgeGenderAdj_rename</span><span class="o">.</span><span class="n">txt</span>  \
  <span class="o">--</span><span class="n">wd</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">Project</span><span class="o">/</span><span class="n">est_hsq</span> \
  <span class="o">--</span><span class="n">genotype_list</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">Project</span><span class="o">/</span><span class="n">Rosmap</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">Rosmap_wgs_genotype_list</span><span class="o">.</span><span class="n">txt</span> \
  <span class="o">--</span><span class="n">region_list</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">Project</span><span class="o">/</span><span class="n">est_hsq</span><span class="o">/</span><span class="n">est_hsq_region_list</span><span class="o">.</span><span class="n">txt</span> \
  <span class="o">--</span><span class="n">region_name</span> <span class="mi">1</span> \
  <span class="o">--</span><span class="n">data_start</span> <span class="mi">2</span> \
  <span class="o">--</span><span class="n">window</span> <span class="mi">500000</span> \
  <span class="o">--</span><span class="n">container</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">system_file</span><span class="o">/</span><span class="n">twas_latest</span><span class="o">.</span><span class="n">sif</span> \
  <span class="o">--</span><span class="n">model</span> <span class="n">blup</span> <span class="n">lasso</span> <span class="n">top1</span> <span class="n">enet</span> \
  <span class="o">--</span><span class="n">job_size</span> <span class="mi">50</span>\
  <span class="o">-</span><span class="n">J</span> <span class="mi">10</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">neuro</span><span class="o">-</span><span class="n">twas</span><span class="o">/</span><span class="n">template</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">s</span> <span class="n">build</span> <span class="o">&amp;</span>
</pre></div>

    </div>
</div>
</div>

</div>
<hr>
&copy 2020-2021 Hao Sun & Gao Wang, Department of Neurology, Columbia University
<p><small>Exported from <a href="http://github.com/hsun3163/jnbinder/blob/93314c88684ab4704a5db1f03c434d85b038a8f8/analysis/Data_description.ipynb"><code>analysis/Data_description.ipynb</code></a> committed by Gao Wang on Thu May 6 00:26:26 2021 <a href="http://github.com/hsun3163/jnbinder/commit/93314c88684ab4704a5db1f03c434d85b038a8f8">revision 1, 93314c8</a> <a href="https://stephenslab.github.io/ipynb-website/notes.html#Note-about-commit-ids"><span class="fa fa-question-circle"></span></a></small></p>
</div>
</div>
</body>
</html>
