<p><small>Exported from <a href="http://github.com/hsun3163/jnbinder/blob/93314c88684ab4704a5db1f03c434d85b038a8f8/analysis/Training_Testing_imputed_expression.ipynb"><code>analysis/Training_Testing_imputed_expression.ipynb</code></a> committed by Gao Wang on Wed May 5 20:26:26 2021 <a href="http://github.com/hsun3163/jnbinder/commit/93314c88684ab4704a5db1f03c434d85b038a8f8">revision 1, 93314c8</a> <a href="https://stephenslab.github.io/ipynb-website/notes.html#Note-about-commit-ids"><span class="fa fa-question-circle"></span></a></small></p>
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="ipynb_website:version" content="0.9.7" />
<meta name="viewport" content="width=device-width, initial-scale=1" />

<link rel="stylesheet" type="text/css" href="../css/jt.css">

<link rel="stylesheet" type="text/css" href="../css/toc2.css">

<link href="../site_libs/jqueryui-1.11.4/jquery-ui.css">
<link rel="stylesheet" href="../site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" href="../site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.9.1/jquery-ui.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>

<link rel="stylesheet"
      href="../site_libs/highlightjs/null.min.css"
      type="text/css" />

<script src="../site_libs/highlightjs/highlight.pack.js"></script>
<script>hljs.initHighlightingOnLoad();</script>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>

<script src="../js/doc_toc.js"></script>
<script src="../js/docs.js"></script>

<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-MML-AM_CHTML"></script>
<script>
    MathJax.Hub.Config({
        extensions: ["tex2jax.js"],
        jax: ["input/TeX", "output/HTML-CSS"],
        tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true
        },
        "HTML-CSS": {
            preferredFont: "TeX",
            availableFonts: ["TeX"],
            styles: {
                scale: 110,
                ".MathJax_Display": {
                    "font-size": "110%",
                }
            }
        }
    });
</script>
<script>
function filterDataFrame(id) {
    var input = document.getElementById("search_" + id);
    var filter = input.value.toUpperCase();
    var table = document.getElementById("dataframe_" + id);
    var tr = table.getElementsByTagName("tr");
    // Loop through all table rows, and hide those who don't match the search query
    for (var i = 1; i < tr.length; i++) {
        for (var j = 0; j < tr[i].cells.length; ++j) {
            var matched = false;
            if (tr[i].cells[j].innerHTML.toUpperCase().indexOf(filter) != -1) {
                tr[i].style.display = "";
                matched = true
                break;
            }
            if (!matched)
                tr[i].style.display = "none";
        }
    }
}
function sortDataFrame(id, n, dtype) {
    var table = document.getElementById("dataframe_" + id);
    var tb = table.tBodies[0]; // use `<tbody>` to ignore `<thead>` and `<tfoot>` rows
    var tr = Array.prototype.slice.call(tb.rows, 0); // put rows into array
    if (dtype === 'numeric') {
        var fn = function(a, b) { 
            return parseFloat(a.cells[n].textContent) <= parseFloat(b.cells[n].textContent) ? -1 : 1;
        }
    } else {
        var fn = function(a, b) {
            var c = a.cells[n].textContent.trim().localeCompare(b.cells[n].textContent.trim()); 
            return c > 0 ? 1 : (c < 0 ? -1 : 0) }
    }
    var isSorted = function(array, fn) {
        if (array.length < 2)
            return 1;
        var direction = fn(array[0], array[1]); 
        for (var i = 1; i < array.length - 1; ++i) {
            var d = fn(array[i], array[i+1]);
            if (d == 0)
                continue;
            else if (direction == 0)
                direction = d;
            else if (direction != d)
                return 0;
            }
        return direction;
    }
    var sorted = isSorted(tr, fn);
    if (sorted == 1 || sorted == -1) {
        // if sorted already, reverse it
        for(var i = tr.length - 1; i >= 0; --i)
            tb.appendChild(tr[i]); // append each row in order
    } else {
        tr = tr.sort(fn);
        for(var i = 0; i < tr.length; ++i)
            tb.appendChild(tr[i]); // append each row in order
    }
}
</script>

<script>
$( document ).ready(function(){
            var cfg={'threshold':3,     // depth of toc (number of levels)
             'number_sections': false,
             'toc_cell': false,          // useless here
             'toc_window_display': true, // display the toc window
             "toc_section_display": "block", // display toc contents in the window
             'sideBar':true,       // sidebar or floating window
             'navigate_menu':false       // navigation menu (only in liveNotebook -- do not change)
            }
            var st={};                  // some variables used in the script
            st.rendering_toc_cell = false;
            st.config_loaded = false;
            st.extension_initialized=false;
            st.nbcontainer_marginleft = $('#notebook-container').css('margin-left')
            st.nbcontainer_marginright = $('#notebook-container').css('margin-right')
            st.nbcontainer_width = $('#notebook-container').css('width')
            st.oldTocHeight = undefined
            st.cell_toc = undefined;
            st.toc_index=0;
            // fire the main function with these parameters
            table_of_contents(cfg, st);
            var file=analysisDict[$("h1:first").attr("id")];
            $("#toc-level0 a").css("color","#126dce");
            $('a[href="#'+$("h1:first").attr("id")+'"]').hide()
            var docs=analysisArray;
            var docs_map=analysisArrayMap;
            var pos=analysisArray.indexOf(file);
            for (var a=pos;a>=0;a--){
                  $('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>').insertBefore("#toc-level0 li:eq(0)");
            }
            $('a[href="'+file+'.html'+'"]').css("color","#126dce");
            for (var a=pos+1;a<docs.length;a++){
                  $(".toc #toc-level0").append('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>');
            }
            // $("#toc-header").hide(); // comment out because it prevents search bar from displaying
    });
</script>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');
  // mark it active
  menuAnchor.parent().addClass('active');
  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>
<div class="container-fluid main-container">
<!-- tabsets -->
<script src="../site_libs/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>



<title>Multi-tissue brain TWAS</title>

<style type = "text/css">
body {
  font-family: "Droid Sans";
  padding-top: 66px;
  padding-bottom: 40px;
}
</style>
</head>

<body>
<div tabindex="-1" id="notebook" class="border-box-sizing">
<div class="container" id="notebook-container">

<!-- code folding -->

<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">Multi-tissue brain TWAS</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
<li>
  <a href="../index.html">Overview</a>
</li>
        
<li>
  <a href="../writeup.html">Writeup</a>
</li>
        
<li>
  <a href="../analysis.html">Analysis</a>
</li>
        
<li>
  <a href="../workflow.html">Workflow</a>
</li>
        
      </ul>
        
<ul class="nav navbar-nav navbar-right">
<li>
   <a href="http://github.com/hsun3163/jnbinder"> <span class="fa fa-github"></span> </a>
</li>
</ul>
        
      </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Training-and-testing-for-imputed-expression">Training and testing for imputed expression<a class="anchor-link" href="#Training-and-testing-for-imputed-expression">&#182;</a></h1><p>This notebook contains the codes that used to select a training vs testing set and then use the sets to evaluate how good is the prediction.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Sample-selection">Sample selection<a class="anchor-link" href="#Sample-selection">&#182;</a></h2><p>The following codes are used to generate a random collection of samples that are treated as the training set. The amount of training set consists of 80% (682) of the samples.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="kn">R:</span> 
<span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span>
<span class="c1"># Set the seed of the process</span>
<span class="n">setseed</span><span class="p">(</span><span class="mi">999</span><span class="p">)</span>
<span class="c1"># Load in the expression data</span>
<span class="n">gene_exp</span> <span class="o">=</span> <span class="n">read_delim</span><span class="p">(</span><span class="s2">&quot;/home/hs3163/Project/Rosmap/data/gene_exp/DLPFC/geneTpmResidualsAgeGenderAdj_rename.txt&quot;</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="c1"># Randomly chose the training set</span>
<span class="n">training_index</span> <span class="o">=</span> <span class="n">sample</span><span class="p">(</span><span class="mi">2</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">gene_exp</span><span class="p">),</span><span class="n">ncol</span><span class="p">(</span><span class="n">gene_exp</span><span class="p">)</span><span class="o">*</span><span class="mf">0.8</span><span class="p">)</span>
<span class="n">training_samples</span> <span class="o">=</span> <span class="n">colnames</span><span class="p">(</span><span class="n">gene_exp</span><span class="p">)[</span><span class="n">training_index</span><span class="p">]</span>
<span class="n">gene_exp_training</span> <span class="o">=</span> <span class="n">gene_exp</span><span class="o">%&gt;%</span><span class="n">dplyr</span><span class="p">::</span><span class="n">select</span><span class="p">(</span><span class="n">gene_ID</span><span class="p">,</span><span class="n">training_samples</span><span class="p">)</span>
<span class="n">gene_exp_testing</span> <span class="o">=</span> <span class="n">gene_exp</span><span class="o">%&gt;%</span><span class="n">dplyr</span><span class="p">::</span><span class="n">select</span><span class="p">(</span><span class="o">-</span><span class="n">training_samples</span><span class="p">)</span>

<span class="c1"># </span>
<span class="n">gene_exp_training</span><span class="o">%&gt;%</span><span class="n">readr</span><span class="p">::</span><span class="n">write_delim</span><span class="p">(</span><span class="s2">&quot;/home/hs3163/Project/Rosmap/data/gene_exp/DLPFC/Training_Testing/geneTpmResidualsAgeGenderAdj_rename_training.txt&quot;</span><span class="p">,</span><span class="n">delim</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="n">gene_exp_testing</span><span class="o">%&gt;%</span><span class="n">readr</span><span class="p">::</span><span class="n">write_delim</span><span class="p">(</span><span class="s2">&quot;/home/hs3163/Project/Rosmap/data/gene_exp/DLPFC/Training_Testing/geneTpmResidualsAgeGenderAdj_rename_testing.txt&quot;</span><span class="p">,</span><span class="n">delim</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Gene-selections">Gene selections<a class="anchor-link" href="#Gene-selections">&#182;</a></h2><p>For the purpose of testing of the imputed gene expression, to avoid additional complecation, the genes that are proven to passed the heratibility gene and succefully produce a weight was selected. The genes are acquired from the test run of rosmap data (first 200 genes from the gene expression file) and the gene list acquired from the Alz data set (143 genes that are in both tables).
Further information please see the following notebook:</p>
<p>{link to be added}</p>
<p>The genes are combined into one region_list by the following commands, all of the 89 genes are unique.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="kn">bash:</span>
<span class="c1"># Combining the genes</span>
<span class="n">cd</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">Project</span><span class="o">/</span><span class="n">traning_testing</span><span class="o">/</span><span class="n">data</span>
<span class="n">awk</span> <span class="s1">&#39;{ print $3 &quot; &quot; $4 &quot; &quot; $5 &quot; &quot; $2}&#39;</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">Project</span><span class="o">/</span><span class="n">test</span><span class="o">/</span><span class="mi">1</span><span class="o">-</span><span class="mi">30</span><span class="o">/</span><span class="n">WEIGHTS</span><span class="o">/*</span><span class="n">weights_list</span><span class="o">*</span> <span class="o">&gt;</span> <span class="n">TT_region_list</span><span class="o">.</span><span class="n">txt</span>
<span class="n">tail</span> <span class="o">-</span><span class="n">n</span> <span class="o">+</span><span class="mi">2</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">Project</span><span class="o">/</span><span class="n">test</span><span class="o">/</span><span class="mi">30</span><span class="o">-</span><span class="mi">200</span><span class="o">/</span><span class="n">WEIGHTS</span><span class="o">/*</span><span class="n">weights_list</span><span class="o">*</span> <span class="o">|</span> <span class="n">awk</span> <span class="s1">&#39;{ print $3 &quot; &quot; $4 &quot; &quot; $5 &quot; &quot; $2}&#39;</span> <span class="o">-</span> <span class="o">&gt;&gt;</span> <span class="n">TT_region_list</span><span class="o">.</span><span class="n">txt</span>
<span class="n">tail</span> <span class="o">-</span><span class="n">n</span> <span class="o">+</span><span class="mi">2</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">Project</span><span class="o">/</span><span class="n">Alz</span><span class="o">/</span><span class="n">WEIGHTS</span><span class="o">/*</span><span class="n">weights_list</span><span class="o">*</span> <span class="o">|</span> <span class="n">awk</span> <span class="s1">&#39;{ print $3 &quot; &quot; $4 &quot; &quot; $5 &quot; &quot; $2}&#39;</span> <span class="o">-</span> <span class="o">&gt;&gt;</span> <span class="n">TT_region_list</span><span class="o">.</span><span class="n">txt</span>

<span class="kn">R:</span>
<span class="c1"># See if the genes are unique</span>
<span class="n">region_list</span> <span class="o">=</span> <span class="n">readr</span><span class="p">::</span><span class="n">read_delim</span><span class="p">(</span><span class="s2">&quot;/home/hs3163/Project/traning_testing/data/TT_region_list.txt&quot;</span><span class="p">,</span><span class="s2">&quot; &quot;</span><span class="p">)</span>
<span class="n">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">region_list</span><span class="err">$</span><span class="n">ID</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Recomputing-weight-for-the-training-set">Recomputing weight for the training set<a class="anchor-link" href="#Recomputing-weight-for-the-training-set">&#182;</a></h2><p>The weights for each of the genes, using only the sampless in the training sets, are recomputed using the following commands</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="n">nohup</span> <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">neuro</span><span class="o">-</span><span class="n">twas</span><span class="o">/</span><span class="n">Workflow</span><span class="o">/</span><span class="n">twas_fusion</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">compute_wgt</span>  \
  <span class="o">--</span><span class="n">molecular</span><span class="o">-</span><span class="n">pheno</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">Project</span><span class="o">/</span><span class="n">Rosmap</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">gene_exp</span><span class="o">/</span><span class="n">DLPFC</span><span class="o">/</span><span class="n">Training_Testing</span><span class="o">/</span><span class="n">geneTpmResidualsAgeGenderAdj_rename_training</span><span class="o">.</span><span class="n">txt</span>  \
  <span class="o">--</span><span class="n">wd</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">Project</span><span class="o">/</span><span class="n">traning_testing</span> \
  <span class="o">--</span><span class="n">genotype_list</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">Project</span><span class="o">/</span><span class="n">Rosmap</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">Rosmap_wgs_genotype_list</span><span class="o">.</span><span class="n">txt</span> \
  <span class="o">--</span><span class="n">region_list</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">Project</span><span class="o">/</span><span class="n">traning_testing</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">TT_region_list</span><span class="o">.</span><span class="n">txt</span> \
  <span class="o">--</span><span class="n">region_name</span> <span class="mi">1</span> \
  <span class="o">--</span><span class="n">data_start</span> <span class="mi">2</span> \
  <span class="o">--</span><span class="n">window</span> <span class="mi">500000</span> \
  <span class="o">--</span><span class="n">container</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">system_file</span><span class="o">/</span><span class="n">twas_latest</span><span class="o">.</span><span class="n">sif</span> \
  <span class="o">--</span><span class="n">model</span> <span class="n">bslmm</span> <span class="n">blup</span> <span class="n">lasso</span> <span class="n">top1</span> <span class="n">enet</span> \
  <span class="o">--</span><span class="n">job_size</span> <span class="mi">1</span>\
  <span class="o">-</span><span class="n">J</span> <span class="mi">6</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">neuro</span><span class="o">-</span><span class="n">twas</span><span class="o">/</span><span class="n">template</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">s</span> <span class="n">build</span> <span class="o">&amp;</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>It took 14 hours for the pipeline to be ran on the 89 genes. Among them, 18 genes did not passed the heritability check. The present of new genes that fail to pass the heritability checked is anticipated. I suspect the 18 genes that did not passed the test are those with borderline heritabilities.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Imputed-expression-estimation-and-comparison.">Imputed expression estimation and comparison.<a class="anchor-link" href="#Imputed-expression-estimation-and-comparison.">&#182;</a></h1>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[3]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="n">suppressMessages</span><span class="p">(</span><span class="n">library</span><span class="p">(</span><span class="s1">&#39;plink2R&#39;</span><span class="p">))</span>
<span class="n">read_plink</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[5]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="o">%</span><span class="n">run</span>
<span class="p">[</span><span class="n">susie_1</span><span class="p">]</span>
<span class="n">geno_file</span> <span class="o">=</span> <span class="s2">&quot;/Users/haosun/Documents/WG_Reasearch_Assisstant/&quot;</span>
<span class="kn">R:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">volumes</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span><span class="s1">&#39;{geno_file}:{geno_file}&#39;</span><span class="p">]</span>
   <span class="n">library</span><span class="p">(</span><span class="s2">&quot;susieR&quot;</span><span class="p">)</span>
   <span class="n">library</span><span class="p">(</span><span class="s2">&quot;plink2R&quot;</span><span class="p">)</span>
    <span class="n">geno</span> <span class="o">=</span> <span class="n">read_plink</span><span class="p">(</span><span class="s2">&quot;/Users/haosun/Documents/WG_Reasearch_Assisstant/Remote_Proj/Alz/Alz_AC_SNP/cache/geneTpmResidualsAgeGenderAdj_rename.ENSG00000256294&quot;</span><span class="p">)</span>
    <span class="n">X</span> <span class="o">=</span> <span class="n">scale</span><span class="p">(</span><span class="n">geno</span><span class="err">$</span><span class="n">bed</span><span class="p">)</span>
    <span class="n">Y</span> <span class="o">=</span> <span class="n">geno</span><span class="err">$</span><span class="n">fam</span><span class="p">[</span><span class="mi">6</span><span class="p">,]</span>
    <span class="n">fitted</span> <span class="o">=</span> <span class="n">susie</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">Y</span><span class="p">[,</span><span class="mi">1</span><span class="p">],</span>
              <span class="n">L</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
              <span class="n">estimate_residual_variance</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> 
              <span class="n">estimate_prior_variance</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span> 
              <span class="n">scaled_prior_variance</span> <span class="o">=</span> <span class="mf">0.2</span><span class="p">)</span>
    <span class="n">Hsq</span> <span class="o">=</span> <span class="n">var</span><span class="p">(</span><span class="n">predict</span><span class="p">(</span><span class="n">fitted</span><span class="p">))</span> <span class="o">/</span> <span class="n">var</span><span class="p">(</span><span class="n">Y</span><span class="p">)</span>
    <span class="n">Hsq</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>



<div class="output_html rendered_html output_subarea ">
<div class="sos_logging sos_info">INFO: Running <span class="sos_highlight">susie_1</span>: 
</div>
</div>

</div>

<div class="output_area">

    <div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>HINT: Pulling docker image gaow/twas
Error in library(&#34;susieR&#34;) : there is no package called ‘susieR’
Execution halted
<span class="ansi-red-intense-fg">ERROR</span>: <span class="ansi-red-intense-fg">[susie_1]: [0]: Executing script in docker returns an error (exitcode=1).
The script has been saved to /Users/haosun/Documents/WG_Reasearch_Assisstant/GIT/freshcopy/neuro-twas/Analysis/ROSMAP_TWAS/.sos/docker_run_24279.R. To reproduce the error please run:
</span><span class="ansi-green-fg">docker run --rm  -v /Users/haosun/Documents/WG_Reasearch_Assisstant:/Users/haosun/Documents/WG_Reasearch_Assisstant -v /Users/haosun/Documents/WG_Reasearch_Assisstant/GIT/freshcopy/neuro-twas/Analysis/ROSMAP_TWAS/.sos/docker_run_24279.R:/var/lib/sos/docker_run_24279.R    -t  -w=/Users/haosun/Documents/WG_Reasearch_Assisstant/GIT/freshcopy/neuro-twas/Analysis/ROSMAP_TWAS -u 501:20    gaow/twas Rscript /var/lib/sos/docker_run_24279.R</span>
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[7]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="n">docker</span> <span class="n">run</span> <span class="o">--</span><span class="n">rm</span>  <span class="o">-</span><span class="n">v</span> <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">haosun</span><span class="o">/</span><span class="n">Documents</span><span class="o">/</span><span class="n">WG_Reasearch_Assisstant</span><span class="p">:</span><span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">haosun</span><span class="o">/</span><span class="n">Documents</span><span class="o">/</span><span class="n">WG_Reasearch_Assisstant</span> <span class="o">-</span><span class="n">v</span> <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">haosun</span><span class="o">/</span><span class="n">Documents</span><span class="o">/</span><span class="n">WG_Reasearch_Assisstant</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">freshcopy</span><span class="o">/</span><span class="n">neuro</span><span class="o">-</span><span class="n">twas</span><span class="o">/</span><span class="n">Analysis</span><span class="o">/</span><span class="n">ROSMAP_TWAS</span><span class="o">/.</span><span class="n">sos</span><span class="o">/</span><span class="n">docker_run_24279</span><span class="o">.</span><span class="n">R</span><span class="p">:</span><span class="o">/</span><span class="n">var</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">sos</span><span class="o">/</span><span class="n">docker_run_24279</span><span class="o">.</span><span class="n">R</span>    <span class="o">-</span><span class="n">t</span>  <span class="o">-</span><span class="n">w</span><span class="o">=/</span><span class="n">Users</span><span class="o">/</span><span class="n">haosun</span><span class="o">/</span><span class="n">Documents</span><span class="o">/</span><span class="n">WG_Reasearch_Assisstant</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">freshcopy</span><span class="o">/</span><span class="n">neuro</span><span class="o">-</span><span class="n">twas</span><span class="o">/</span><span class="n">Analysis</span><span class="o">/</span><span class="n">ROSMAP_TWAS</span> <span class="o">-</span><span class="n">u</span> <span class="mi">501</span><span class="p">:</span><span class="mi">20</span>    <span class="n">gaow</span><span class="o">/</span><span class="n">twas</span> <span class="n">Rscript</span> <span class="o">/</span><span class="n">var</span><span class="o">/</span><span class="n">lib</span><span class="o">/</span><span class="n">sos</span><span class="o">/</span><span class="n">docker_run_24279</span><span class="o">.</span><span class="n">R</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>Error in library(&#34;susieR&#34;) : there is no package called ‘susieR’
Execution halted

</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span>
</pre></div>

    </div>
</div>
</div>

</div>
<hr>
&copy 2020-2021 Hao Sun & Gao Wang, Department of Neurology, Columbia University
<p><small>Exported from <a href="http://github.com/hsun3163/jnbinder/blob/93314c88684ab4704a5db1f03c434d85b038a8f8/analysis/Training_Testing_imputed_expression.ipynb"><code>analysis/Training_Testing_imputed_expression.ipynb</code></a> committed by Gao Wang on Thu May 6 00:26:26 2021 <a href="http://github.com/hsun3163/jnbinder/commit/93314c88684ab4704a5db1f03c434d85b038a8f8">revision 1, 93314c8</a> <a href="https://stephenslab.github.io/ipynb-website/notes.html#Note-about-commit-ids"><span class="fa fa-question-circle"></span></a></small></p>
</div>
</div>
</body>
</html>
