{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "Bash"
   },
   "source": [
    "## Enviorment\n",
    "The following section are used to specify the hyperparameter and the path the the software that were used. \n",
    "No need to changed if the pipeline stay the same and were applied to different data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "##!/bin/sh\n",
    "# MAKE SURE FUSION.compute_weights.R IS IN YOUR PATH\n",
    "# FILL IN THESE PATHS\n",
    "# For mac user, the mac version of GCTA shall be downloaded saperately, the one came with the Fusion package will not work.\n",
    "GCTA=\"/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/gcta_1.93.2beta_mac/gcta64\"\n",
    "PLINK=\"/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/plink_mac_20200616/plink\"\n",
    "GEMMA=\"/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/GEMMA\"\n",
    "PLINKED_file=\"/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles\"\n",
    "\n",
    "# Required the customized fusion.compute_weight.mod.R script, other wise will not work\n",
    "Rscp=\"/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/FUSION.compute_weights.mod.R\"\n",
    "# Path to the input data\n",
    "Datarep=\"/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/\"\n",
    "# ALTERNATIVELY: ENSURE THAT plink, gcta, gemma CAN BE CALLED FROM PATH AND REMOVE --PATH_* FLAGS BELOW\n",
    "\n",
    "# PATH TO WORKING DIRECTORY\n",
    "wd=\"/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working\"\n",
    "\n",
    "# PATH TO OUTPUT DIRECTORY (population-specific subdirs will be made)\n",
    "OUT_DIR=\"./WEIGHTS\"\n",
    "\n",
    "# ROWS IN THE MATRIX TO ANALYZE (FOR BATCHED RUNS)\n",
    "BATCH_START=1\n",
    "BATCH_END=49"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Input \n",
    "\n",
    "This is where the actual RNA Seq data of our own should be used as input.\n",
    "\n",
    "Note: GD462.GeneQuantRPKM.50FN.samplename.resk10.txt is the rna seq data for gene expression from the \"465 lymphoblastoid cell lines from the 1000 Genomes\" projects. There is 23722 genes from 462 samples. The first 4 col is \"TargetID, Gene_Symbol, Chr, Coord\". The rest are each samples.\n",
    "\n",
    "GD462.GeneQuantRPKM.50FN.samplename.resk10.txt is the RNA seq data for the first 10 ppl\n",
    "\n",
    "LDREF/1000G.EUR.*.bim contains the linkage disequilibrium info for each snp."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "YRI.120samples.hg19\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# PATH TO DIRECTORY CONTAINING LDREF DATA (FROM FUSION WEBSITE or https://data.broadinstitute.org/alkesgroup/FUSION/LDREF.tar.bz2)\n",
    "LDREF=\"/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles\"\n",
    "# THIS IS USED TO RESTRICT INPUT SNPS TO REFERENCE IDS ONLY\n",
    "\n",
    "# PATH TO GEUVADIS GENE EXPRESSION MATRIX:\n",
    "PRE_GEXP=\"fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt\"\n",
    "# GEUVADIS DATA WAS DOWNLOADED FROM https://www.ebi.ac.uk/arrayexpress/experiments/E-GEUV-1/files/analysis_results/\n",
    "\n",
    "# PATH TO PREFIX FOR GEUVADIS GENOTYPES SPLIT BY CHROMOSOME\n",
    "# SUBSAMPLE THESE TO THE LDREF SNPS FOR EFFICIENCY\n",
    "PRE_GENO=\"YRI.120samples.hg19\"\n",
    "echo $PRE_GENO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Pipeline\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# --- BEGIN SCRIPT:\n",
    "cd $wd\n",
    "NR=\"${BATCH_START}_${BATCH_END}\"\n",
    "mkdir -p tmp/$NR\n",
    "mkdir -p hsq/$NR\n",
    "mkdir -p out/$NR\n",
    "# THIS IS DIRECTORY WHERE THE OUTPUT WILL GO:\n",
    "mkdir $OUT_DIR"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "Bash"
   },
   "source": [
    "Note 1: the first two column of the out.pheno shall be the same as that of the .fam file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# makes necessary component that are useful for follow up pipeline\n",
    "cd $wd\n",
    "\n",
    "# get the within family sample ID, need to customized for it to match those of the LDREF file\n",
    "head -1 $Datarep$PRE_GEXP | awk '{$1=$2=$3=$4=\"\"; print substr($0,5)}' | fmt -1 | awk '$0=\"NA\"$0' > PRE_GEXPID\n",
    "# make a diretory containing the per_gene file\n",
    "mkdir $wd/${PRE_GEXP%.*}_per_gene\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "Bash"
   },
   "source": [
    "## SOS Pipeline prep (WIP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000225972.1\n",
      "ENSG00000225630.1\n",
      "ENSG00000237973.1\n",
      "ENSG00000229344.1\n",
      "ENSG00000248527.1\n",
      "ENSG00000198744.5\n",
      "ENSG00000187634.6\n",
      "ENSG00000188976.6\n",
      "ENSG00000187961.9\n",
      "ENSG00000187583.6\n",
      "ENSG00000188290.6\n",
      "ENSG00000187608.5\n",
      "ENSG00000188157.9\n",
      "ENSG00000242590.1\n",
      "ENSG00000217801.5\n",
      "ENSG00000273443.1\n",
      "ENSG00000131591.13\n",
      "ENSG00000186891.9\n",
      "ENSG00000186827.6\n",
      "ENSG00000078808.12\n",
      "ENSG00000176022.3\n",
      "ENSG00000160087.16\n",
      "ENSG00000131584.14\n",
      "ENSG00000169972.7\n",
      "ENSG00000127054.14\n",
      "ENSG00000240731.1\n",
      "ENSG00000224051.2\n",
      "ENSG00000107404.13\n",
      "ENSG00000175756.9\n",
      "ENSG00000221978.7\n",
      "ENSG00000224870.3\n",
      "ENSG00000242485.1\n",
      "ENSG00000264293.1\n",
      "ENSG00000272455.1\n",
      "ENSG00000179403.10\n",
      "ENSG00000160072.15\n",
      "ENSG00000197785.9\n",
      "ENSG00000160075.10\n",
      "ENSG00000215791.2\n",
      "ENSG00000228594.1\n",
      "ENSG00000272106.1\n",
      "ENSG00000197530.8\n",
      "ENSG00000248333.3\n",
      "ENSG00000272004.1\n",
      "ENSG00000189339.7\n",
      "ENSG00000269737.1\n",
      "ENSG00000269227.1\n",
      "ENSG00000008128.18\n",
      "ENSG00000268575.1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cd $wd\n",
    "\n",
    "# extract each gene from the data matrixs to a file\n",
    "cat $Datarep$PRE_GEXP | awk -v s=$BATCH_START -v e=$BATCH_END 'NR > s && NR <= e' |  while read PARAM; do\n",
    "\n",
    "GNAME=`echo $PARAM | awk '{ print $4 }'`\n",
    "echo $GNAME\n",
    "touch $wd/${PRE_GEXP%.*}_per_gene/${PRE_GEXP%.*}_$GNAME.txt\n",
    "echo $PARAM > $wd/${PRE_GEXP%.*}_per_gene/${PRE_GEXP%.*}_$GNAME.txt\n",
    "\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "Bash"
   },
   "source": [
    "## Original Pipeline (Working but needs to transformed into sos pipeline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000225972.1 chr1 64442 1064442\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000225972.1.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 64442\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000225972.1.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000225972.1\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000225972.1.pheno\n",
      "  --to-bp 1064442\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "3399 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000225972.1.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 3399 variants remaining.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.983795.\n",
      "3399 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000225972.1.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000225972.1.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000225972.1.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000225972.1.tmp : heritability -0.252611; LRT P-value 0.078845 : skipping gene\n",
      "end of circle\n",
      "ENSG00000225630.1 chr1 65020 1065020\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000225630.1.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 65020\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000225630.1.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000225630.1\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000225630.1.pheno\n",
      "  --to-bp 1065020\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "3405 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000225630.1.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 3405 variants remaining.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.983806.\n",
      "3405 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000225630.1.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000225630.1.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000225630.1.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000225630.1.tmp : heritability 0.004331; LRT P-value 0.49149 : skipping gene\n",
      "end of circle\n",
      "ENSG00000237973.1 chr1 66454 1066454\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000237973.1.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 66454\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000237973.1.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000237973.1\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000237973.1.pheno\n",
      "  --to-bp 1066454\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "3428 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000237973.1.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 3428 variants remaining.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.983812.\n",
      "3428 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000237973.1.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000237973.1.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000237973.1.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000237973.1.tmp : heritability -0.128193; LRT P-value 0.3009 : skipping gene\n",
      "end of circle\n",
      "ENSG00000229344.1 chr1 68137 1068137\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000229344.1.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 68137\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000229344.1.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000229344.1\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000229344.1.pheno\n",
      "  --to-bp 1068137\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "3448 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000229344.1.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 3448 variants remaining.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.983834.\n",
      "3448 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000229344.1.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000229344.1.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000229344.1.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000229344.1.tmp : heritability -0.357426; LRT P-value 0.015634 : skipping gene\n",
      "end of circle\n",
      "ENSG00000248527.1 chr1 69076 1069076\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000248527.1.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 69076\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000248527.1.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000248527.1\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000248527.1.pheno\n",
      "  --to-bp 1069076\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "3457 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000248527.1.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 3457 variants remaining.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.983859.\n",
      "3457 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000248527.1.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000248527.1.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000248527.1.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000248527.1.tmp : heritability 0.00143; LRT P-value 0.49767 : skipping gene\n",
      "end of circle\n",
      "ENSG00000198744.5 chr1 69756 1069756\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000198744.5.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 69756\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000198744.5.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000198744.5\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000198744.5.pheno\n",
      "  --to-bp 1069756\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "3470 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000198744.5.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 3470 variants remaining.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.983787.\n",
      "3470 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000198744.5.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000198744.5.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000198744.5.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000198744.5.tmp : heritability -0.138645; LRT P-value 0.23358 : skipping gene\n",
      "end of circle\n",
      "ENSG00000187634.6 chr1 360260 1360260\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187634.6.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 360260\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187634.6.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187634.6\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187634.6.pheno\n",
      "  --to-bp 1360260\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "6229 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187634.6.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 6229 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.987085.\n",
      "6229 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187634.6.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187634.6.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187634.6.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187634.6.tmp : heritability -0.159623; LRT P-value 0.31851 : skipping gene\n",
      "end of circle\n",
      "ENSG00000188976.6 chr1 379584 1379584\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188976.6.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 379584\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188976.6.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188976.6\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188976.6.pheno\n",
      "  --to-bp 1379584\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "6415 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188976.6.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 6415 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.986662.\n",
      "6415 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188976.6.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188976.6.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188976.6.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188976.6.tmp : heritability 0.379109; LRT P-value 0.089861 : skipping gene\n",
      "end of circle\n",
      "ENSG00000187961.9 chr1 395967 1395967\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187961.9.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 395967\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187961.9.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187961.9\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187961.9.pheno\n",
      "  --to-bp 1395967\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "6585 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187961.9.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 6585 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.98615.\n",
      "6585 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187961.9.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187961.9.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187961.9.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187961.9.tmp : heritability 0.440384; LRT P-value 0.019087 : skipping gene\n",
      "end of circle\n",
      "ENSG00000187583.6 chr1 401877 1401877\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187583.6.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 401877\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187583.6.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187583.6\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187583.6.pheno\n",
      "  --to-bp 1401877\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "6662 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187583.6.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 6662 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.98606.\n",
      "6662 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187583.6.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187583.6.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187583.6.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187583.6.tmp : heritability 0.206654; LRT P-value 0.12634 : skipping gene\n",
      "end of circle\n",
      "ENSG00000188290.6 chr1 434342 1434342\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188290.6.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 434342\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188290.6.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188290.6\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188290.6.pheno\n",
      "  --to-bp 1434342\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "7053 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188290.6.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 7053 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.985863.\n",
      "7053 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188290.6.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188290.6.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188290.6.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188290.6.tmp : heritability -0.008582; LRT P-value 0.48582 : skipping gene\n",
      "end of circle\n",
      "ENSG00000187608.5 chr1 448803 1448803\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187608.5.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 448803\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187608.5.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187608.5\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187608.5.pheno\n",
      "  --to-bp 1448803\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "7206 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187608.5.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 7206 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.985955.\n",
      "7206 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187608.5.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187608.5.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187608.5.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000187608.5.tmp : heritability 0.240545; LRT P-value 0.13128 : skipping gene\n",
      "end of circle\n",
      "ENSG00000188157.9 chr1 455503 1455503\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188157.9.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 455503\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188157.9.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188157.9\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188157.9.pheno\n",
      "  --to-bp 1455503\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "7276 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188157.9.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 7276 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.986002.\n",
      "7276 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188157.9.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188157.9.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188157.9.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000188157.9.tmp : heritability -0.081477; LRT P-value 0.36505 : skipping gene\n",
      "end of circle\n",
      "ENSG00000242590.1 chr1 490413 1490413\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000242590.1.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 490413\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000242590.1.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000242590.1\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000242590.1.pheno\n",
      "  --to-bp 1490413\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "7639 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000242590.1.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 7639 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.986235.\n",
      "7639 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000242590.1.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000242590.1.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000242590.1.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000242590.1.tmp : heritability 0.01071; LRT P-value 0.48215 : skipping gene\n",
      "end of circle\n",
      "ENSG00000217801.5 chr1 495114 1495114\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000217801.5.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 495114\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000217801.5.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000217801.5\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000217801.5.pheno\n",
      "  --to-bp 1495114\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "7672 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000217801.5.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 7672 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.986268.\n",
      "7672 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000217801.5.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000217801.5.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000217801.5.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000217801.5.tmp : heritability -0.443257; LRT P-value 0.048901 : skipping gene\n",
      "end of circle\n",
      "ENSG00000273443.1 chr1 497588 1497588\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000273443.1.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 497588\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000273443.1.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000273443.1\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000273443.1.pheno\n",
      "  --to-bp 1497588\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "7692 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000273443.1.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 7692 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.986299.\n",
      "7692 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000273443.1.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000273443.1.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000273443.1.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000273443.1.tmp : heritability 0.022891; LRT P-value 0.46835 : skipping gene\n",
      "end of circle\n",
      "ENSG00000131591.13 chr1 517198 1517198\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000131591.13.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 517198\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000131591.13.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000131591.13\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000131591.13.pheno\n",
      "  --to-bp 1517198\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "7851 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000131591.13.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 7851 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.986461.\n",
      "7851 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000131591.13.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000131591.13.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000131591.13.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000131591.13.tmp : heritability 0.394741; LRT P-value 0.045339 : skipping gene\n",
      "end of circle\n",
      "ENSG00000186891.9 chr1 638888 1638888\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000186891.9.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 638888\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000186891.9.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000186891.9\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000186891.9.pheno\n",
      "  --to-bp 1638888\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "8742 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000186891.9.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 8742 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.985408.\n",
      "8742 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000186891.9.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000186891.9.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000186891.9.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000186891.9.tmp : heritability 0.13376; LRT P-value 0.29771 : skipping gene\n",
      "end of circle\n",
      "ENSG00000186827.6 chr1 646706 1646706\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000186827.6.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 646706\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000186827.6.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000186827.6\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000186827.6.pheno\n",
      "  --to-bp 1646706\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "8759 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000186827.6.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 8759 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.985345.\n",
      "8759 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000186827.6.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000186827.6.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000186827.6.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000186827.6.tmp : heritability 0.147121; LRT P-value 0.26187 : skipping gene\n",
      "end of circle\n",
      "ENSG00000078808.12 chr1 652288 1652288\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000078808.12.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 652288\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000078808.12.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000078808.12\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000078808.12.pheno\n",
      "  --to-bp 1652288\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "8833 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000078808.12.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 8833 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.985112.\n",
      "8833 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000078808.12.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000078808.12.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000078808.12.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000078808.12.tmp : heritability 0.342345; LRT P-value 0.11329 : skipping gene\n",
      "end of circle\n",
      "ENSG00000176022.3 chr1 667629 1667629\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000176022.3.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 667629\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000176022.3.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000176022.3\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000176022.3.pheno\n",
      "  --to-bp 1667629\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "8968 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000176022.3.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 8968 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.985116.\n",
      "8968 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000176022.3.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000176022.3.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000176022.3.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000176022.3.tmp : heritability 0.385077; LRT P-value 0.058903 : skipping gene\n",
      "end of circle\n",
      "ENSG00000160087.16 chr1 689289 1689289\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160087.16.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 689289\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160087.16.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160087.16\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160087.16.pheno\n",
      "  --to-bp 1689289\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "9116 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160087.16.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 9116 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.985306.\n",
      "9116 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160087.16.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160087.16.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160087.16.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160087.16.tmp : heritability 0.55225; LRT P-value 0.015293 : skipping gene\n",
      "end of circle\n",
      "ENSG00000131584.14 chr1 727756 1727756\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000131584.14.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 727756\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000131584.14.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000131584.14\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000131584.14.pheno\n",
      "  --to-bp 1727756\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "9253 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000131584.14.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 9253 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.985747.\n",
      "9253 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000131584.14.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000131584.14.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000131584.14.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000131584.14.tmp : heritability 0.536299; LRT P-value 0.012837 : skipping gene\n",
      "end of circle\n",
      "ENSG00000169972.7 chr1 743947 1743947\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000169972.7.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 743947\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000169972.7.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000169972.7\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000169972.7.pheno\n",
      "  --to-bp 1743947\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "9254 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000169972.7.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 9254 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.986068.\n",
      "9254 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000169972.7.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000169972.7.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000169972.7.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "WARNING : 2457 SNPs could not be scaled and were zeroed out, make sure all SNPs are polymorphic\n",
      "end of circle\n",
      "ENSG00000127054.14 chr1 746965 1746965\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000127054.14.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 746965\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000127054.14.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000127054.14\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000127054.14.pheno\n",
      "  --to-bp 1746965\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "9264 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000127054.14.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 9264 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.98612.\n",
      "9264 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000127054.14.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000127054.14.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000127054.14.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "WARNING : 2463 SNPs could not be scaled and were zeroed out, make sure all SNPs are polymorphic\n",
      "end of circle\n",
      "ENSG00000240731.1 chr1 752961 1752961\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000240731.1.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 752961\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000240731.1.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000240731.1\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000240731.1.pheno\n",
      "  --to-bp 1752961\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "9263 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000240731.1.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 9263 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.986193.\n",
      "9263 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000240731.1.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000240731.1.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000240731.1.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000240731.1.tmp : heritability 0.207026; LRT P-value 0.18524 : skipping gene\n",
      "end of circle\n",
      "ENSG00000224051.2 chr1 760136 1760136\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000224051.2.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 760136\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000224051.2.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000224051.2\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000224051.2.pheno\n",
      "  --to-bp 1760136\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "9225 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000224051.2.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 9225 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.986223.\n",
      "9225 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000224051.2.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000224051.2.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000224051.2.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "WARNING : 2446 SNPs could not be scaled and were zeroed out, make sure all SNPs are polymorphic\n",
      "end of circle\n",
      "ENSG00000107404.13 chr1 770656 1770656\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000107404.13.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 770656\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000107404.13.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000107404.13\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000107404.13.pheno\n",
      "  --to-bp 1770656\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "9204 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000107404.13.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 9204 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.986377.\n",
      "9204 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000107404.13.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000107404.13.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000107404.13.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000107404.13.tmp : heritability 0.308693; LRT P-value 0.13504 : skipping gene\n",
      "end of circle\n",
      "ENSG00000175756.9 chr1 809110 1809110\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000175756.9.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 809110\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000175756.9.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000175756.9\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000175756.9.pheno\n",
      "  --to-bp 1809110\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "9028 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000175756.9.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 9028 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.986941.\n",
      "9028 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000175756.9.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000175756.9.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000175756.9.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000175756.9.tmp : heritability 0.078136; LRT P-value 0.39285 : skipping gene\n",
      "end of circle\n",
      "ENSG00000221978.7 chr1 821091 1821091\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000221978.7.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 821091\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000221978.7.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000221978.7\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000221978.7.pheno\n",
      "  --to-bp 1821091\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "9015 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000221978.7.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 9015 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.987083.\n",
      "9015 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000221978.7.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000221978.7.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000221978.7.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000221978.7.tmp : heritability 0.116915; LRT P-value 0.36833 : skipping gene\n",
      "end of circle\n",
      "ENSG00000224870.3 chr1 834902 1834902\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000224870.3.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 834902\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000224870.3.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000224870.3\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000224870.3.pheno\n",
      "  --to-bp 1834902\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "8997 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000224870.3.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 8997 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.987265.\n",
      "8997 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000224870.3.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000224870.3.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000224870.3.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000224870.3.tmp : heritability 0.16293; LRT P-value 0.2784 : skipping gene\n",
      "end of circle\n",
      "ENSG00000242485.1 chr1 837288 1837288\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000242485.1.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 837288\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000242485.1.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000242485.1\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000242485.1.pheno\n",
      "  --to-bp 1837288\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "8980 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000242485.1.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 8980 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.987289.\n",
      "8980 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000242485.1.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000242485.1.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000242485.1.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000242485.1.tmp : heritability 0.487102; LRT P-value 0.062577 : skipping gene\n",
      "end of circle\n",
      "ENSG00000264293.1 chr1 840841 1840841\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000264293.1.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 840841\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000264293.1.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000264293.1\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000264293.1.pheno\n",
      "  --to-bp 1840841\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "8956 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000264293.1.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 8956 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.987318.\n",
      "8956 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000264293.1.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000264293.1.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000264293.1.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000264293.1.tmp : heritability 0.1007; LRT P-value 0.35961 : skipping gene\n",
      "end of circle\n",
      "ENSG00000272455.1 chr1 844476 1844476\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272455.1.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 844476\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272455.1.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272455.1\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272455.1.pheno\n",
      "  --to-bp 1844476\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "8956 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272455.1.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 8956 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.98738.\n",
      "8956 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272455.1.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272455.1.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272455.1.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "WARNING : 2375 SNPs could not be scaled and were zeroed out, make sure all SNPs are polymorphic\n",
      "end of circle\n",
      "ENSG00000179403.10 chr1 870241 1870241\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000179403.10.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 870241\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000179403.10.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000179403.10\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000179403.10.pheno\n",
      "  --to-bp 1870241\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "8887 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000179403.10.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 8887 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.987805.\n",
      "8887 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000179403.10.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000179403.10.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000179403.10.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000179403.10.tmp : heritability 0.066477; LRT P-value 0.40687 : skipping gene\n",
      "end of circle\n",
      "ENSG00000160072.15 chr1 907143 1907143\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160072.15.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 907143\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160072.15.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160072.15\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160072.15.pheno\n",
      "  --to-bp 1907143\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "8979 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160072.15.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 8979 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.987622.\n",
      "8979 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160072.15.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160072.15.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160072.15.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160072.15.tmp : heritability 0.358549; LRT P-value 0.12251 : skipping gene\n",
      "end of circle\n",
      "ENSG00000197785.9 chr1 947531 1947531\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000197785.9.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 947531\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000197785.9.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000197785.9\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000197785.9.pheno\n",
      "  --to-bp 1947531\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "9107 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000197785.9.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 9107 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.987057.\n",
      "9107 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000197785.9.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000197785.9.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000197785.9.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000197785.9.tmp : heritability -0.053259; LRT P-value 0.43037 : skipping gene\n",
      "end of circle\n",
      "ENSG00000160075.10 chr1 977053 1977053\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160075.10.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 977053\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160075.10.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160075.10\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160075.10.pheno\n",
      "  --to-bp 1977053\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "9116 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160075.10.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 9116 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.987045.\n",
      "9116 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160075.10.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160075.10.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160075.10.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000160075.10.tmp : heritability -0.377878; LRT P-value 0.089595 : skipping gene\n",
      "end of circle\n",
      "ENSG00000215791.2 chr1 997726 1997726\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000215791.2.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 997726\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000215791.2.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000215791.2\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000215791.2.pheno\n",
      "  --to-bp 1997726\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "9152 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000215791.2.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 9152 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.986942.\n",
      "9152 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000215791.2.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000215791.2.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000215791.2.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000215791.2.tmp : heritability 0.116192; LRT P-value 0.34069 : skipping gene\n",
      "end of circle\n",
      "ENSG00000228594.1 chr1 1033392 2033392\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000228594.1.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 1033392\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000228594.1.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000228594.1\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000228594.1.pheno\n",
      "  --to-bp 2033392\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "9302 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000228594.1.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 9302 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.987057.\n",
      "9302 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000228594.1.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000228594.1.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000228594.1.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000228594.1.tmp : heritability -0.150838; LRT P-value 0.29807 : skipping gene\n",
      "end of circle\n",
      "ENSG00000272106.1 chr1 1049138 2049138\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272106.1.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 1049138\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272106.1.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272106.1\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272106.1.pheno\n",
      "  --to-bp 2049138\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "9360 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272106.1.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 9360 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.987146.\n",
      "9360 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272106.1.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272106.1.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272106.1.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272106.1.tmp does not exist, likely GCTA could not converge, skipping gene\n",
      "cat: ./WEIGHTS/ALL/ALL.ENSG00000272106.1.hsq: No such file or directory\n",
      "end of circle\n",
      "ENSG00000197530.8 chr1 1050795 2050795\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000197530.8.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 1050795\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000197530.8.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000197530.8\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000197530.8.pheno\n",
      "  --to-bp 2050795\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "9354 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000197530.8.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 9354 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.987156.\n",
      "9354 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000197530.8.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000197530.8.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000197530.8.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000197530.8.tmp : heritability 0.113682; LRT P-value 0.37601 : skipping gene\n",
      "end of circle\n",
      "ENSG00000248333.3 chr1 1070603 2070603\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000248333.3.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 1070603\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000248333.3.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000248333.3\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000248333.3.pheno\n",
      "  --to-bp 2070603\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "9268 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000248333.3.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 9268 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.987469.\n",
      "9268 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000248333.3.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000248333.3.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000248333.3.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000248333.3.tmp : heritability 0.455032; LRT P-value 0.037003 : skipping gene\n",
      "end of circle\n",
      "ENSG00000272004.1 chr1 1090786 2090786\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272004.1.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 1090786\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272004.1.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272004.1\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272004.1.pheno\n",
      "  --to-bp 2090786\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "9224 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272004.1.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 9224 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.987675.\n",
      "9224 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272004.1.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272004.1.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272004.1.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000272004.1.tmp : heritability 0.626125; LRT P-value 0.02467 : skipping gene\n",
      "end of circle\n",
      "ENSG00000189339.7 chr1 1092939 2092939\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000189339.7.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 1092939\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000189339.7.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000189339.7\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000189339.7.pheno\n",
      "  --to-bp 2092939\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "9207 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000189339.7.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 9207 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.987672.\n",
      "9207 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000189339.7.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000189339.7.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000189339.7.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000189339.7.tmp : heritability 0.215412; LRT P-value 0.19627 : skipping gene\n",
      "end of circle\n",
      "ENSG00000269737.1 chr1 1103429 2103429\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000269737.1.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 1103429\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000269737.1.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000269737.1\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000269737.1.pheno\n",
      "  --to-bp 2103429\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "9164 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000269737.1.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 9164 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.987688.\n",
      "9164 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000269737.1.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000269737.1.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000269737.1.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000269737.1.tmp : heritability -0.025414; LRT P-value 0.45859 : skipping gene\n",
      "end of circle\n",
      "ENSG00000269227.1 chr1 1104714 2104714\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000269227.1.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 1104714\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000269227.1.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000269227.1\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000269227.1.pheno\n",
      "  --to-bp 2104714\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "9159 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000269227.1.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 9159 variants remaining.\n",
      "Warning: At least 25 duplicate IDs in --extract file.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.987688.\n",
      "9159 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000269227.1.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000269227.1.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000269227.1.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000269227.1.tmp : heritability 0.021218; LRT P-value 0.4659 : skipping gene\n",
      "end of circle\n",
      "ENSG00000008128.18 chr1 1134169 2134169\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000008128.18.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 1134169\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000008128.18.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000008128.18\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000008128.18.pheno\n",
      "  --to-bp 2134169\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "8956 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000008128.18.nosex .\n",
      "85 phenotype values present after --pheno.\n",
      "--extract: 8956 variants remaining.\n",
      "--keep: 85 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 85 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate in remaining samples is 0.987722.\n",
      "8956 variants and 85 people pass filters and QC.\n",
      "Phenotype data is quantitative.\n",
      "--make-bed to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000008128.18.bed +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000008128.18.bim +\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000008128.18.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000008128.18.tmp : heritability 0.600777; LRT P-value 0.015346 : skipping gene\n",
      "end of circle\n",
      "ENSG00000268575.1 chr1 1134175 2134175\n",
      "PLINK v1.90b6.18 64-bit (16 Jun 2020)          www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2020 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000268575.1.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1\n",
      "  --chr chr1\n",
      "  --extract /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles/YRI.120samples.hg19.chr1.bim\n",
      "  --from-bp 1134175\n",
      "  --keep tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000268575.1.pheno\n",
      "  --make-bed\n",
      "  --out tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000268575.1\n",
      "  --pheno tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000268575.1.pheno\n",
      "  --to-bp 2134175\n",
      "\n",
      "16384 MB RAM detected; reserving 8192 MB for main workspace.\n",
      "8956 out of 1639740 variants loaded from .bim file.\n",
      "120 people (0 males, 0 females, 120 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000268575.1.nosex .\n",
      "Error: Line 71 of --pheno file has fewer tokens than expected.\n",
      "mkdir: ./WEIGHTS/ALL: File exists\n",
      "ERROR: tmp/1_100/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt.ENSG00000268575.1.bed input file does not exist\n",
      "cat: ./WEIGHTS/ALL/ALL.ENSG00000268575.1.hsq: No such file or directory\n",
      "end of circle\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cd $wd\n",
    "# Loop through each gene expression phenotype in the batch (Only looking at the first 10 genes)\n",
    "cat $Datarep$PRE_GEXP | awk -v s=$BATCH_START -v e=$BATCH_END 'NR > s && NR <= e' |  while read PARAM; do\n",
    "# Get the gene positions +/- 500kb\n",
    "CHR=`echo $PARAM | awk '{ print $1 }'`\n",
    "P0=`echo $PARAM | awk '{ print $2 - 0.5e6 }'`\n",
    "P1=`echo $PARAM | awk '{ print $2 + 0.5e6 }'`\n",
    "GNAME=`echo $PARAM | awk '{ print $4 }'`\n",
    "\n",
    "OUT=\"tmp/$NR/$PRE_GEXP.$GNAME\"\n",
    "echo $GNAME $CHR $P0 $P1\n",
    "\n",
    "## Pull out the current gene expression phenotype (Note 1 )\n",
    "echo $PARAM | tr ' ' '\\n' | tail -n+5 | paste PRE_GEXPID PRE_GEXPID - > $OUT.pheno\n",
    "## Get the locus genotypes for all samples and set current gene expression as the phenotype\n",
    "$PLINK --bfile $LDREF/$PRE_GENO.$CHR --pheno $tmpdir$OUT.pheno --make-bed --out $tmpdir$OUT --chr $CHR --from-bp $P0 --to-bp $P1 --extract $LDREF/$PRE_GENO.$CHR.bim --keep $tmpdir$OUT.pheno --allow-no-sex\n",
    "## Process all samples together (for reference purposes only since this is mult-ethnic data)\n",
    "mkdir $OUT_DIR/ALL\n",
    "FINAL_OUT=\"$OUT_DIR/ALL/ALL.$GNAME\"\n",
    "#echo $FINAL_OUT\n",
    "#\n",
    "#echo $OUT.tmp\n",
    "\n",
    "Rscript $Rscp --bfile $OUT --tmp $OUT.tmp --out $FINAL_OUT --verbose 0 --save_hsq --PATH_gcta $GCTA --PATH_gemma $GEMMA --models blup,lasso,top1,enet\n",
    "\n",
    "#\n",
    "## Append heritability output to hsq file\n",
    "cat $FINAL_OUT.hsq >> hsq/$NR.hsq\n",
    "#\n",
    "## Clean-up just in case\n",
    "rm -f $FINAL_OUT.hsq $OUT.tmp.*\n",
    "#\n",
    "# Remove all intermediate files\n",
    "echo \"end of circle\"\n",
    "\n",
    "# GO TO THE NEXT GENE\n",
    "done\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2020-07-13 14:31:25--  https://github.com/cumc/neuro-twas/blob/master/WIP/pipeline_prototype/Fusion.compute_weights_testing.ipynb\n",
      "Resolving github.com (github.com)... 140.82.114.3\n",
      "Connecting to github.com (github.com)|140.82.114.3|:443... connected.\n",
      "HTTP request sent, awaiting response... 404 Not Found\n",
      "2020-07-13 14:31:26 ERROR 404: Not Found.\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "wget https://github.com/cumc/neuro-twas/blob/master/WIP/pipeline_prototype/Fusion.compute_weights_testing.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "cd tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[0;31m'utf-8' codec can't decode byte 0xa4 in position 1: invalid start byte\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "cd tmp/1_2/\n",
    "ls\n",
    "cat GD462.hsq_succ.test.txt.ENSG00000136237.12..tmp.cv.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[0;31m'utf-8' codec can't decode byte 0x8c in position 3: invalid start byte\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "rm -r *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cat: 1_10.hsq: No such file or directory\n",
      "0\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cat 1_10.hsq | grep . -c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "echo $pop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "head: /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/LDREF/1000G.10.bed: No such file or directory\n",
      "\n"
     ]
    }
   ],
   "source": [
    "head /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/LDREF/1000G.10.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WEIGHTS/ALL/ALL.ENSG00000127054.14.wgt.RDat\n",
      "WEIGHTS/ALL/ALL.ENSG00000169972.7.wgt.RDat\n",
      "WEIGHTS/ALL/ALL.ENSG00000224051.2.wgt.RDat\n",
      "WEIGHTS/ALL/ALL.ENSG00000272455.1.wgt.RDat\n",
      "\n"
     ]
    }
   ],
   "source": [
    "ls WEIGHTS/*/*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "patential working genes\n",
    "ENSG00000225267.1\n",
    "ENSG00000259382.1\n",
    "ENSG00000135046.9"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "$OUT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TargetID\tGene_Symbol\tChr\tCoord\tHG00096\tHG00097\tHG00099\tHG00100\tHG00101\tHG00102\tHG00103\tHG00104\tHG00105\tHG00106\tHG00108\tHG00109\tHG00110\tHG00111\tHG00112\tHG00114\tHG00115\tHG00116\tHG00117\tHG00118\tHG00119\tHG00120\tHG00121\tHG00122\tHG00123\tHG00124\tHG00125\tHG00126\tHG00127\tHG00128\tHG00129\tHG00130\tHG00131\tHG00132\tHG00133\tHG00134\tHG00135\tHG00136\tHG00137\tHG00138\tHG00139\tHG00141\tHG00142\tHG00143\tHG00145\tHG00146\tHG00148\tHG00149\tHG00150\tHG00151\tHG00152\tHG00154\tHG00155\tHG00156\tHG00157\tHG00158\tHG00159\tHG00160\tHG00171\tHG00173\tHG00174\tHG00176\tHG00177\tHG00178\tHG00179\tHG00180\tHG00181\tHG00182\tHG00183\tHG00185\tHG00186\tHG00187\tHG00188\tHG00189\tHG00231\tHG00232\tHG00233\tHG00234\tHG00235\tHG00236\tHG00238\tHG00239\tHG00240\tHG00242\tHG00243\tHG00244\tHG00245\tHG00246\tHG00247\tHG00249\tHG00250\tHG00251\tHG00252\tHG00253\tHG00255\tHG00256\tHG00257\tHG00258\tHG00259\tHG00260\tHG00261\tHG00262\tHG00263\tHG00264\tHG00265\tHG00266\tHG00267\tHG00268\tHG00269\tHG00271\tHG00272\tHG00273\tHG00274\tHG00275\tHG00276\tHG00277\tHG00278\tHG00280\tHG00281\tHG00282\tHG00284\tHG00285\tHG00306\tHG00308\tHG00309\tHG00310\tHG00311\tHG00312\tHG00313\tHG00315\tHG00319\tHG00320\tHG00321\tHG00323\tHG00324\tHG00325\tHG00326\tHG00327\tHG00328\tHG00329\tHG00330\tHG00331\tHG00332\tHG00334\tHG00335\tHG00336\tHG00337\tHG00338\tHG00339\tHG00341\tHG00342\tHG00343\tHG00344\tHG00345\tHG00346\tHG00349\tHG00350\tHG00351\tHG00353\tHG00355\tHG00356\tHG00358\tHG00359\tHG00360\tHG00361\tHG00362\tHG00364\tHG00365\tHG00366\tHG00367\tHG00369\tHG00371\tHG00372\tHG00373\tHG00375\tHG00376\tHG00377\tHG00378\tHG00379\tHG00380\tHG00381\tHG00382\tHG00383\tHG00384\tHG01334\tHG01789\tHG01790\tHG01791\tHG02215\tNA06984\tNA06985\tNA06986\tNA06989\tNA06994\tNA07037\tNA07048\tNA07051\tNA07056\tNA07346\tNA07347\tNA07357\tNA10847\tNA10851\tNA11829\tNA11830\tNA11831\tNA11832\tNA11840\tNA11843\tNA11881\tNA11892\tNA11893\tNA11894\tNA11918\tNA11920\tNA11930\tNA11931\tNA11992\tNA11993\tNA11994\tNA11995\tNA12004\tNA12005\tNA12006\tNA12043\tNA12044\tNA12045\tNA12058\tNA12144\tNA12154\tNA12155\tNA12156\tNA12234\tNA12249\tNA12272\tNA12273\tNA12275\tNA12282\tNA12283\tNA12286\tNA12287\tNA12340\tNA12341\tNA12342\tNA12347\tNA12348\tNA12383\tNA12399\tNA12400\tNA12413\tNA12489\tNA12546\tNA12716\tNA12717\tNA12718\tNA12749\tNA12750\tNA12751\tNA12760\tNA12761\tNA12762\tNA12763\tNA12775\tNA12776\tNA12777\tNA12778\tNA12812\tNA12813\tNA12814\tNA12815\tNA12827\tNA12829\tNA12830\tNA12842\tNA12843\tNA12872\tNA12873\tNA12874\tNA12889\tNA12890\tNA18486\tNA18487\tNA18488\tNA18489\tNA18498\tNA18499\tNA18502\tNA18505\tNA18508\tNA18510\tNA18511\tNA18517\tNA18519\tNA18520\tNA18858\tNA18861\tNA18867\tNA18868\tNA18870\tNA18873\tNA18907\tNA18908\tNA18909\tNA18910\tNA18912\tNA18916\tNA18917\tNA18923\tNA18933\tNA18934\tNA19092\tNA19093\tNA19095\tNA19096\tNA19098\tNA19099\tNA19102\tNA19107\tNA19108\tNA19113\tNA19114\tNA19116\tNA19117\tNA19118\tNA19119\tNA19121\tNA19129\tNA19130\tNA19131\tNA19137\tNA19138\tNA19141\tNA19143\tNA19144\tNA19146\tNA19147\tNA19149\tNA19150\tNA19152\tNA19153\tNA19159\tNA19160\tNA19171\tNA19172\tNA19175\tNA19184\tNA19185\tNA19189\tNA19190\tNA19197\tNA19198\tNA19200\tNA19201\tNA19204\tNA19206\tNA19207\tNA19209\tNA19210\tNA19213\tNA19214\tNA19222\tNA19223\tNA19225\tNA19235\tNA19236\tNA19247\tNA19248\tNA19256\tNA19257\tNA20502\tNA20503\tNA20504\tNA20505\tNA20506\tNA20507\tNA20508\tNA20509\tNA20510\tNA20512\tNA20513\tNA20514\tNA20515\tNA20516\tNA20517\tNA20518\tNA20519\tNA20520\tNA20521\tNA20524\tNA20525\tNA20527\tNA20528\tNA20529\tNA20530\tNA20531\tNA20532\tNA20534\tNA20535\tNA20536\tNA20537\tNA20538\tNA20539\tNA20540\tNA20541\tNA20542\tNA20543\tNA20544\tNA20581\tNA20582\tNA20585\tNA20586\tNA20588\tNA20589\tNA20752\tNA20754\tNA20756\tNA20757\tNA20758\tNA20759\tNA20760\tNA20761\tNA20765\tNA20766\tNA20768\tNA20769\tNA20770\tNA20771\tNA20772\tNA20773\tNA20774\tNA20778\tNA20783\tNA20785\tNA20786\tNA20787\tNA20790\tNA20792\tNA20795\tNA20796\tNA20797\tNA20798\tNA20799\tNA20800\tNA20801\tNA20802\tNA20803\tNA20804\tNA20805\tNA20806\tNA20807\tNA20808\tNA20809\tNA20810\tNA20811\tNA20812\tNA20813\tNA20814\tNA20815\tNA20816\tNA20819\tNA20826\tNA20828\n",
      "\n"
     ]
    }
   ],
   "source": [
    "head -1 $Datarep$PRE_GEXP"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "Bash"
   },
   "source": [
    "# Association Test\n",
    "The following section is for the assoc test, which is ready to use off the shelf, tne triky part is getting the weights"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "Rscr2=\"<>\"\n",
    "# This is the new input file\n",
    "sumstats=\"<>\"\n",
    "weights=FINAL_OUT\n",
    "weights_dir=$wd/WEIGHTS/\n",
    "ref_ld_chr=$LDREF/PRE_GENO\n",
    "chr=\"<>\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [],
   "source": [
    "Rscript $Rscr2 \\\n",
    "--sumstats $sumstats \\\n",
    "--weights $weights \\\n",
    "--weights_dir $weights_dir \\\n",
    "--ref_ld_chr $LDREF/PRE_GENO \\\n",
    "--chr 22 \\\n",
    "--out PGC2.SCZ.22.dat"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "Bash",
     "calysto_bash",
     "Bash",
     "#E6EEFF",
     ""
    ]
   ],
   "version": "0.20.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
