{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Chicago Data TWAS\n",
    "## Motivation\n",
    "Before actual analysis of our own data, the pipeline shall first be used to try to reproduce the work that are done previously. So that the result can be compared.\n",
    "\n",
    "## Data set\n",
    "The data set is the gene expresion data from African population got from chicago university. It contains the expression profile of 50 genes from 89 individuals. The gene expression is processed by STAR and recorded as RPKM. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Tools\n",
    "The tools used in this analysis is the \"SOS_weight_cpt_template\" pipeline, with the options outlined as followed. Developed on the fusion foundation, this pipeline takes in gene expression data ans SNP to compute the weight for the follow ups Correlation analysis. association testing are then conducted by the standard Fusion scripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: sos run SOS_weight_cpt_template.ipynb\n",
      "               [workflow_name | -t targets] [options] [workflow_options]\n",
      "  workflow_name:        Single or combined workflows defined in this script\n",
      "  targets:              One or more targets to generate\n",
      "  options:              Single-hyphen sos parameters (see \"sos run -h\" for details)\n",
      "  workflow_options:     Double-hyphen workflow-specific parameters\n",
      "\n",
      "Workflows:\n",
      "  STEP\n",
      "\n",
      "Global Workflow Options:\n",
      "  --GCTA '/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/gcta_1.93.2beta_mac/gcta64'\n",
      "                        !/bin/sh MAKE SURE FUSION.compute_weights.R IS IN YOUR\n",
      "                        PATH FILL IN THESE PATHS For mac user, the mac version\n",
      "                        of GCTA shall be downloaded saperately, the one came\n",
      "                        with the Fusion package will not work.\n",
      "  --PLINK '/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/plink_mac_20200616/plink'\n",
      "  --GEMMA '/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/GEMMA'\n",
      "  --Rscp '/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/FUSION.compute_weights.mod.R'\n",
      "                        Required the customized fusion.compute_weight.mod.R\n",
      "                        script, other wise will not work\n",
      "  --Datafile '/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/Testing/Data/GD462.hsq_succ.test.txt'\n",
      "                        Path to the input data\n",
      "  --wd '/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/SOS'\n",
      "                        PATH TO WORKING DIRECTORY\n",
      "  --OUT-DIR './WEIGHTS'\n",
      "                        PATH TO OUTPUT DIRECTORY (population-specific subdirs\n",
      "                        will be made)\n",
      "  --BATCH-START 1 (as int)\n",
      "                        ROWS IN THE MATRIX TO ANALYZE (FOR BATCHED RUNS)\n",
      "  --BATCH-END 3 (as int)\n",
      "  --LDREF '/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/LDREF'\n",
      "                        PATH TO DIRECTORY CONTAINING LDREF DATA (FROM FUSION\n",
      "                        WEBSITE or https://data.broadinstitute.org/alkesgroup/FU\n",
      "                        SION/LDREF.tar.bz2)\n",
      "  --PRE-GEXP 'GD462.hsq_succ.test.txt'\n",
      "                        PATH TO GEUVADIS GENE EXPRESSION MATRIX:\n",
      "  --PRE-GENO '1000G.EUR'\n",
      "                        PATH TO PREFIX FOR GEUVADIS GENOTYPES SPLIT BY\n",
      "                        CHROMOSOME SUBSAMPLE THESE TO THE LDREF SNPS FOR\n",
      "                        EFFICIENCY\n",
      "  --chrm 3 (as int)\n",
      "                        Specify the column in the genexpression file that\n",
      "                        contains chromosome\n",
      "  --pos 4 (as int)\n",
      "                        Specify the column in the genexpression file that\n",
      "                        contains the position of the gene\n",
      "  --start VAL (as int, required)\n",
      "                        If both the start and end region are specified, then\n",
      "                        their column can be specified saperately\n",
      "  --end VAL (as int, required)\n",
      "  --gname 1 (as int)\n",
      "                        Specify the column in the genexpression file that\n",
      "                        contains the name of the gene\n",
      "  --window 50000 (as int)\n",
      "                        Specify the scanning window for the gene position, set\n",
      "                        default to 50000 if start = end\n",
      "\n",
      "Sections\n",
      "  STEP_1:\n",
      "  STEP_2:\n",
      "\n"
     ]
    }
   ],
   "source": [
    "sos run SOS_weight_cpt_template.ipynb -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "Bash"
   },
   "source": [
    "## Test pipeline with example datas\n",
    "The sos pipeline were tested using the provided working example, exhibiting similar behavior as described in the documents.\n",
    "\n",
    "sos run SOS_weight_cpt_template.ipynb \\\n",
    "  --GCTA \"/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/gcta_1.93.2beta_mac/gcta64\" \\\n",
    "  --PLINK `which plink` \\\n",
    "  --GEMMA `which gemma` \\\n",
    "  --Rscp  \"/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/FUSION.compute_weights.mod.R\" \\\n",
    "  --Datafile \"/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/Testing/Data/GD462.hsq_succ.test.txt\" \\\n",
    "  --wd  \"/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/SOS\" \\\n",
    "  --OUT_DIR \"./\" \\\n",
    "  --LDREF  \"/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/LDREF\" \\\n",
    "  --PRE_GENO \"1000G.EUR\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "Bash"
   },
   "source": [
    "# Data wrangling\n",
    "Before conducting analysis on the chicago data, the following issues needed to be fixed via the data wrangling steps:\n",
    "1. The sample ID is not equal to the bfile ID\n",
    "Note: Please disregard the warning message that"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "kernel": "R"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Parsed with column specification:\n",
      "cols(\n",
      "  .default = col_double(),\n",
      "  `#Chr` = \u001b[31mcol_character()\u001b[39m,\n",
      "  ID = \u001b[31mcol_character()\u001b[39m\n",
      ")\n",
      "\n",
      "See spec(...) for full column specifications.\n",
      "\n",
      "Warning message:\n",
      "“1 parsing failure.\n",
      "row col   expected     actual                                                                                                                                      file\n",
      " 49  -- 89 columns 74 columns '/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt'\n",
      "”\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'#Chr'</li><li>'start'</li><li>'end'</li><li>'ID'</li><li>'NA18486'</li><li>'NA18487'</li><li>'NA18488'</li><li>'NA18489'</li><li>'NA18498'</li><li>'NA18499'</li><li>'NA18502'</li><li>'NA18505'</li><li>'NA18508'</li><li>'NA18510'</li><li>'NA18511'</li><li>'NA18517'</li><li>'NA18519'</li><li>'NA18520'</li><li>'NA18858'</li><li>'NA18861'</li><li>'NA18867'</li><li>'NA18868'</li><li>'NA18870'</li><li>'NA18873'</li><li>'NA18907'</li><li>'NA18909'</li><li>'NA18910'</li><li>'NA18912'</li><li>'NA18916'</li><li>'NA18917'</li><li>'NA18923'</li><li>'NA18933'</li><li>'NA18934'</li><li>'NA19093'</li><li>'NA19095'</li><li>'NA19096'</li><li>'NA19098'</li><li>'NA19099'</li><li>'NA19102'</li><li>'NA19107'</li><li>'NA19108'</li><li>'NA19113'</li><li>'NA19114'</li><li>'NA19116'</li><li>'NA19117'</li><li>'NA19118'</li><li>'NA19119'</li><li>'NA19121'</li><li>'NA19129'</li><li>'NA19130'</li><li>'NA19131'</li><li>'NA19137'</li><li>'NA19138'</li><li>'NA19141'</li><li>'NA19143'</li><li>'NA19144'</li><li>'NA19146'</li><li>'NA19147'</li><li>'NA19149'</li><li>'NA19150'</li><li>'NA19152'</li><li>'NA19153'</li><li>'NA19159'</li><li>'NA19171'</li><li>'NA19172'</li><li>'NA19175'</li><li>'NA19184'</li><li>'NA19185'</li><li>'NA19189'</li><li>'NA19190'</li><li>'NA19197'</li><li>'NA19198'</li><li>'NA19200'</li><li>'NA19201'</li><li>'NA19204'</li><li>'NA19206'</li><li>'NA19207'</li><li>'NA19209'</li><li>'NA19210'</li><li>'NA19213'</li><li>'NA19222'</li><li>'NA19223'</li><li>'NA19225'</li><li>'NA19235'</li><li>'NA19236'</li><li>'NA19247'</li><li>'NA19248'</li><li>'NA19256'</li><li>'NA19257'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item '\\#Chr'\n",
       "\\item 'start'\n",
       "\\item 'end'\n",
       "\\item 'ID'\n",
       "\\item 'NA18486'\n",
       "\\item 'NA18487'\n",
       "\\item 'NA18488'\n",
       "\\item 'NA18489'\n",
       "\\item 'NA18498'\n",
       "\\item 'NA18499'\n",
       "\\item 'NA18502'\n",
       "\\item 'NA18505'\n",
       "\\item 'NA18508'\n",
       "\\item 'NA18510'\n",
       "\\item 'NA18511'\n",
       "\\item 'NA18517'\n",
       "\\item 'NA18519'\n",
       "\\item 'NA18520'\n",
       "\\item 'NA18858'\n",
       "\\item 'NA18861'\n",
       "\\item 'NA18867'\n",
       "\\item 'NA18868'\n",
       "\\item 'NA18870'\n",
       "\\item 'NA18873'\n",
       "\\item 'NA18907'\n",
       "\\item 'NA18909'\n",
       "\\item 'NA18910'\n",
       "\\item 'NA18912'\n",
       "\\item 'NA18916'\n",
       "\\item 'NA18917'\n",
       "\\item 'NA18923'\n",
       "\\item 'NA18933'\n",
       "\\item 'NA18934'\n",
       "\\item 'NA19093'\n",
       "\\item 'NA19095'\n",
       "\\item 'NA19096'\n",
       "\\item 'NA19098'\n",
       "\\item 'NA19099'\n",
       "\\item 'NA19102'\n",
       "\\item 'NA19107'\n",
       "\\item 'NA19108'\n",
       "\\item 'NA19113'\n",
       "\\item 'NA19114'\n",
       "\\item 'NA19116'\n",
       "\\item 'NA19117'\n",
       "\\item 'NA19118'\n",
       "\\item 'NA19119'\n",
       "\\item 'NA19121'\n",
       "\\item 'NA19129'\n",
       "\\item 'NA19130'\n",
       "\\item 'NA19131'\n",
       "\\item 'NA19137'\n",
       "\\item 'NA19138'\n",
       "\\item 'NA19141'\n",
       "\\item 'NA19143'\n",
       "\\item 'NA19144'\n",
       "\\item 'NA19146'\n",
       "\\item 'NA19147'\n",
       "\\item 'NA19149'\n",
       "\\item 'NA19150'\n",
       "\\item 'NA19152'\n",
       "\\item 'NA19153'\n",
       "\\item 'NA19159'\n",
       "\\item 'NA19171'\n",
       "\\item 'NA19172'\n",
       "\\item 'NA19175'\n",
       "\\item 'NA19184'\n",
       "\\item 'NA19185'\n",
       "\\item 'NA19189'\n",
       "\\item 'NA19190'\n",
       "\\item 'NA19197'\n",
       "\\item 'NA19198'\n",
       "\\item 'NA19200'\n",
       "\\item 'NA19201'\n",
       "\\item 'NA19204'\n",
       "\\item 'NA19206'\n",
       "\\item 'NA19207'\n",
       "\\item 'NA19209'\n",
       "\\item 'NA19210'\n",
       "\\item 'NA19213'\n",
       "\\item 'NA19222'\n",
       "\\item 'NA19223'\n",
       "\\item 'NA19225'\n",
       "\\item 'NA19235'\n",
       "\\item 'NA19236'\n",
       "\\item 'NA19247'\n",
       "\\item 'NA19248'\n",
       "\\item 'NA19256'\n",
       "\\item 'NA19257'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. '#Chr'\n",
       "2. 'start'\n",
       "3. 'end'\n",
       "4. 'ID'\n",
       "5. 'NA18486'\n",
       "6. 'NA18487'\n",
       "7. 'NA18488'\n",
       "8. 'NA18489'\n",
       "9. 'NA18498'\n",
       "10. 'NA18499'\n",
       "11. 'NA18502'\n",
       "12. 'NA18505'\n",
       "13. 'NA18508'\n",
       "14. 'NA18510'\n",
       "15. 'NA18511'\n",
       "16. 'NA18517'\n",
       "17. 'NA18519'\n",
       "18. 'NA18520'\n",
       "19. 'NA18858'\n",
       "20. 'NA18861'\n",
       "21. 'NA18867'\n",
       "22. 'NA18868'\n",
       "23. 'NA18870'\n",
       "24. 'NA18873'\n",
       "25. 'NA18907'\n",
       "26. 'NA18909'\n",
       "27. 'NA18910'\n",
       "28. 'NA18912'\n",
       "29. 'NA18916'\n",
       "30. 'NA18917'\n",
       "31. 'NA18923'\n",
       "32. 'NA18933'\n",
       "33. 'NA18934'\n",
       "34. 'NA19093'\n",
       "35. 'NA19095'\n",
       "36. 'NA19096'\n",
       "37. 'NA19098'\n",
       "38. 'NA19099'\n",
       "39. 'NA19102'\n",
       "40. 'NA19107'\n",
       "41. 'NA19108'\n",
       "42. 'NA19113'\n",
       "43. 'NA19114'\n",
       "44. 'NA19116'\n",
       "45. 'NA19117'\n",
       "46. 'NA19118'\n",
       "47. 'NA19119'\n",
       "48. 'NA19121'\n",
       "49. 'NA19129'\n",
       "50. 'NA19130'\n",
       "51. 'NA19131'\n",
       "52. 'NA19137'\n",
       "53. 'NA19138'\n",
       "54. 'NA19141'\n",
       "55. 'NA19143'\n",
       "56. 'NA19144'\n",
       "57. 'NA19146'\n",
       "58. 'NA19147'\n",
       "59. 'NA19149'\n",
       "60. 'NA19150'\n",
       "61. 'NA19152'\n",
       "62. 'NA19153'\n",
       "63. 'NA19159'\n",
       "64. 'NA19171'\n",
       "65. 'NA19172'\n",
       "66. 'NA19175'\n",
       "67. 'NA19184'\n",
       "68. 'NA19185'\n",
       "69. 'NA19189'\n",
       "70. 'NA19190'\n",
       "71. 'NA19197'\n",
       "72. 'NA19198'\n",
       "73. 'NA19200'\n",
       "74. 'NA19201'\n",
       "75. 'NA19204'\n",
       "76. 'NA19206'\n",
       "77. 'NA19207'\n",
       "78. 'NA19209'\n",
       "79. 'NA19210'\n",
       "80. 'NA19213'\n",
       "81. 'NA19222'\n",
       "82. 'NA19223'\n",
       "83. 'NA19225'\n",
       "84. 'NA19235'\n",
       "85. 'NA19236'\n",
       "86. 'NA19247'\n",
       "87. 'NA19248'\n",
       "88. 'NA19256'\n",
       "89. 'NA19257'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"#Chr\"    \"start\"   \"end\"     \"ID\"      \"NA18486\" \"NA18487\" \"NA18488\"\n",
       " [8] \"NA18489\" \"NA18498\" \"NA18499\" \"NA18502\" \"NA18505\" \"NA18508\" \"NA18510\"\n",
       "[15] \"NA18511\" \"NA18517\" \"NA18519\" \"NA18520\" \"NA18858\" \"NA18861\" \"NA18867\"\n",
       "[22] \"NA18868\" \"NA18870\" \"NA18873\" \"NA18907\" \"NA18909\" \"NA18910\" \"NA18912\"\n",
       "[29] \"NA18916\" \"NA18917\" \"NA18923\" \"NA18933\" \"NA18934\" \"NA19093\" \"NA19095\"\n",
       "[36] \"NA19096\" \"NA19098\" \"NA19099\" \"NA19102\" \"NA19107\" \"NA19108\" \"NA19113\"\n",
       "[43] \"NA19114\" \"NA19116\" \"NA19117\" \"NA19118\" \"NA19119\" \"NA19121\" \"NA19129\"\n",
       "[50] \"NA19130\" \"NA19131\" \"NA19137\" \"NA19138\" \"NA19141\" \"NA19143\" \"NA19144\"\n",
       "[57] \"NA19146\" \"NA19147\" \"NA19149\" \"NA19150\" \"NA19152\" \"NA19153\" \"NA19159\"\n",
       "[64] \"NA19171\" \"NA19172\" \"NA19175\" \"NA19184\" \"NA19185\" \"NA19189\" \"NA19190\"\n",
       "[71] \"NA19197\" \"NA19198\" \"NA19200\" \"NA19201\" \"NA19204\" \"NA19206\" \"NA19207\"\n",
       "[78] \"NA19209\" \"NA19210\" \"NA19213\" \"NA19222\" \"NA19223\" \"NA19225\" \"NA19235\"\n",
       "[85] \"NA19236\" \"NA19247\" \"NA19248\" \"NA19256\" \"NA19257\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "##Ad hoc Data wrangling for the chicago data\n",
    "library(\"dplyr\")\n",
    "to_fix = readr::read_delim(\"/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/fastqtl_qqnorm_RNAseqGeuvadis_phase2.txt\",delim = \"\\t\")\n",
    "colnames(to_fix)[5:ncol(to_fix)] = paste(\"NA\",colnames(to_fix)[5:ncol(to_fix)])%>%gsub(\" \",\"\",.)\n",
    "to_fix%>%readr::write_tsv( path = \"/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.txt\", na = \"NA\", append = FALSE, col_names = TRUE, quote_escape = \"double\")\n",
    "colnames(to_fix)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "R"
   },
   "source": [
    "## SVA correction (In progress)\n",
    "Hidden factor is addressed for each genes using the sva method. Current obstacle of implementing this section is to create a model matrix with both the gene expression data and the snp information in the same matrixs for every genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "R"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: Running \u001b[32mSTEP_1\u001b[0m: Make folder structure for the pipeline\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=0) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=1) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=2) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=3) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=4) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=5) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=6) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=7) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=8) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=9) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=10) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=11) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=12) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=13) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=14) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=15) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=16) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=17) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=18) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=19) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=20) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=21) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=22) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=23) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=24) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=27) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=28) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=29) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=30) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=31) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=32) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=33) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=34) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=35) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=36) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=37) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=39) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=40) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=41) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=43) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=44) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=45) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=46) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=47) is \u001b[32mignored\u001b[0m due to saved signature\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=26) is \u001b[32mcompleted\u001b[0m.\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=25) is \u001b[32mcompleted\u001b[0m.\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=38) is \u001b[32mcompleted\u001b[0m.\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m (index=42) is \u001b[32mcompleted\u001b[0m.\n",
      "INFO: \u001b[32mSTEP_1\u001b[0m output:   \u001b[32m/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed_per_gene/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000217801.5.txt /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000217801.5.pheno... (240 items in 48 groups)\u001b[0m\n",
      "INFO: Running \u001b[32mSTEP_2\u001b[0m: Actual weight computation analysis\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000187608.5.tmp/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000217801.5.tmp : heritability  : heritability 0.0152820.07194; LRT P-value ; LRT P-value 0.439980.24629 : skipping gene\n",
      " : skipping gene\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000175756.9.tmp : heritability -0.038087; LRT P-value 0.26877 : skipping gene\n",
      "end of circle\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=0) is \u001b[32mcompleted\u001b[0m.\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=2) is \u001b[32mcompleted\u001b[0m.\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=1) is \u001b[32mcompleted\u001b[0m.\n",
      "WARNING : 7 SNPs could not be scaled and were zeroed out, make sure all SNPs are polymorphic\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=3) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000189339.7.tmp : heritability 0.050823; LRT P-value 0.34259 : skipping gene\n",
      "WARNING : 5 SNPs could not be scaled and were zeroed out, make sure all SNPs are polymorphic\n",
      "WARNING : 34 SNPs could not be scaled and were zeroed out, make sure all SNPs are polymorphic\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=4) is \u001b[32mcompleted\u001b[0m.\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=6) is \u001b[32mcompleted\u001b[0m.\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=5) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000197530.8.tmp : heritability 0.094787; LRT P-value 0.30092 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=7) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000188976.6.tmp : heritability -0.056268; LRT P-value 0.18715 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=8) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000160075.10.tmp : heritability -0.02687; LRT P-value 0.31087 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=9) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000187634.6.tmp : heritability 0.010971; LRT P-value 0.44755 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=10) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000269227.1.tmp : heritability -0.024469; LRT P-value 0.35222 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=11) is \u001b[32mcompleted\u001b[0m.\n",
      "WARNING : 78 SNPs could not be scaled and were zeroed out, make sure all SNPs are polymorphic\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000242485.1.tmp : heritability 0.070258; LRT P-value 0.23683 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=13) is \u001b[32mcompleted\u001b[0m.\n",
      "WARNING : 37 SNPs could not be scaled and were zeroed out, make sure all SNPs are polymorphic\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000224870.3.tmp : heritability 0.065597; LRT P-value 0.21875 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=15) is \u001b[32mcompleted\u001b[0m.\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=12) is \u001b[32mcompleted\u001b[0m.\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=14) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000228594.1.tmp : heritability 0.081225; LRT P-value 0.23179 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=16) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000248527.1.tmp does not exist, likely GCTA could not converge, skipping gene\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000188157.9.tmp : heritability 0.042024; LRT P-value 0.2373 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=18) is \u001b[32mcompleted\u001b[0m.\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=17) is \u001b[32mcompleted\u001b[0m.\n",
      "WARNING : 46 SNPs could not be scaled and were zeroed out, make sure all SNPs are polymorphic\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000198744.5.tmp does not exist, likely GCTA could not converge, skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=20) is \u001b[32mcompleted\u001b[0m.\n",
      "WARNING : 12 SNPs could not be scaled and were zeroed out, make sure all SNPs are polymorphic\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=19) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000187961.9.tmp : heritability 0.018761; LRT P-value 0.38942 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=22) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000169972.7.tmp : heritability 0.026709; LRT P-value 0.40638 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=23) is \u001b[32mcompleted\u001b[0m.\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=21) is \u001b[32mcompleted\u001b[0m.\n",
      "WARNING : 33 SNPs could not be scaled and were zeroed out, make sure all SNPs are polymorphic\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000078808.12.tmp : heritability 0.096015; LRT P-value 0.056635 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=25) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000179403.10.tmp : heritability 0.099418; LRT P-value 0.21384 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=26) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000225630.1.tmp does not exist, likely GCTA could not converge, skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=27) is \u001b[32mcompleted\u001b[0m.\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=24) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000186827.6.tmp : heritability 0.001913; LRT P-value 0.49225 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=28) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000237973.1.tmp does not exist, likely GCTA could not converge, skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=29) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000107404.13.tmp : heritability -0.075739; LRT P-value 0.1327 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=30) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000221978.7.tmp : heritability -0.122755; LRT P-value 0.10656 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=31) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000225972.1.tmp does not exist, likely GCTA could not converge, skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=32) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000264293.1.tmp does not exist, likely GCTA could not converge, skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=33) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000186891.9.tmp : heritability 0.084467; LRT P-value 0.14518 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=34) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000176022.3.tmp : heritability 0.139452; LRT P-value 0.056858 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=35) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000242590.1.tmp : heritability -0.875086; LRT P-value 0.0071512 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=36) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000229344.1.tmp does not exist, likely GCTA could not converge, skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=37) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000160072.15.tmp : heritability 0.114979; LRT P-value 0.27578 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=38) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000215791.2.tmp does not exist, likely GCTA could not converge, skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=39) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000269737.1.tmp : heritability 0.161498; LRT P-value 0.092918 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=40) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000272106.1.tmp does not exist, likely GCTA could not converge, skipping gene\n",
      "cat: /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/WEIGHTS/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000272106.1.hsq: No such file or directory\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=41) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000224051.2.tmp : heritability 0.088467; LRT P-value 0.10675 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=42) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000188290.6.tmp does not exist, likely GCTA could not converge, skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=43) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000160087.16.tmp : heritability 0.154213; LRT P-value 0.078429 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=44) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000131591.13.tmp : heritability 0.193209; LRT P-value 0.038544 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=45) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000240731.1.tmp does not exist, likely GCTA could not converge, skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=46) is \u001b[32mcompleted\u001b[0m.\n",
      "/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/tmp/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000248333.3.tmp : heritability 0.294325; LRT P-value 0.043677 : skipping gene\n",
      "end of circle\n",
      "INFO: \u001b[32mSTEP_2\u001b[0m (index=47) is \u001b[32mcompleted\u001b[0m.\n",
      "\u001b[91mERROR\u001b[0m: \u001b[91m[STEP_2]: [STEP_2]: Output target /Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working/WEIGHTS/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.ENSG00000217801.5.wgt.RDat does not exist after the completion of step STEP_2\u001b[0m\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "## Test pipeline with chicago data\n",
    "sos_pipeline=\"../Workflow/SOS_weight_cpt_template.ipynb\"\n",
    "sos run ../Workflow/SOS_weight_cpt_template.ipynb \\\n",
    "  --GCTA \"/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/gcta_1.93.2beta_mac/gcta64\" \\\n",
    "  --PLINK `which plink` \\\n",
    "  --GEMMA `which gemma` \\\n",
    "  --Rscp  \"/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/FUSION.compute_weights.mod.R\" \\\n",
    "  --Datafile \"/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/fastqtl_qqnorm_RNAseqGeuvadis_phase2_fixed.txt\" \\\n",
    "  --wd  \"/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/working\" \\\n",
    "  --LDREF  \"/Users/haosun/Documents/WG_Reasearch_Assisstant/Fusion/install/fusion_twas-master/chicago/Data/genotype_plink_bfiles\" \\\n",
    "  --PRE_GENO \"YRI.120samples.hg19\" \\\n",
    "  --chrm 1 \\\n",
    "  --gname 4 \\\n",
    "  --start 2 \\\n",
    "  --end 3 \\\n",
    "  --window 1000\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Result\n",
    "The above scripts was ran 3 times before the desired output was generated.\n",
    "\n",
    "When scanning 1000bp before and after the gene region. Out of 50 genes, 10 of them passed the heritability test and can be passed down toward the followup analysis.\n",
    "\n",
    "When scanning 500000bp before and after the gene starting site, out of 50 genes, 4 of them passed the heritability test and can be passed down toward the followup analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [],
   "source": [
    "## Potential graphing with the hsq distribution"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Association testing\n",
    "Association testing shall be done using the following scripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"sos_logging sos_info\">INFO: Running <span class=\"sos_highlight\">example</span>: \n",
       "</div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[91mERROR\u001b[0m: \u001b[91m[example]: [example]: Output target $a.txt does not exist after the completion of step example\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "Rscr2=\"\"\n",
    "sumstats=\"\"\n",
    "weights=\"\"\n",
    "weights_di=\"\"\n",
    "LDREF=\"\"\n",
    "PRE_GENO=\"\"\n",
    "out=\"\"\n",
    "## \n",
    "Rscript $Rscr2 \\\n",
    "--sumstats $sumstats \\\n",
    "--weights $weights \\\n",
    "--weights_dir $weights_dir \\\n",
    "--ref_ld_chr $LDREF/PRE_GENO \\\n",
    "--chr 22 \\\n",
    "--out $out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "Python3"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "R"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "Python3"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "Bash",
     "calysto_bash",
     "Bash",
     "#E6EEFF",
     ""
    ],
    [
     "Python3",
     "python3",
     "Python3",
     "#FFD91A",
     {
      "name": "ipython",
      "version": 3
     }
    ],
    [
     "R",
     "ir",
     "R",
     "#DCDCDA",
     "r"
    ],
    [
     "SoS",
     "sos",
     "",
     "",
     "sos"
    ]
   ],
   "version": "0.20.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
