<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="ipynb_website:version" content="0.9.7" />
<meta name="viewport" content="width=device-width, initial-scale=1" />

<link rel="stylesheet" type="text/css" href="../css/jt.css">

<link rel="stylesheet" type="text/css" href="../css/toc2.css">

<link href="../site_libs/jqueryui-1.11.4/jquery-ui.css">
<link rel="stylesheet" href="../site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" href="../site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.9.1/jquery-ui.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>

<link rel="stylesheet"
      href="../site_libs/highlightjs/null.min.css"
      type="text/css" />

<script src="../site_libs/highlightjs/highlight.pack.js"></script>
<script>hljs.initHighlightingOnLoad();</script>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>

<script src="../js/doc_toc.js"></script>
<script src="../js/docs.js"></script>

<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-MML-AM_CHTML"></script>
<script>
    MathJax.Hub.Config({
        extensions: ["tex2jax.js"],
        jax: ["input/TeX", "output/HTML-CSS"],
        tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true
        },
        "HTML-CSS": {
            preferredFont: "TeX",
            availableFonts: ["TeX"],
            styles: {
                scale: 110,
                ".MathJax_Display": {
                    "font-size": "110%",
                }
            }
        }
    });
</script>
<script>
function filterDataFrame(id) {
    var input = document.getElementById("search_" + id);
    var filter = input.value.toUpperCase();
    var table = document.getElementById("dataframe_" + id);
    var tr = table.getElementsByTagName("tr");
    // Loop through all table rows, and hide those who don't match the search query
    for (var i = 1; i < tr.length; i++) {
        for (var j = 0; j < tr[i].cells.length; ++j) {
            var matched = false;
            if (tr[i].cells[j].innerHTML.toUpperCase().indexOf(filter) != -1) {
                tr[i].style.display = "";
                matched = true
                break;
            }
            if (!matched)
                tr[i].style.display = "none";
        }
    }
}
function sortDataFrame(id, n, dtype) {
    var table = document.getElementById("dataframe_" + id);
    var tb = table.tBodies[0]; // use `<tbody>` to ignore `<thead>` and `<tfoot>` rows
    var tr = Array.prototype.slice.call(tb.rows, 0); // put rows into array
    if (dtype === 'numeric') {
        var fn = function(a, b) { 
            return parseFloat(a.cells[n].textContent) <= parseFloat(b.cells[n].textContent) ? -1 : 1;
        }
    } else {
        var fn = function(a, b) {
            var c = a.cells[n].textContent.trim().localeCompare(b.cells[n].textContent.trim()); 
            return c > 0 ? 1 : (c < 0 ? -1 : 0) }
    }
    var isSorted = function(array, fn) {
        if (array.length < 2)
            return 1;
        var direction = fn(array[0], array[1]); 
        for (var i = 1; i < array.length - 1; ++i) {
            var d = fn(array[i], array[i+1]);
            if (d == 0)
                continue;
            else if (direction == 0)
                direction = d;
            else if (direction != d)
                return 0;
            }
        return direction;
    }
    var sorted = isSorted(tr, fn);
    if (sorted == 1 || sorted == -1) {
        // if sorted already, reverse it
        for(var i = tr.length - 1; i >= 0; --i)
            tb.appendChild(tr[i]); // append each row in order
    } else {
        tr = tr.sort(fn);
        for(var i = 0; i < tr.length; ++i)
            tb.appendChild(tr[i]); // append each row in order
    }
}
</script>

<script>
$( document ).ready(function(){
            var cfg={'threshold':3,     // depth of toc (number of levels)
             'number_sections': false,
             'toc_cell': false,          // useless here
             'toc_window_display': true, // display the toc window
             "toc_section_display": "block", // display toc contents in the window
             'sideBar':true,       // sidebar or floating window
             'navigate_menu':false       // navigation menu (only in liveNotebook -- do not change)
            }
            var st={};                  // some variables used in the script
            st.rendering_toc_cell = false;
            st.config_loaded = false;
            st.extension_initialized=false;
            st.nbcontainer_marginleft = $('#notebook-container').css('margin-left')
            st.nbcontainer_marginright = $('#notebook-container').css('margin-right')
            st.nbcontainer_width = $('#notebook-container').css('width')
            st.oldTocHeight = undefined
            st.cell_toc = undefined;
            st.toc_index=0;
            // fire the main function with these parameters
            table_of_contents(cfg, st);
            var file=workflowDict[$("h1:first").attr("id")];
            $("#toc-level0 a").css("color","#126dce");
            $('a[href="#'+$("h1:first").attr("id")+'"]').hide()
            var docs=workflowArray;
            var docs_map=workflowArrayMap;
            var pos=workflowArray.indexOf(file);
            for (var a=pos;a>=0;a--){
                  $('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>').insertBefore("#toc-level0 li:eq(0)");
            }
            $('a[href="'+file+'.html'+'"]').css("color","#126dce");
            for (var a=pos+1;a<docs.length;a++){
                  $(".toc #toc-level0").append('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>');
            }
            // $("#toc-header").hide(); // comment out because it prevents search bar from displaying
    });
</script>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');
  // mark it active
  menuAnchor.parent().addClass('active');
  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>
<div class="container-fluid main-container">
<!-- tabsets -->
<script src="../site_libs/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>



<title>Multi-tissue brain TWAS</title>

<style type = "text/css">
body {
  font-family: "Droid Sans";
  padding-top: 66px;
  padding-bottom: 40px;
}
</style>
</head>

<body>
<div tabindex="-1" id="notebook" class="border-box-sizing">
<div class="container" id="notebook-container">

<!-- code folding -->

<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">Multi-tissue brain TWAS</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
<li>
  <a href="../index.html">Overview</a>
</li>
        
<li>
  <a href="../writeup.html">Writeup</a>
</li>
        
<li>
  <a href="../analysis.html">Analysis</a>
</li>
        
<li>
  <a href="../workflow.html">Workflow</a>
</li>
        
      </ul>
        
<ul class="nav navbar-nav navbar-right">
<li>
   <a href="http://github.com/hsun3163/jnbinder"> <span class="fa fa-github"></span> </a>
</li>
</ul>
        
      </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Name-Change">Name Change<a class="anchor-link" href="#Name-Change">&#182;</a></h1><h2 id="Aim">Aim<a class="anchor-link" href="#Aim">&#182;</a></h2><p>The aim of this pipeline is to change the header of gene expression file based on a reference file that are made externally.</p>
<p>An example of the pipeline running is shown below</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[1]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="n">SOS</span> <span class="n">run</span> <span class="o">/</span><span class="n">Users</span><span class="o">/</span><span class="n">haosun</span><span class="o">/</span><span class="n">Documents</span><span class="o">/</span><span class="n">WG_Reasearch_Assisstant</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">freshcopy</span><span class="o">/</span><span class="n">neuro</span><span class="o">-</span><span class="n">twas</span><span class="o">/</span><span class="n">Workflow</span><span class="o">/</span><span class="n">SOS_ROSMAP_gene_exp_processing</span><span class="o">.</span><span class="n">ipynb</span> \
    <span class="o">--</span><span class="n">ref</span> <span class="s2">&quot;/Users/haosun/Documents/WG_Reasearch_Assisstant/Rosmap/Data/new_gene_keys_WGS_plink.txt&quot;</span> \
    <span class="o">--</span><span class="n">name_col</span> <span class="mi">2</span> \
    <span class="o">--</span><span class="n">real_name_col</span> <span class="mi">1</span> \
    <span class="o">--</span><span class="n">gene_exp</span> <span class="s2">&quot;/Users/haosun/Documents/WG_Reasearch_Assisstant/Rosmap/Data/gene_exp/geneCountsResidualsPlusBaselineAgeGenderAdj.txt&quot;</span> \
    <span class="o">--</span><span class="n">start_at</span> <span class="mi">2</span> \
    <span class="o">--</span><span class="n">output</span> <span class="s2">&quot;/Users/haosun/Documents/WG_Reasearch_Assisstant/Rosmap/Data/gene_exp/&quot;</span> \
    <span class="o">-</span><span class="n">j</span> <span class="mi">2</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>


<div class="output_subarea output_stream output_stderr output_text">
<pre>Keyboard Interrupt
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[71]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>/Users/haosun/Documents/WG_Reasearch_Assisstant/GIT/twas-dev/WIP/pipeline_prototype

</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[73]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># This pipeline changes the patient names of a gene_exp file based on a index reference file </span>
<span class="kn">parameter:</span> <span class="n">ref</span> <span class="o">=</span> <span class="s2">&quot;/Users/haosun/Documents/WG_Reasearch_Assisstant/Rosmap/Data/new_gene_keys_plink.txt&quot;</span>
<span class="c1"># A table contaiing the patient ID in the gene_exp file and the patient ID in the plink file</span>
<span class="kn">parameter:</span> <span class="n">name_col</span> <span class="o">=</span> <span class="mi">2</span>
<span class="c1"># Column position that specify the header of gene_exp in the index</span>
<span class="kn">parameter:</span> <span class="n">real_name_col</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Column position that specify the plink patient ID in the index</span>
<span class="kn">parameter:</span> <span class="n">gene_exp</span> <span class="o">=</span> <span class="s2">&quot;/Users/haosun/Documents/WG_Reasearch_Assisstant/Rosmap/Data/gene_exp/geneCountsResidualsPlusBaselineAgeGenderAdj.txt&quot;</span>
<span class="c1"># the patient ID to be changed</span>
<span class="kn">parameter:</span> <span class="n">start_at</span> <span class="o">=</span> <span class="mi">2</span>
<span class="c1"># the column where the patient ID start</span>

<span class="kn">parameter:</span> <span class="n">output</span> <span class="o">=</span> <span class="s2">&quot;/Users/haosun/Documents/WG_Reasearch_Assisstant/Rosmap/Data/&quot;</span>
<span class="c1"># where to store the output file</span>

<span class="c1"># Container option for software to run the analysis: docker or singularity</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;/home/hs3163/system_file/twas_latest.sif&#39;</span>

<span class="c1"># change the name of gene_exp, save the modified expression file and the file containing the patients ID whose genotype information are absent.</span>
<span class="p">[</span><span class="n">Processing_1</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">ref</span><span class="p">,</span> <span class="n">gene_exp</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{output}/{_input[1]:bn}_rename.txt&#39;</span><span class="p">,</span>
        <span class="n">f</span><span class="s1">&#39;{output}/{_input[1]:bn}_discardID.txt&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;4h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;6G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output[0]:bn}&#39;</span>
<span class="kn">R:</span>  <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span> <span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[1]:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[1]:n}.stdout&#39;</span><span class="p">,</span><span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">gene_exp</span> <span class="o">=</span> <span class="n">readr</span><span class="p">::</span><span class="n">read_delim</span><span class="p">(</span><span class="s2">&quot;$[_input[1]]&quot;</span><span class="p">,</span><span class="n">delim</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="p">)</span>
    <span class="n">index</span> <span class="o">=</span> <span class="n">readr</span><span class="p">::</span><span class="n">read_delim</span><span class="p">(</span><span class="s2">&quot;$[_input[0]]&quot;</span><span class="p">,</span><span class="n">delim</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="kp">name</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">Name</span> <span class="o">=</span> <span class="n">index</span><span class="o">%&gt;%</span><span class="n">pull</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">index</span><span class="p">)[</span><span class="err">$</span><span class="p">[</span><span class="n">name_col</span><span class="p">]]),</span><span class="n">Real_name</span> <span class="o">=</span> <span class="n">index</span><span class="o">%&gt;%</span><span class="n">pull</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">index</span><span class="p">)[</span><span class="err">$</span><span class="p">[</span><span class="n">real_name_col</span><span class="p">]]))</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">gene_exp</span><span class="p">[,</span><span class="err">$</span><span class="p">[</span><span class="n">start_at</span><span class="p">]:</span><span class="n">ncol</span><span class="p">(</span><span class="n">gene_exp</span><span class="p">)]</span>
    <span class="n">pos</span> <span class="o">=</span> <span class="n">match</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">df</span><span class="p">),</span> <span class="kp">name</span><span class="err">$</span><span class="n">Name</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">character</span><span class="p">())</span><span class="o">%&gt;%</span><span class="n">na</span><span class="o">.</span><span class="n">omit</span><span class="p">()</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span> <span class="o">%&gt;%</span>
      <span class="n">select_at</span><span class="p">(</span><span class="k">as</span><span class="o">.</span><span class="n">character</span><span class="p">(</span><span class="kp">name</span><span class="err">$</span><span class="n">Name</span><span class="p">[</span><span class="n">pos</span><span class="p">]),</span> 
                <span class="o">~</span><span class="k">as</span><span class="o">.</span><span class="n">character</span><span class="p">((</span><span class="kp">name</span><span class="err">$</span><span class="n">Real_name</span><span class="p">[</span><span class="n">pos</span><span class="p">])))</span>
    <span class="n">final</span> <span class="o">=</span> <span class="n">cbind</span><span class="p">(</span><span class="n">gene_exp</span><span class="p">[,</span><span class="mi">1</span><span class="p">:</span><span class="err">$</span><span class="p">[</span><span class="n">start_at</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span><span class="n">df</span><span class="p">)</span>
    <span class="n">final</span><span class="o">%&gt;%</span><span class="n">readr</span><span class="p">::</span><span class="n">write_tsv</span><span class="p">(</span> <span class="n">path</span> <span class="o">=</span> <span class="s2">&quot;$[_output[0]]&quot;</span><span class="p">,</span> <span class="n">na</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span><span class="p">,</span> <span class="n">append</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span> <span class="n">col_names</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">quote_escape</span> <span class="o">=</span> <span class="s2">&quot;double&quot;</span><span class="p">)</span>
    <span class="kp">name</span><span class="err">$</span><span class="n">Name</span><span class="p">[</span><span class="n">attr</span><span class="p">(</span><span class="n">pos</span><span class="p">,</span><span class="s2">&quot;na.action&quot;</span><span class="p">)]</span><span class="o">%&gt;%</span><span class="n">cat</span><span class="p">(</span><span class="nb">file</span> <span class="o">=</span> <span class="s2">&quot;$[_output[1]]&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot; &quot;</span><span class="p">)</span>

<span class="c1"># add position to gene_ID based on ensembl </span>
<span class="p">[</span><span class="n">Processing_2</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="mi">2</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{output}/{_input[0]:bn}_region_list.txt&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;4h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;6G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output[0]:bn}&#39;</span>
<span class="kn">R:</span>  <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.stdout&#39;</span><span class="p">,</span><span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;biomaRt&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
    <span class="n">gene_exp</span> <span class="o">=</span> <span class="n">readr</span><span class="p">::</span><span class="n">read_delim</span><span class="p">(</span><span class="s2">&quot;$[_input[0]]&quot;</span><span class="p">,</span><span class="n">delim</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">ensembl</span> <span class="o">=</span> <span class="n">useDataset</span><span class="p">(</span><span class="s2">&quot;hsapiens_gene_ensembl&quot;</span><span class="p">,</span><span class="n">mart</span><span class="o">=</span><span class="n">useMart</span><span class="p">(</span><span class="s2">&quot;ENSEMBL_MART_ENSEMBL&quot;</span><span class="p">))</span>
    <span class="n">ensembl_df</span> <span class="o">&lt;-</span> <span class="n">getBM</span><span class="p">(</span><span class="n">attributes</span><span class="o">=</span><span class="n">c</span><span class="p">(</span><span class="s2">&quot;ensembl_gene_id&quot;</span><span class="p">,</span><span class="s2">&quot;chromosome_name&quot;</span><span class="p">,</span> <span class="s2">&quot;start_position&quot;</span><span class="p">,</span> <span class="s2">&quot;end_position&quot;</span><span class="p">),</span><span class="n">mart</span><span class="o">=</span><span class="n">ensembl</span><span class="p">)</span>
    <span class="n">my_genes</span> <span class="o">=</span> <span class="n">gene_exp</span><span class="err">$</span><span class="n">gene_ID</span>
    <span class="n">my_genes_ann</span> <span class="o">=</span> <span class="n">ensembl_df</span><span class="p">[</span><span class="n">match</span><span class="p">(</span><span class="n">my_genes</span><span class="p">,</span> <span class="n">ensembl_df</span><span class="err">$</span><span class="n">ensembl_gene_id</span><span class="p">),]</span><span class="o">%&gt;%</span><span class="nb">filter</span><span class="p">(</span><span class="n">chromosome_name</span><span class="o">%</span><span class="ow">in</span><span class="o">%</span><span class="mi">1</span><span class="p">:</span><span class="mi">23</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">dplyr</span><span class="p">::</span><span class="n">select</span><span class="p">(</span> <span class="s2">&quot;#chr&quot;</span> <span class="o">=</span> <span class="n">chromosome_name</span><span class="p">,</span> <span class="n">start_position</span><span class="p">,</span> <span class="n">end_position</span><span class="p">,</span><span class="s2">&quot;gene_ID&quot;</span> <span class="o">=</span> <span class="n">ensembl_gene_id</span><span class="p">)</span><span class="o">%&gt;%</span><span class="nb">filter</span><span class="p">(</span><span class="n">gene_ID</span><span class="o">!=</span><span class="s2">&quot;NA&quot;</span><span class="p">)</span>
    <span class="n">my_genes_ann</span><span class="o">%&gt;%</span><span class="n">readr</span><span class="p">::</span><span class="n">write_tsv</span><span class="p">(</span> <span class="n">path</span> <span class="o">=</span> <span class="s2">&quot;$[_output]&quot;</span><span class="p">,</span> <span class="n">na</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span><span class="p">,</span> <span class="n">append</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span> <span class="n">col_names</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">quote_escape</span> <span class="o">=</span> <span class="s2">&quot;double&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
<hr>
&copy 2020-2021 Hao Sun & Gao Wang, Department of Neurology, Columbia University
<p><small>Exported from <a href="http://github.com/hsun3163/jnbinder/blob/93314c88684ab4704a5db1f03c434d85b038a8f8/workflow/SOS_ROSMAP_gene_exp_processing.ipynb"><code>workflow/SOS_ROSMAP_gene_exp_processing.ipynb</code></a> committed by Gao Wang on Thu May 6 00:26:26 2021 <a href="http://github.com/hsun3163/jnbinder/commit/93314c88684ab4704a5db1f03c434d85b038a8f8">revision 1, 93314c8</a> <a href="https://stephenslab.github.io/ipynb-website/notes.html#Note-about-commit-ids"><span class="fa fa-question-circle"></span></a></small></p>
</div>
</div>
</body>
</html>
