<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="ipynb_website:version" content="0.9.7" />
<meta name="viewport" content="width=device-width, initial-scale=1" />

<link rel="stylesheet" type="text/css" href="../css/jt.css">

<link rel="stylesheet" type="text/css" href="../css/toc2.css">

<link href="../site_libs/jqueryui-1.11.4/jquery-ui.css">
<link rel="stylesheet" href="../site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" href="../site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.9.1/jquery-ui.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>

<link rel="stylesheet"
      href="../site_libs/highlightjs/null.min.css"
      type="text/css" />

<script src="../site_libs/highlightjs/highlight.pack.js"></script>
<script>hljs.initHighlightingOnLoad();</script>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>

<script src="../js/doc_toc.js"></script>
<script src="../js/docs.js"></script>

<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-MML-AM_CHTML"></script>
<script>
    MathJax.Hub.Config({
        extensions: ["tex2jax.js"],
        jax: ["input/TeX", "output/HTML-CSS"],
        tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true
        },
        "HTML-CSS": {
            preferredFont: "TeX",
            availableFonts: ["TeX"],
            styles: {
                scale: 110,
                ".MathJax_Display": {
                    "font-size": "110%",
                }
            }
        }
    });
</script>
<script>
function filterDataFrame(id) {
    var input = document.getElementById("search_" + id);
    var filter = input.value.toUpperCase();
    var table = document.getElementById("dataframe_" + id);
    var tr = table.getElementsByTagName("tr");
    // Loop through all table rows, and hide those who don't match the search query
    for (var i = 1; i < tr.length; i++) {
        for (var j = 0; j < tr[i].cells.length; ++j) {
            var matched = false;
            if (tr[i].cells[j].innerHTML.toUpperCase().indexOf(filter) != -1) {
                tr[i].style.display = "";
                matched = true
                break;
            }
            if (!matched)
                tr[i].style.display = "none";
        }
    }
}
function sortDataFrame(id, n, dtype) {
    var table = document.getElementById("dataframe_" + id);
    var tb = table.tBodies[0]; // use `<tbody>` to ignore `<thead>` and `<tfoot>` rows
    var tr = Array.prototype.slice.call(tb.rows, 0); // put rows into array
    if (dtype === 'numeric') {
        var fn = function(a, b) { 
            return parseFloat(a.cells[n].textContent) <= parseFloat(b.cells[n].textContent) ? -1 : 1;
        }
    } else {
        var fn = function(a, b) {
            var c = a.cells[n].textContent.trim().localeCompare(b.cells[n].textContent.trim()); 
            return c > 0 ? 1 : (c < 0 ? -1 : 0) }
    }
    var isSorted = function(array, fn) {
        if (array.length < 2)
            return 1;
        var direction = fn(array[0], array[1]); 
        for (var i = 1; i < array.length - 1; ++i) {
            var d = fn(array[i], array[i+1]);
            if (d == 0)
                continue;
            else if (direction == 0)
                direction = d;
            else if (direction != d)
                return 0;
            }
        return direction;
    }
    var sorted = isSorted(tr, fn);
    if (sorted == 1 || sorted == -1) {
        // if sorted already, reverse it
        for(var i = tr.length - 1; i >= 0; --i)
            tb.appendChild(tr[i]); // append each row in order
    } else {
        tr = tr.sort(fn);
        for(var i = 0; i < tr.length; ++i)
            tb.appendChild(tr[i]); // append each row in order
    }
}
</script>

<script>
$( document ).ready(function(){
            var cfg={'threshold':3,     // depth of toc (number of levels)
             'number_sections': false,
             'toc_cell': false,          // useless here
             'toc_window_display': true, // display the toc window
             "toc_section_display": "block", // display toc contents in the window
             'sideBar':true,       // sidebar or floating window
             'navigate_menu':false       // navigation menu (only in liveNotebook -- do not change)
            }
            var st={};                  // some variables used in the script
            st.rendering_toc_cell = false;
            st.config_loaded = false;
            st.extension_initialized=false;
            st.nbcontainer_marginleft = $('#notebook-container').css('margin-left')
            st.nbcontainer_marginright = $('#notebook-container').css('margin-right')
            st.nbcontainer_width = $('#notebook-container').css('width')
            st.oldTocHeight = undefined
            st.cell_toc = undefined;
            st.toc_index=0;
            // fire the main function with these parameters
            table_of_contents(cfg, st);
            var file=workflowDict[$("h1:first").attr("id")];
            $("#toc-level0 a").css("color","#126dce");
            $('a[href="#'+$("h1:first").attr("id")+'"]').hide()
            var docs=workflowArray;
            var docs_map=workflowArrayMap;
            var pos=workflowArray.indexOf(file);
            for (var a=pos;a>=0;a--){
                  $('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>').insertBefore("#toc-level0 li:eq(0)");
            }
            $('a[href="'+file+'.html'+'"]').css("color","#126dce");
            for (var a=pos+1;a<docs.length;a++){
                  $(".toc #toc-level0").append('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>');
            }
            // $("#toc-header").hide(); // comment out because it prevents search bar from displaying
    });
</script>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');
  // mark it active
  menuAnchor.parent().addClass('active');
  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>
<div class="container-fluid main-container">
<!-- tabsets -->
<script src="../site_libs/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>



<title>Multi-tissue brain TWAS</title>

<style type = "text/css">
body {
  font-family: "Droid Sans";
  padding-top: 66px;
  padding-bottom: 40px;
}
</style>
</head>

<body>
<div tabindex="-1" id="notebook" class="border-box-sizing">
<div class="container" id="notebook-container">

<!-- code folding -->

<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">Multi-tissue brain TWAS</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
<li>
  <a href="../index.html">Overview</a>
</li>
        
<li>
  <a href="../workflow.html">Workflow</a>
</li>
        
      </ul>
        
<ul class="nav navbar-nav navbar-right">
<li>
   <a href="http://github.com/hsun3163/neuro-twas"> <span class="fa fa-github"></span> </a>
</li>
</ul>
        
      </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Susie-whole-genome-sequencing">Susie whole genome sequencing<a class="anchor-link" href="#Susie-whole-genome-sequencing">&#182;</a></h1><p>Ready to use pipeline to estimate univariate association between a molecular phenotype and the genotypes via SuSiE.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Pre-requisites">Pre-requisites<a class="anchor-link" href="#Pre-requisites">&#182;</a></h2><p>We provide a container image <code>docker://gaow/twas</code> that contains all software needed to run the pipeline. If you would like to configure it by yourself, please make sure you install the following software before running this notebook.</p>
<p>Bash:</p>
<ul>
<li>GCTA</li>
<li>PLINK</li>
</ul>
<p>R:</p>
<ul>
<li>Tidyverse</li>
<li>susieR</li>
<li>modelr</li>
<li>abind</li>
</ul>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Input-and-Output">Input and Output<a class="anchor-link" href="#Input-and-Output">&#182;</a></h1><h2 id="Input">Input<a class="anchor-link" href="#Input">&#182;</a></h2><ul>
<li><code>--genotype_list</code> An index text file with two columns of chromosome and the corresponding PLINK bed file.</li>
<li><code>--molecular-pheno</code>, The text file containing the table describing the molecular phenotype. It shall have regions(genes) as rows and samples as columnes</li>
<li><code>--region_list</code> The text file with 4 columns specifying the #Chr, P0 (Start position), P1(End position) and names of regions to analyze. The name of the column is not important but the order of the columns. It is also important that the column name of the first column starts with a #. The region_list can can be generated by using another sos pipeline SOS_ROSMAP_gene_exp_processing.ipynb.</li>
</ul>
<h2 id="Output">Output<a class="anchor-link" href="#Output">&#182;</a></h2><ul>
<li><code>uni_weight.RDS</code> a RDS file that served as the input for the mixture pipeline.</li>
<li><code>susie.RData</code> a R object containing all the susie output for each of the regions</li>
</ul>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Command-interface">Command interface<a class="anchor-link" href="#Command-interface">&#182;</a></h1>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[2]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="err">!</span><span class="n">sos</span> <span class="n">run</span> <span class="n">twas_fusion</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre>usage: sos run twas_fusion.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &#34;sos run -h&#34; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  twas_fusion
  association_test

Global Workflow Options:
  --molecular-pheno VAL (as path, required)
                        Path to the input molecular phenotype data.
  --gwas-sumstat VAL (as path, required)
                        Path to GWAS summary statistics data (association
                        results between SNP and disease in a GWAS)
  --genotype-list VAL (as path, required)
                        An index text file with two columns of chromosome and
                        the corresponding PLINK bed file.
  --region-list VAL (as path, required)
                        An index text file with 4 columns specifying the chr,
                        start, end and names of regions to analyze
  --wd . (as path)
                        Path to the work directory of the weight computation:
                        output weights and cache will be saved to this
                        directory.
  --weights-path  f&#39;{wd:a}/WEIGHTS&#39;

                        Specify the directory to save fitted weights
  --weights-list  f&#39;{weights_path}/{molecular_pheno:bn}.weights_list.txt&#39;

                        Path to list of weights
  --region-name VAL (as int, required)
                        Specify the column in the molecular_pheno file that
                        contains the name of the region of interest
  --data-start VAL (as int, required)
                        Specify the column in the molecular_pheno file where the
                        actual data start
  --window 50000 (as int)
                        Specify the scanning window for the up and downstream
                        radius to analyze around the region of interest, in
                        units of Kb
  --model blsmm blup lasso top1 enet (as list)
                        Specify what model are used to compute weights. Notice
                        that `blsmm` can be very resource consuming.
  --container &#39;gaow/twas&#39;
                        Container option for software to run the analysis:
                        docker or singularity

Sections
  twas_fusion_1:
  twas_fusion_2:
  twas_fusion_3:        Actual weight computation
  twas_fusion_4:
  twas_fusion_5, association_test: Association test
  twas_fusion_6:        Clean up dummy file
</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Working-example">Working example<a class="anchor-link" href="#Working-example">&#182;</a></h1><p>The MWE file is availble at :
"<a href="https://www.synapse.org/#!Synapse:syn24179064/files/">https://www.synapse.org/#!Synapse:syn24179064/files/</a>"</p>
<p>The time it take to run this MWE shall be around 2 minutes. Pay extra attention to the gene_start and gene_end position  when using following command on gene_exp file that are not this MWE. Also, when there is too few or too many genes that passed the heritability check, consider increasing or decreasing the --window options.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[1]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1">## Test pipeline with test data</span>
<span class="c1">## Switch back to abosolute path, otherwise there will be file not found error in step 5</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">susie</span><span class="o">-</span><span class="n">wgs</span><span class="o">-</span><span class="n">prior</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">susie</span> \
  <span class="o">--</span><span class="n">molecular</span><span class="o">-</span><span class="n">pheno</span> <span class="o">./</span><span class="n">molecular_phenotype</span> \
  <span class="o">--</span><span class="n">wd</span> <span class="o">./</span> \
  <span class="o">--</span><span class="n">genotype_list</span> <span class="o">./</span><span class="n">geno_dir</span>\
  <span class="o">--</span><span class="n">region_list</span> <span class="o">./</span><span class="n">region_list</span> \
  <span class="o">--</span><span class="n">region_name</span> <span class="mi">1</span> \
  <span class="o">--</span><span class="n">data_start</span> <span class="mi">5</span> \
  <span class="o">--</span><span class="n">window</span> <span class="mi">500000</span> \
  <span class="o">--</span><span class="n">container</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">containers</span><span class="o">/</span><span class="n">twas_latest</span><span class="o">.</span><span class="n">sif</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">


<div class="output_area">

    <div class="prompt"></div>


<div class="output_subarea output_stream output_stdout output_text">
<pre><span class="ansi-red-intense-fg">ERROR</span>: <span class="ansi-red-intense-fg">Failed to locate twas_fusion.ipynb.sos</span>

</pre>
</div>
</div>

</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="Global-parameter-settings">Global parameter settings<a class="anchor-link" href="#Global-parameter-settings">&#182;</a></h1><p>The section outlined the parameters that can be set in the command interface.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[5]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># Path to the input molecular phenotype data.</span>
<span class="kn">parameter:</span> <span class="n">molecular_pheno</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># An index text file with two columns of chromosome and the corresponding PLINK bed file.</span>
<span class="kn">parameter:</span> <span class="n">genotype_list</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># An index text file with 4 columns specifying the chr, start, end and names of regions to analyze</span>
<span class="kn">parameter:</span> <span class="n">region_list</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Path to the work directory of the weight computation: output weights and cache will be saved to this directory.</span>
<span class="kn">parameter:</span> <span class="n">wd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;./&#39;</span><span class="p">)</span>
<span class="c1"># Specify the directory to save fitted weights</span>
<span class="kn">parameter:</span> <span class="n">weights_path</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{wd:a}/WEIGHTS&#39;</span>
<span class="c1"># Path to list of weights</span>
<span class="kn">parameter:</span> <span class="n">weights_list</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{weights_path}/{molecular_pheno:bn}.weights_list.txt&#39;</span>
<span class="c1"># Path to store the output folder</span>
<span class="kn">parameter:</span> <span class="n">output_path</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{wd:a}/result&#39;</span>
<span class="c1"># Specify the column in the molecular_pheno file that contains the name of the region of interest</span>
<span class="kn">parameter:</span> <span class="n">region_name</span> <span class="o">=</span> <span class="nb">int</span>
<span class="c1"># Specify the column in the molecular_pheno file where the actual data start</span>
<span class="kn">parameter:</span> <span class="n">data_start</span> <span class="o">=</span> <span class="nb">int</span>
<span class="c1"># Specify the scanning window for the up and downstream radius to analyze around the region of interest, in units of Kb</span>
<span class="kn">parameter:</span> <span class="n">window</span> <span class="o">=</span> <span class="mi">50000</span>
<span class="c1"># Specify the number of jobs per run.</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">2</span>

<span class="c1"># Container option for software to run the analysis: docker or singularity</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;gaow/twas&#39;</span>

<span class="c1"># Propotion of samples set into testing, set to zero if no cv are needed.</span>
<span class="kn">parameter:</span> <span class="n">testing_prop</span> <span class="o">=</span> <span class="mf">0.2</span>
<span class="c1"># Number of training &amp; testing samples used</span>
<span class="kn">parameter:</span> <span class="n">cv_times</span> <span class="o">=</span> <span class="mi">2</span>

<span class="c1"># parameters for the susie pipelines.</span>
<span class="kn">parameter:</span> <span class="n">causal_variables_L</span> <span class="o">=</span> <span class="mi">10</span>
<span class="kn">parameter:</span> <span class="n">scaled_prior_variance</span> <span class="o">=</span> <span class="mf">0.1</span>

<span class="c1"># Get regions of interest to focus on.</span>
<span class="n">regions</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">region_list</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">)]</span>

<span class="n">geno_inventory</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">genotype_list</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">)])</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="k">def</span> <span class="nf">get_genotype_file</span><span class="p">(</span><span class="n">chrom</span><span class="p">,</span> <span class="n">genotype_list</span><span class="p">,</span> <span class="n">geno_inventory</span><span class="p">):</span>
    <span class="n">chrom</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{chrom}&#39;</span>
    <span class="k">if</span> <span class="n">chrom</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;chr&#39;</span><span class="p">):</span>
        <span class="n">chrom</span> <span class="o">=</span> <span class="n">chrom</span><span class="p">[</span><span class="mi">3</span><span class="p">:]</span>
    <span class="k">if</span> <span class="n">chrom</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">geno_inventory</span><span class="p">:</span>
        <span class="n">geno_file</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{chrom}&#39;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">geno_file</span> <span class="o">=</span> <span class="n">geno_inventory</span><span class="p">[</span><span class="n">chrom</span><span class="p">]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">geno_file</span><span class="p">):</span>
        <span class="c1"># relative path</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{genotype_list:ad}/&#39;</span> <span class="o">+</span> <span class="n">geno_file</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="n">f</span><span class="s2">&quot;Cannot find genotype file {geno_file}&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">geno_file</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{genotype_list:ad}/&#39;</span> <span class="o">+</span> <span class="n">geno_file</span>
    <span class="k">return</span> <span class="n">path</span><span class="p">(</span><span class="n">geno_file</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Partition-of-the-molecular-phenotype-for-each-genes">Partition of the molecular phenotype for each genes<a class="anchor-link" href="#Partition-of-the-molecular-phenotype-for-each-genes">&#182;</a></h2><p>This step extracts the molecular phenotype for each gene and transposes them into the formats needed in the follow-up analysis.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[11]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">hsq_1</span><span class="p">,</span><span class="n">susie_1</span><span class="p">,</span><span class="n">susie_cv_1</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">molecular_pheno</span><span class="p">,</span> <span class="n">for_each</span> <span class="o">=</span> <span class="s2">&quot;regions&quot;</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{wd:a}/cache/{_input:bn}.{_regions[3]}.exp&#39;</span><span class="p">,</span>
        <span class="n">f</span><span class="s1">&#39;{wd:a}/cache/{_input:bn}.{_regions[3]}.pheno&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;4h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;6G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output[0]:bn}&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]}.stdout&#39;</span><span class="p">,</span><span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="c1"># Get the line number for the region in the file</span>
    <span class="n">line_num</span> <span class="o">=</span> <span class="n">system</span><span class="p">(</span><span class="s2">&quot;awk &#39;($$[region_name]==</span><span class="se">\&quot;</span><span class="s2">$[_regions[3]]</span><span class="se">\&quot;</span><span class="s2">) {print NR}&#39; $[_input]&quot;</span><span class="p">,</span> <span class="nb">intern</span><span class="o">=</span><span class="n">T</span><span class="p">)</span>
    <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">line_num</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">){</span>
      <span class="n">stop</span><span class="p">(</span> <span class="s2">&quot;Cannot find $[_regions[3]] in column $[region_name]  $[_input]&quot;</span><span class="p">)}</span>
    <span class="n">yi</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">::</span><span class="n">fread</span><span class="p">(</span><span class="nb">file</span> <span class="o">=</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">],</span> <span class="n">skip</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">integer</span><span class="p">(</span><span class="n">line_num</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">nrows</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">samplenames_yi</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">table</span><span class="p">::</span><span class="n">fread</span><span class="p">(</span><span class="nb">file</span> <span class="o">=</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">:</span><span class="n">r</span><span class="p">],</span> <span class="n">skip</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">nrows</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">yi</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="n">colnames</span><span class="p">(</span><span class="n">samplenames_yi</span><span class="p">)</span>
    <span class="n">readr</span><span class="p">::</span><span class="n">write_tsv</span><span class="p">(</span><span class="n">yi</span><span class="p">,</span> <span class="n">path</span> <span class="o">=</span> <span class="s2">&quot;$[_output[0]]&quot;</span><span class="p">,</span> <span class="n">na</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span><span class="p">,</span> <span class="n">append</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span> <span class="n">col_names</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">quote_escape</span> <span class="o">=</span> <span class="s2">&quot;double&quot;</span><span class="p">)</span>
    <span class="n">yi</span> <span class="o">&lt;-</span> <span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">yi</span><span class="p">[,</span> <span class="err">$</span><span class="p">[</span><span class="n">data_start</span><span class="p">]:</span><span class="n">ncol</span><span class="p">(</span><span class="n">yi</span><span class="p">),</span> <span class="n">drop</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">])</span>
    <span class="n">yj</span> <span class="o">&lt;-</span> <span class="n">rbind</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">yi</span><span class="p">),</span><span class="n">colnames</span><span class="p">(</span><span class="n">yi</span><span class="p">),</span><span class="n">yi</span><span class="p">)</span>
    <span class="n">readr</span><span class="p">::</span><span class="n">write_tsv</span><span class="p">(</span><span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">t</span><span class="p">(</span><span class="n">yj</span><span class="p">)),</span> <span class="n">path</span> <span class="o">=</span> <span class="s2">&quot;$[_output[1]]&quot;</span><span class="p">,</span> <span class="n">na</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span><span class="p">,</span> <span class="n">append</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span> <span class="n">col_names</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">quote_escape</span> <span class="o">=</span> <span class="s2">&quot;double&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Construction-of-Plink-trio-for-each-gene">Construction of Plink trio for each gene<a class="anchor-link" href="#Construction-of-Plink-trio-for-each-gene">&#182;</a></h2><p>This step constructs the plink file for each gene based on the output of previous steps. Specifically it:</p>
<ol>
<li>Selects only the SNPs within the start and end position of the corresponding region (gene)</li>
<li>Replaces the Phenotype value (last column) of the .fam based on the input</li>
</ol>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[225]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">hsq_2</span><span class="p">,</span><span class="n">susie_2</span><span class="p">,</span><span class="n">susie_cv_2</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">group_with</span> <span class="o">=</span> <span class="s1">&#39;regions&#39;</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{_input[0]:n}.bed&#39;</span><span class="p">,</span>
        <span class="n">f</span><span class="s1">&#39;{_input[0]:n}.bim&#39;</span><span class="p">,</span>
        <span class="n">f</span><span class="s1">&#39;{_input[0]:n}.fam&#39;</span>

<span class="c1"># look up for genotype file</span>
<span class="n">geno_file</span> <span class="o">=</span> <span class="n">get_genotype_file</span><span class="p">(</span><span class="n">_regions</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span><span class="n">genotype_list</span><span class="p">,</span><span class="n">geno_inventory</span><span class="p">)</span>

<span class="kn">parameter:</span> <span class="n">extract_snp</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{geno_file:an}.bim&#39;</span>
<span class="kn">parameter:</span> <span class="n">exclude_snp</span> <span class="o">=</span> <span class="s2">&quot;./.&quot;</span>
<span class="kn">parameter:</span> <span class="n">keep_sample</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_input[1]}&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;12h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;6G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output[0]:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]}.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span> <span class="n">volumes</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span><span class="s1">&#39;{geno_file:ad}:{geno_file:ad}&#39;</span><span class="p">]</span>
    <span class="c1">##### Get the locus genotypes for $[_regions[3]]</span>
    <span class="n">plink</span> <span class="o">--</span><span class="n">bfile</span> <span class="err">$</span><span class="p">[</span><span class="n">geno_file</span><span class="p">:</span><span class="n">an</span><span class="p">]</span> \
    <span class="o">--</span><span class="n">pheno</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> \
    <span class="o">--</span><span class="n">make</span><span class="o">-</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">out</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">n</span><span class="p">]</span> \
    <span class="o">--</span><span class="nb">chr</span> <span class="err">$</span><span class="p">[</span><span class="n">_regions</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> \
    <span class="o">--</span><span class="n">from</span><span class="o">-</span><span class="n">bp</span> <span class="err">$</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">_regions</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">-</span> <span class="n">window</span> <span class="p">]</span> \
    <span class="o">--</span><span class="n">to</span><span class="o">-</span><span class="n">bp</span> <span class="err">$</span><span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">_regions</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="n">window</span> <span class="p">]</span> \
    <span class="o">--</span><span class="n">keep</span> <span class="err">$</span><span class="p">[</span><span class="n">keep_sample</span><span class="p">]</span> \
    <span class="o">--</span><span class="n">extract</span> <span class="err">$</span><span class="p">[</span><span class="n">extract_snp</span><span class="p">]</span>
    <span class="o">--</span><span class="n">exclude</span> <span class="err">$</span><span class="p">[</span><span class="n">exclude_snp</span><span class="p">]</span>
    <span class="o">--</span><span class="n">allow</span><span class="o">-</span><span class="n">no</span><span class="o">-</span><span class="n">sex</span> <span class="o">||</span> <span class="n">true</span>
    <span class="n">touch</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Heritability-Estimation">Heritability Estimation<a class="anchor-link" href="#Heritability-Estimation">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">hsq_3</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="n">group_with</span> <span class="o">=</span> <span class="s1">&#39;regions&#39;</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{_input[0]:n}.bed&#39;</span><span class="p">,</span>
        <span class="n">f</span><span class="s1">&#39;{_input[0]:n}.bim&#39;</span><span class="p">,</span>
        <span class="n">f</span><span class="s1">&#39;{_input[0]:n}.fam&#39;</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="n">skip_if</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">getsize</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{_input[0]}&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;4h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;6G&quot;</span> <span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output[0]:bn}&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span><span class="n">volumes</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span><span class="s1">&#39;{wd:a}:{wd:a}&#39;</span><span class="p">]</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;dplyr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;tibble&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;susieR&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;plink2R&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;readr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;modelr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;purrr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;abind&quot;</span><span class="p">)</span>



    <span class="c1"># Perform i/o checks here:</span>
    <span class="n">files</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span><span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]],</span><span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">]])</span>

    <span class="k">for</span> <span class="p">(</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">files</span> <span class="p">)</span> <span class="p">{</span>
        <span class="k">if</span> <span class="p">(</span> <span class="err">!</span><span class="nb">file</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">f</span><span class="p">)</span> <span class="p">){</span>
            <span class="n">cat</span><span class="p">(</span> <span class="s2">&quot;ERROR: &quot;</span><span class="p">,</span> <span class="n">f</span> <span class="p">,</span> <span class="s2">&quot; input file does not exist</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">stderr</span><span class="p">()</span> <span class="p">)</span>
            <span class="n">cleanup</span><span class="p">()</span>
            <span class="n">q</span><span class="p">()</span>
        <span class="p">}</span>
    <span class="p">}</span>

    <span class="k">if</span> <span class="p">(</span> <span class="n">system</span><span class="p">(</span> <span class="n">paste</span><span class="p">(</span><span class="s2">&quot;plink&quot;</span><span class="p">,</span><span class="s2">&quot;--help&quot;</span><span class="p">)</span> <span class="p">,</span> <span class="n">ignore</span><span class="o">.</span><span class="n">stdout</span><span class="o">=</span><span class="n">T</span><span class="p">,</span><span class="n">ignore</span><span class="o">.</span><span class="n">stderr</span><span class="o">=</span><span class="n">T</span> <span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span> <span class="p">)</span> <span class="p">{</span>
        <span class="n">cat</span><span class="p">(</span> <span class="s2">&quot;ERROR: plink could not be executed </span><span class="se">\n</span><span class="s2">&quot;</span> <span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">stderr</span><span class="p">()</span> <span class="p">)</span>
        <span class="n">cleanup</span><span class="p">()</span>
        <span class="n">q</span><span class="p">()</span>
    <span class="p">}</span>

    <span class="k">if</span> <span class="p">(</span> <span class="n">system</span><span class="p">(</span> <span class="s2">&quot;gcta64&quot;</span> <span class="p">,</span> <span class="n">ignore</span><span class="o">.</span><span class="n">stdout</span><span class="o">=</span><span class="n">T</span><span class="p">,</span><span class="n">ignore</span><span class="o">.</span><span class="n">stderr</span><span class="o">=</span><span class="n">T</span> <span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span> <span class="p">){</span>
        <span class="n">cat</span><span class="p">(</span> <span class="s2">&quot;ERROR: gcta (gcta64) could not be executed &quot;</span> <span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">stderr</span><span class="p">()</span> <span class="p">)</span>
        <span class="n">cleanup</span><span class="p">()</span>
        <span class="n">q</span><span class="p">()</span>
    <span class="p">}</span>

    <span class="c1"># --- </span>
    <span class="c1"># Set up the &quot;input&quot;</span>
    <span class="n">opt</span> <span class="o">=</span> <span class="nb">list</span><span class="p">()</span>
    <span class="n">opt</span><span class="err">$</span><span class="n">bfile</span> <span class="o">=</span> <span class="s2">&quot;$[_input[0]:n]&quot;</span>
    <span class="n">opt</span><span class="err">$</span><span class="n">tmp</span> <span class="o">=</span> <span class="s2">&quot;$[_input[0]:n].tmp&quot;</span>
    <span class="n">opt</span><span class="err">$</span><span class="n">PATH_plink</span> <span class="o">=</span> <span class="s2">&quot;plink&quot;</span>
    <span class="n">opt</span><span class="err">$</span><span class="n">PATH_gcta</span> <span class="o">=</span> <span class="s2">&quot;gcta64&quot;</span>
    
    
    <span class="c1"># ---</span>

    <span class="n">fam</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">paste</span><span class="p">(</span><span class="n">opt</span><span class="err">$</span><span class="n">bfile</span><span class="p">,</span><span class="s2">&quot;.fam&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">),</span><span class="k">as</span><span class="o">.</span><span class="ow">is</span><span class="o">=</span><span class="n">T</span><span class="p">)</span>

    <span class="c1"># Make/fetch the phenotype file</span>
    
        <span class="n">pheno</span><span class="o">.</span><span class="n">file</span> <span class="o">=</span> <span class="n">paste</span><span class="p">(</span><span class="n">opt</span><span class="err">$</span><span class="n">tmp</span><span class="p">,</span><span class="s2">&quot;.pheno&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
        <span class="n">pheno</span> <span class="o">=</span> <span class="n">fam</span><span class="p">[,</span><span class="n">c</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">6</span><span class="p">)]</span>
        <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">pheno</span><span class="p">,</span><span class="n">quote</span><span class="o">=</span><span class="n">F</span><span class="p">,</span><span class="n">row</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">F</span><span class="p">,</span><span class="n">col</span><span class="o">.</span><span class="n">names</span><span class="o">=</span><span class="n">F</span><span class="p">,</span><span class="nb">file</span><span class="o">=</span><span class="n">pheno</span><span class="o">.</span><span class="n">file</span><span class="p">)</span>
   
    <span class="n">geno</span><span class="o">.</span><span class="n">file</span> <span class="o">=</span> <span class="n">opt</span><span class="err">$</span><span class="n">tmp</span>
    <span class="c1"># recode to the intersection of samples and new phenotype</span>
    <span class="n">arg</span> <span class="o">=</span> <span class="n">paste</span><span class="p">(</span> <span class="n">opt</span><span class="err">$</span><span class="n">PATH_plink</span> <span class="p">,</span><span class="s2">&quot; --allow-no-sex --bfile &quot;</span><span class="p">,</span><span class="n">opt</span><span class="err">$</span><span class="n">bfile</span><span class="p">,</span><span class="s2">&quot; --pheno &quot;</span><span class="p">,</span><span class="n">pheno</span><span class="o">.</span><span class="n">file</span><span class="p">,</span><span class="s2">&quot; --keep &quot;</span><span class="p">,</span><span class="n">pheno</span><span class="o">.</span><span class="n">file</span><span class="p">,</span><span class="s2">&quot; --make-bed --out &quot;</span><span class="p">,</span><span class="n">geno</span><span class="o">.</span><span class="n">file</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">system</span><span class="p">(</span><span class="n">arg</span> <span class="p">,</span> <span class="n">ignore</span><span class="o">.</span><span class="n">stdout</span><span class="o">=</span><span class="n">SYS_PRINT</span><span class="p">,</span><span class="n">ignore</span><span class="o">.</span><span class="n">stderr</span><span class="o">=</span><span class="n">SYS_PRINT</span><span class="p">)</span>

    <span class="c1"># --- HERITABILITY ANALYSIS</span>
    
    <span class="c1"># 1. generate GRM</span>
    <span class="n">arg</span> <span class="o">=</span> <span class="n">paste</span><span class="p">(</span> <span class="n">opt</span><span class="err">$</span><span class="n">PATH_plink</span><span class="p">,</span><span class="s2">&quot; --allow-no-sex --bfile &quot;</span><span class="p">,</span><span class="n">opt</span><span class="err">$</span><span class="n">tmp</span><span class="p">,</span><span class="s2">&quot; --make-grm-bin --out &quot;</span><span class="p">,</span><span class="n">opt</span><span class="err">$</span><span class="n">tmp</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="n">system</span><span class="p">(</span><span class="n">arg</span> <span class="p">,</span> <span class="n">ignore</span><span class="o">.</span><span class="n">stdout</span><span class="o">=</span><span class="n">SYS_PRINT</span><span class="p">,</span><span class="n">ignore</span><span class="o">.</span><span class="n">stderr</span><span class="o">=</span><span class="n">SYS_PRINT</span><span class="p">)</span>

    <span class="c1"># 2. estimate heritability</span>
    <span class="k">if</span> <span class="p">(</span> <span class="err">!</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">opt</span><span class="err">$</span><span class="n">covar</span><span class="p">)</span> <span class="p">)</span> <span class="p">{</span>
    <span class="n">arg</span> <span class="o">=</span> <span class="n">paste</span><span class="p">(</span> <span class="n">opt</span><span class="err">$</span><span class="n">PATH_gcta</span> <span class="p">,</span><span class="s2">&quot; --grm &quot;</span><span class="p">,</span><span class="n">opt</span><span class="err">$</span><span class="n">tmp</span><span class="p">,</span><span class="s2">&quot; --pheno &quot;</span><span class="p">,</span><span class="n">raw</span><span class="o">.</span><span class="n">pheno</span><span class="o">.</span><span class="n">file</span><span class="p">,</span><span class="s2">&quot; --qcovar &quot;</span><span class="p">,</span><span class="n">opt</span><span class="err">$</span><span class="n">covar</span><span class="p">,</span><span class="s2">&quot; --out &quot;</span><span class="p">,</span><span class="n">opt</span><span class="err">$</span><span class="n">tmp</span><span class="p">,</span><span class="s2">&quot; --reml --reml-no-constrain --reml-lrt 1&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
    <span class="n">arg</span> <span class="o">=</span> <span class="n">paste</span><span class="p">(</span> <span class="n">opt</span><span class="err">$</span><span class="n">PATH_gcta</span> <span class="p">,</span><span class="s2">&quot; --grm &quot;</span><span class="p">,</span><span class="n">opt</span><span class="err">$</span><span class="n">tmp</span><span class="p">,</span><span class="s2">&quot; --pheno &quot;</span><span class="p">,</span><span class="n">pheno</span><span class="o">.</span><span class="n">file</span><span class="p">,</span><span class="s2">&quot; --out &quot;</span><span class="p">,</span><span class="n">opt</span><span class="err">$</span><span class="n">tmp</span><span class="p">,</span><span class="s2">&quot; --reml --reml-no-constrain --reml-lrt 1&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="n">system</span><span class="p">(</span><span class="n">arg</span> <span class="p">,</span> <span class="n">ignore</span><span class="o">.</span><span class="n">stdout</span><span class="o">=</span><span class="n">SYS_PRINT</span><span class="p">,</span><span class="n">ignore</span><span class="o">.</span><span class="n">stderr</span><span class="o">=</span><span class="n">SYS_PRINT</span><span class="p">)</span>

    <span class="c1"># 3. evaluate LRT and V(G)/Vp</span>
    <span class="k">if</span> <span class="p">(</span> <span class="err">!</span><span class="nb">file</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span> <span class="n">paste</span><span class="p">(</span><span class="n">opt</span><span class="err">$</span><span class="n">tmp</span><span class="p">,</span><span class="s2">&quot;.hsq&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span> <span class="p">)</span> <span class="p">)</span> <span class="p">{</span>
        <span class="n">cat</span><span class="p">(</span><span class="n">opt</span><span class="err">$</span><span class="n">tmp</span><span class="p">,</span><span class="s2">&quot;does not exist, likely GCTA could not converge, skipping gene</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span><span class="nb">file</span><span class="o">=</span><span class="n">stderr</span><span class="p">())</span>
        <span class="n">cleanup</span><span class="p">()</span>
        <span class="n">q</span><span class="p">()</span>
    <span class="p">}</span>

    <span class="n">hsq</span><span class="o">.</span><span class="n">file</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="nb">file</span><span class="o">=</span><span class="n">paste</span><span class="p">(</span><span class="n">opt</span><span class="err">$</span><span class="n">tmp</span><span class="p">,</span><span class="s2">&quot;.hsq&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">),</span><span class="k">as</span><span class="o">.</span><span class="ow">is</span><span class="o">=</span><span class="n">T</span><span class="p">,</span><span class="n">fill</span><span class="o">=</span><span class="n">T</span><span class="p">)</span>
    <span class="n">hsq</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">unlist</span><span class="p">(</span><span class="n">hsq</span><span class="o">.</span><span class="n">file</span><span class="p">[</span><span class="n">hsq</span><span class="o">.</span><span class="n">file</span><span class="p">[,</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;V(G)/Vp&quot;</span><span class="p">,</span><span class="mi">2</span><span class="p">:</span><span class="mi">3</span><span class="p">]))</span>
    <span class="n">hsq</span><span class="o">.</span><span class="n">pv</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">numeric</span><span class="p">(</span><span class="n">unlist</span><span class="p">(</span><span class="n">hsq</span><span class="o">.</span><span class="n">file</span><span class="p">[</span><span class="n">hsq</span><span class="o">.</span><span class="n">file</span><span class="p">[,</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Pval&quot;</span><span class="p">,</span><span class="mi">2</span><span class="p">]))</span>

    <span class="k">if</span> <span class="p">(</span> <span class="n">opt</span><span class="err">$</span><span class="n">verbose</span> <span class="o">&gt;=</span> <span class="mi">1</span> <span class="p">)</span> <span class="n">cat</span><span class="p">(</span><span class="s2">&quot;Heritability (se):&quot;</span><span class="p">,</span><span class="n">hsq</span><span class="p">,</span><span class="s2">&quot;LRT P-value:&quot;</span><span class="p">,</span><span class="n">hsq</span><span class="o">.</span><span class="n">pv</span><span class="p">,</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="p">(</span> <span class="n">opt</span><span class="err">$</span><span class="n">save_hsq</span> <span class="p">)</span> <span class="n">cat</span><span class="p">(</span> <span class="n">opt</span><span class="err">$</span><span class="n">out</span> <span class="p">,</span> <span class="n">hsq</span> <span class="p">,</span> <span class="n">hsq</span><span class="o">.</span><span class="n">pv</span> <span class="p">,</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span> <span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">paste</span><span class="p">(</span><span class="n">opt</span><span class="err">$</span><span class="n">out</span><span class="p">,</span><span class="s2">&quot;.hsq&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span> <span class="p">)</span>

    <span class="c1"># 4. stop if insufficient</span>
    <span class="k">if</span> <span class="p">(</span> <span class="n">hsq</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="o">||</span> <span class="n">hsq</span><span class="o">.</span><span class="n">pv</span> <span class="o">&gt;</span> <span class="n">opt</span><span class="err">$</span><span class="n">hsq_p</span> <span class="p">)</span> <span class="p">{</span>
        <span class="n">cat</span><span class="p">(</span><span class="n">opt</span><span class="err">$</span><span class="n">tmp</span><span class="p">,</span><span class="s2">&quot; : heritability &quot;</span><span class="p">,</span><span class="n">hsq</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="s2">&quot;; LRT P-value &quot;</span><span class="p">,</span><span class="n">hsq</span><span class="o">.</span><span class="n">pv</span><span class="p">,</span><span class="s2">&quot; : skipping gene</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">,</span><span class="nb">file</span><span class="o">=</span><span class="n">stderr</span><span class="p">())</span>
        <span class="n">cleanup</span><span class="p">()</span>
        <span class="n">q</span><span class="p">()</span>
    <span class="p">}</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Conducting-univariate-test-for-all-the-genes-and-save-the-sumstat">Conducting univariate test for all the genes and save the sumstat<a class="anchor-link" href="#Conducting-univariate-test-for-all-the-genes-and-save-the-sumstat">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1">#Susie test</span>
<span class="p">[</span><span class="n">susie_3</span><span class="p">,</span><span class="n">susie_cv_3</span><span class="p">]</span>
<span class="kn">input:</span>  <span class="n">group_by</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="n">group_with</span> <span class="o">=</span> <span class="s1">&#39;regions&#39;</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{wd:a}/susie/{_input[0]:bn}.susie.model.RData&#39;</span><span class="p">,</span>
        <span class="n">f</span><span class="s1">&#39;{wd:a}/susie/{_input[0]:bn}.uni_weight.rds&#39;</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="n">skip_if</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">getsize</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{_input[0]}&#39;</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;4h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;6G&quot;</span> <span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output[0]:bn}&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]:n}.stdout&#39;</span><span class="p">,</span> <span class="n">container</span> <span class="o">=</span> <span class="n">container</span><span class="p">,</span><span class="n">volumes</span> <span class="o">=</span> <span class="p">[</span><span class="n">f</span><span class="s1">&#39;{wd:a}:{wd:a}&#39;</span><span class="p">]</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;dplyr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;tibble&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;susieR&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;plink2R&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;readr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;modelr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;purrr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;abind&quot;</span><span class="p">)</span>

    <span class="c1"># Define functions</span>
    <span class="c1">###Functions to compute MAF and missing genotype rate</span>
    <span class="n">compute_maf</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">geno</span><span class="p">){</span>
      <span class="n">f</span> <span class="o">&lt;-</span> <span class="n">mean</span><span class="p">(</span><span class="n">geno</span><span class="p">,</span><span class="n">na</span><span class="o">.</span><span class="n">rm</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span>
      <span class="k">return</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="mi">1</span><span class="o">-</span><span class="n">f</span><span class="p">))</span>
    <span class="p">}</span>
    
    <span class="n">compute_missing</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">geno</span><span class="p">){</span>
      <span class="n">miss</span> <span class="o">&lt;-</span> <span class="nb">sum</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">geno</span><span class="p">))</span><span class="o">/</span><span class="n">length</span><span class="p">(</span><span class="n">geno</span><span class="p">)</span>
      <span class="k">return</span><span class="p">(</span><span class="n">miss</span><span class="p">)</span>
    <span class="p">}</span>
    
    <span class="n">mean_impute</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">geno</span><span class="p">){</span>
      <span class="n">f</span> <span class="o">&lt;-</span> <span class="nb">apply</span><span class="p">(</span><span class="n">geno</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">mean</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="n">na</span><span class="o">.</span><span class="n">rm</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">))</span>
      <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">f</span><span class="p">))</span> <span class="n">geno</span><span class="p">[,</span><span class="n">i</span><span class="p">][</span><span class="n">which</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">geno</span><span class="p">[,</span><span class="n">i</span><span class="p">]))]</span> <span class="o">&lt;-</span> <span class="n">f</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
      <span class="k">return</span><span class="p">(</span><span class="n">geno</span><span class="p">)</span>
    <span class="p">}</span>
    
    <span class="n">is_zero_variance</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="p">{</span>
      <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">x</span><span class="o">%&gt;%</span><span class="n">na</span><span class="o">.</span><span class="n">omit</span><span class="p">()))</span><span class="o">==</span><span class="mi">1</span><span class="p">)</span> <span class="k">return</span><span class="p">(</span><span class="n">T</span><span class="p">)</span>
      <span class="k">else</span> <span class="k">return</span><span class="p">(</span><span class="n">F</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="c1">### Filter X matrix</span>
    <span class="n">filter_X</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">missing_rate_thresh</span><span class="p">,</span> <span class="n">maf_thresh</span><span class="p">)</span> <span class="p">{</span>
      <span class="n">rm_col</span> <span class="o">&lt;-</span> <span class="n">which</span><span class="p">(</span><span class="nb">apply</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">compute_missing</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">missing_rate_thresh</span><span class="p">)</span>
      <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">rm_col</span><span class="p">))</span> <span class="n">X</span> <span class="o">&lt;-</span> <span class="n">X</span><span class="p">[,</span> <span class="o">-</span><span class="n">rm_col</span><span class="p">]</span>
      <span class="n">rm_col</span> <span class="o">&lt;-</span> <span class="n">which</span><span class="p">(</span><span class="nb">apply</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">compute_maf</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">maf_thresh</span><span class="p">)</span>
      <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">rm_col</span><span class="p">))</span> <span class="n">X</span> <span class="o">&lt;-</span> <span class="n">X</span><span class="p">[,</span> <span class="o">-</span><span class="n">rm_col</span><span class="p">]</span>
      <span class="n">rm_col</span> <span class="o">&lt;-</span> <span class="n">which</span><span class="p">(</span><span class="nb">apply</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">is_zero_variance</span><span class="p">))</span>
      <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">rm_col</span><span class="p">))</span> <span class="n">X</span> <span class="o">&lt;-</span> <span class="n">X</span><span class="p">[,</span> <span class="o">-</span><span class="n">rm_col</span><span class="p">]</span>
      <span class="k">return</span><span class="p">(</span><span class="n">mean_impute</span><span class="p">(</span><span class="n">X</span><span class="p">))}</span>
      
    <span class="c1">###Function to impute the missing X with means and then scale and center X</span>
      <span class="n">impute_and_transform</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">genos</span><span class="p">,</span><span class="n">impute</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">){</span>
      <span class="n">tmp</span> <span class="o">=</span> <span class="n">genos</span>
      <span class="k">if</span><span class="p">(</span><span class="n">impute</span> <span class="o">==</span> <span class="n">TRUE</span><span class="p">){</span>
      <span class="k">for</span><span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">tmp</span><span class="p">)){</span>
        <span class="n">tmp</span><span class="p">[,</span><span class="n">i</span><span class="p">]</span><span class="o">=</span><span class="n">coalesce</span><span class="p">(</span><span class="n">tmp</span><span class="p">[,</span><span class="n">i</span><span class="p">],</span><span class="n">mean</span><span class="p">(</span><span class="n">tmp</span><span class="p">[,</span><span class="n">i</span><span class="p">]</span><span class="o">%&gt;%</span><span class="n">na</span><span class="o">.</span><span class="n">omit</span><span class="p">()))</span><span class="o">%&gt;%</span><span class="n">scale</span><span class="p">()}</span>
        <span class="k">return</span><span class="p">(</span><span class="n">tmp</span><span class="p">)</span>
      <span class="p">}</span> <span class="k">else</span> <span class="p">{</span>
    <span class="k">for</span><span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">tmp</span><span class="p">)){</span>
        <span class="n">tmp</span><span class="p">[,</span><span class="n">i</span><span class="p">]</span><span class="o">=</span><span class="n">tmp</span><span class="p">[,</span><span class="n">i</span><span class="p">]</span><span class="o">%&gt;%</span><span class="n">scale</span><span class="p">()}</span>
        <span class="k">return</span><span class="p">(</span><span class="n">tmp</span><span class="p">)}}</span>
        
    <span class="c1">###Function to impute the weight and standard errors</span>
    
    <span class="n">mm_regression</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">Y</span><span class="p">,</span> <span class="n">Z</span><span class="o">=</span><span class="n">NULL</span><span class="p">,</span><span class="n">center</span><span class="o">=</span><span class="n">TRUE</span><span class="p">,</span><span class="n">scale</span><span class="o">=</span><span class="n">TRUE</span><span class="p">)</span> <span class="p">{</span>
    <span class="c1">## HS: Make sure X and Y is matrixs as well otherwise the ncol(Y) give error</span>
    <span class="n">X</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
    <span class="n">Y</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">Y</span><span class="p">)</span>
    <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">Z</span><span class="p">))</span> <span class="p">{</span>
    <span class="n">Z</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">Z</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="k">if</span><span class="p">(</span><span class="nb">any</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">Y</span><span class="p">))){</span>
    <span class="n">reg</span> <span class="o">=</span> <span class="n">lapply</span><span class="p">(</span><span class="n">seq_len</span><span class="p">(</span><span class="n">ncol</span><span class="p">(</span><span class="n">Y</span><span class="p">)),</span> <span class="n">function</span> <span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="n">simplify2array</span><span class="p">(</span><span class="n">susieR</span><span class="p">:::</span><span class="n">univariate_regression</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">Y</span><span class="p">[,</span><span class="n">i</span><span class="p">],</span> <span class="n">Z</span><span class="p">,</span> <span class="n">center</span><span class="p">,</span> <span class="n">scale</span><span class="p">)))</span>
    <span class="n">reg</span> <span class="o">=</span> <span class="n">do</span><span class="o">.</span><span class="n">call</span><span class="p">(</span><span class="n">abind</span><span class="p">,</span> <span class="n">c</span><span class="p">(</span><span class="n">reg</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="mi">0</span><span class="p">)))</span>
    <span class="c1"># return array: out[1,,] is betahat, out[2,,] is shat</span>
    <span class="n">out</span> <span class="o">=</span> <span class="n">aperm</span><span class="p">(</span><span class="n">reg</span><span class="p">,</span> <span class="n">c</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
    <span class="c1">#HS Force dimension for the matrix slice.</span>
    <span class="n">out</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">bhat</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">out</span><span class="p">[</span><span class="mi">1</span><span class="p">,,]),</span> <span class="n">sbhat</span><span class="o">=</span><span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">out</span><span class="p">[</span><span class="mi">2</span><span class="p">,,]))</span>
    <span class="p">}</span><span class="k">else</span><span class="p">{</span>
    <span class="n">out</span> <span class="o">=</span> <span class="n">univariate_regression</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">Y</span><span class="p">,</span> <span class="n">Z</span><span class="p">,</span> <span class="n">center</span> <span class="o">=</span> <span class="n">F</span><span class="p">,</span> <span class="n">scale</span> <span class="o">=</span> <span class="n">F</span><span class="p">)</span>
    <span class="c1">#HS Original out dont has betahat instead of bhat, fixed now.</span>
    <span class="n">out</span><span class="err">$</span><span class="n">bhat</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">out</span><span class="err">$</span><span class="n">betahat</span><span class="p">)</span>
    <span class="n">out</span><span class="err">$</span><span class="n">sbhat</span> <span class="o">=</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">out</span><span class="err">$</span><span class="n">sebetahat</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">X</span><span class="p">)))</span> <span class="p">{</span>
    <span class="n">rownames</span><span class="p">(</span><span class="n">out</span><span class="err">$</span><span class="n">bhat</span><span class="p">)</span> <span class="o">=</span> <span class="n">colnames</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
    <span class="n">rownames</span><span class="p">(</span><span class="n">out</span><span class="err">$</span><span class="n">sbhat</span><span class="p">)</span> <span class="o">=</span> <span class="n">colnames</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="k">if</span> <span class="p">(</span><span class="err">!</span><span class="ow">is</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="n">Y</span><span class="p">)))</span> <span class="p">{</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">out</span><span class="err">$</span><span class="n">bhat</span><span class="p">)</span> <span class="o">=</span> <span class="n">colnames</span><span class="p">(</span><span class="n">Y</span><span class="p">)</span>
    <span class="n">colnames</span><span class="p">(</span><span class="n">out</span><span class="err">$</span><span class="n">sbhat</span><span class="p">)</span> <span class="o">=</span> <span class="n">colnames</span><span class="p">(</span><span class="n">Y</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="c1"># `out` is a list of bhat and sbhat</span>
    <span class="k">return</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>
    <span class="p">}</span>

    <span class="c1"># Load data and transform</span>
    <span class="n">genos</span> <span class="o">=</span> <span class="n">read_plink</span><span class="p">(</span><span class="s2">&quot;$[_input[0]:n]&quot;</span><span class="p">)</span>
  
    <span class="c1"># Filter X by 0.1 NA and 0.01 MAF, and then transfrom X</span>
    
    <span class="n">X_ftr</span> <span class="o">=</span> <span class="n">filter_X</span><span class="p">(</span><span class="n">genos</span><span class="err">$</span><span class="n">bed</span><span class="p">,</span><span class="mf">0.1</span><span class="p">,</span><span class="mf">0.01</span><span class="p">)</span>
    <span class="n">X</span> <span class="o">=</span> <span class="n">impute_and_transform</span><span class="p">(</span><span class="n">X_ftr</span><span class="p">,</span><span class="n">impute</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)</span>
    
    <span class="n">Y</span> <span class="o">=</span> <span class="n">genos</span><span class="err">$</span><span class="n">fam</span><span class="o">%&gt;%</span><span class="n">as_tibble</span><span class="p">()</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span><span class="kp">name</span> <span class="o">=</span> <span class="n">paste</span><span class="p">(</span><span class="n">V1</span><span class="p">,</span><span class="s2">&quot;:&quot;</span><span class="p">,</span><span class="n">V2</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">))</span>
    
    <span class="c1"># Make sure X and Y have the same order</span>
    <span class="n">Y</span> <span class="o">=</span> <span class="n">Y</span><span class="o">%&gt;%</span><span class="n">arrange</span><span class="p">(</span><span class="n">match</span><span class="p">(</span><span class="kp">name</span><span class="p">,</span><span class="n">rownames</span><span class="p">(</span><span class="n">X</span><span class="p">)))</span><span class="o">%&gt;%</span><span class="n">select</span><span class="p">(</span><span class="n">V6</span><span class="p">)</span>
    
    
    <span class="c1"># Center and scale Y </span>
    
    <span class="n">Y</span> <span class="o">=</span> <span class="n">impute_and_transform</span><span class="p">(</span><span class="n">Y</span><span class="p">,</span> <span class="n">impute</span> <span class="o">=</span> <span class="n">FALSE</span> <span class="p">)</span>
    <span class="n">Y</span> <span class="o">=</span> <span class="n">Y</span><span class="err">$</span><span class="n">V6</span>
    
    <span class="c1"># Susie with full samples</span>
    <span class="n">full_model</span> <span class="o">=</span> <span class="n">susie</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">Y</span><span class="p">,</span>
                  <span class="n">L</span> <span class="o">=</span> <span class="err">$</span><span class="p">[</span><span class="n">causal_variables_L</span><span class="p">],</span>
                  <span class="n">estimate_residual_variance</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> 
                  <span class="n">estimate_prior_variance</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span>
                  <span class="n">scaled_prior_variance</span> <span class="o">=</span> <span class="err">$</span><span class="p">[</span><span class="n">scaled_prior_variance</span><span class="p">])</span>
                  

    <span class="c1"># Get sumstat from the univariate regression</span>
    
    <span class="n">uni</span> <span class="o">=</span> <span class="n">mm_regression</span><span class="p">(</span><span class="n">X</span><span class="p">,</span><span class="n">Y</span><span class="p">,</span><span class="n">center</span><span class="o">=</span><span class="n">FALSE</span><span class="p">,</span><span class="n">scale</span><span class="o">=</span><span class="n">FALSE</span><span class="p">)</span>
    
    <span class="c1"># Save the sumstat objects</span>
    
    <span class="n">uni</span><span class="o">%&gt;%</span><span class="n">saveRDS</span><span class="p">(</span><span class="s2">&quot;$[_output[1]]&quot;</span><span class="p">)</span>
    
    <span class="c1"># Save the model for future use</span>
    <span class="n">full_model</span><span class="err">$</span><span class="n">X</span> <span class="o">=</span> <span class="n">X</span>
    <span class="n">full_model</span><span class="err">$</span><span class="n">Y</span> <span class="o">=</span> <span class="n">Y</span>
    <span class="n">save</span><span class="p">(</span><span class="n">full_model</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="s2">&quot;$[_output[0]]&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Crossvalidation">Crossvalidation<a class="anchor-link" href="#Crossvalidation">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1"># CV with univariate susie</span>
<span class="p">[</span><span class="n">susie_cv_4</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="n">group_with</span> <span class="o">=</span> <span class="s1">&#39;regions&#39;</span>
<span class="kn">output:</span>  <span class="n">f</span><span class="s1">&#39;{wd:a}/susie/{_input[2]:bn}.cv.RData&#39;</span><span class="p">,</span>
         <span class="n">f</span><span class="s1">&#39;{wd:a}/susie/{_input[2]:bn}.cv_diag.RData&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;12h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;8cG&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output[0]:bn}&#39;</span>
<span class="kn">R:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[0]}.stdout&#39;</span><span class="p">,</span><span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;dplyr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;tibble&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;readr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;plink2R&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;purrr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;modelr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;susieR&quot;</span><span class="p">)</span>
    
    <span class="c1"># Define functions</span>
   
    <span class="c1">## Compute rmse function</span>
    <span class="n">compute_rmse</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">raw</span><span class="p">,</span><span class="n">fitted</span><span class="p">){</span>
    <span class="n">rmse</span> <span class="o">=</span> <span class="n">rep</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">ncol</span><span class="p">(</span><span class="n">raw</span><span class="p">))</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">raw</span><span class="p">)){</span>
      <span class="n">rmse</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="p">((</span><span class="n">fitted</span> <span class="o">-</span> <span class="n">raw</span><span class="p">)[,</span><span class="n">i</span><span class="p">])</span><span class="o">^</span><span class="mi">2</span><span class="o">%&gt;%</span><span class="n">mean</span><span class="p">(</span><span class="n">na</span><span class="o">.</span><span class="n">rm</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">sqrt</span><span class="p">()</span> 
      <span class="p">}</span>
    <span class="k">return</span><span class="p">(</span><span class="n">rmse</span><span class="p">)</span>
    <span class="p">}</span>
    
    <span class="c1">## Compute r2 function</span>
    <span class="n">compute_r2</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">raw</span><span class="p">,</span><span class="n">fitted</span><span class="p">){</span>
      <span class="n">r2</span> <span class="o">=</span> <span class="n">rep</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">ncol</span><span class="p">(</span><span class="n">raw</span><span class="p">))</span>
      <span class="k">for</span> <span class="p">(</span><span class="n">j</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">raw</span><span class="p">)){</span>
       <span class="n">r2</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span> <span class="n">summary</span><span class="p">(</span><span class="n">lm</span><span class="p">(</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">fitted</span><span class="p">[,</span><span class="n">j</span><span class="p">])</span> <span class="o">~</span> <span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">raw</span><span class="p">[,</span><span class="n">j</span><span class="p">])</span> <span class="p">))</span><span class="err">$</span><span class="n">adj</span><span class="o">.</span><span class="n">r</span><span class="o">.</span><span class="n">sq</span>
      <span class="p">}</span>
      <span class="k">return</span><span class="p">(</span><span class="n">r2</span><span class="p">)</span>
    <span class="p">}</span>
    
    <span class="c1">## Compute r2 raw</span>
    
    <span class="n">compute_r2_raw</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">raw</span><span class="p">,</span><span class="n">fitted</span><span class="p">){</span>
      <span class="n">r2</span> <span class="o">=</span> <span class="n">rep</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">ncol</span><span class="p">(</span><span class="n">raw</span><span class="p">))</span>
      <span class="k">for</span> <span class="p">(</span><span class="n">j</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">raw</span><span class="p">)){</span>
        <span class="n">r2</span><span class="p">[</span><span class="n">j</span><span class="p">]</span> <span class="o">=</span>  <span class="n">cor</span><span class="p">(</span><span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">fitted</span><span class="p">[,</span><span class="n">j</span><span class="p">])[</span><span class="n">which</span><span class="p">(</span><span class="err">!</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">raw</span><span class="p">[,</span><span class="n">j</span><span class="p">]))],</span><span class="n">raw</span><span class="p">[,</span><span class="n">j</span><span class="p">]</span><span class="o">%&gt;%</span><span class="n">na</span><span class="o">.</span><span class="n">omit</span><span class="p">())</span><span class="o">^</span><span class="mi">2</span>
      <span class="p">}</span>
      <span class="k">return</span><span class="p">(</span><span class="n">r2</span><span class="p">)</span>
    <span class="p">}</span>
    
    <span class="c1">## Get P.value</span>
    <span class="n">compute_pval</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">raw</span><span class="p">,</span><span class="n">fitted</span><span class="p">){</span>
      <span class="n">pval</span> <span class="o">=</span> <span class="n">rep</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">ncol</span><span class="p">(</span><span class="n">raw</span><span class="p">))</span>
      <span class="k">for</span> <span class="p">(</span><span class="n">k</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">raw</span><span class="p">)){</span>
        <span class="n">pval</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="o">=</span> <span class="n">summary</span><span class="p">(</span><span class="n">lm</span><span class="p">(</span> <span class="n">fitted</span><span class="p">[,</span><span class="n">k</span><span class="p">]</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">matrix</span> <span class="o">~</span> <span class="n">raw</span><span class="p">[,</span><span class="n">k</span><span class="p">]</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">matrix</span> <span class="p">))</span><span class="err">$</span><span class="n">coef</span><span class="p">[</span><span class="mi">2</span><span class="p">,</span><span class="mi">4</span><span class="p">]</span>
      <span class="p">}</span>
      <span class="k">return</span><span class="p">(</span><span class="n">pval</span><span class="p">)</span>
    <span class="p">}</span>
    
    

    <span class="c1">###Functions to compute MAF and missing genotype rate</span>
    <span class="n">compute_maf</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">geno</span><span class="p">){</span>
      <span class="n">f</span> <span class="o">&lt;-</span> <span class="n">mean</span><span class="p">(</span><span class="n">geno</span><span class="p">,</span><span class="n">na</span><span class="o">.</span><span class="n">rm</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span>
      <span class="k">return</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">f</span><span class="p">,</span> <span class="mi">1</span><span class="o">-</span><span class="n">f</span><span class="p">))</span>
    <span class="p">}</span>
    
    <span class="n">compute_missing</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">geno</span><span class="p">){</span>
      <span class="n">miss</span> <span class="o">&lt;-</span> <span class="nb">sum</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">geno</span><span class="p">))</span><span class="o">/</span><span class="n">length</span><span class="p">(</span><span class="n">geno</span><span class="p">)</span>
      <span class="k">return</span><span class="p">(</span><span class="n">miss</span><span class="p">)</span>
    <span class="p">}</span>
    
    <span class="n">mean_impute</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">geno</span><span class="p">){</span>
      <span class="n">f</span> <span class="o">&lt;-</span> <span class="nb">apply</span><span class="p">(</span><span class="n">geno</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">mean</span><span class="p">(</span><span class="n">x</span><span class="p">,</span><span class="n">na</span><span class="o">.</span><span class="n">rm</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">))</span>
      <span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">f</span><span class="p">))</span> <span class="n">geno</span><span class="p">[,</span><span class="n">i</span><span class="p">][</span><span class="n">which</span><span class="p">(</span><span class="ow">is</span><span class="o">.</span><span class="n">na</span><span class="p">(</span><span class="n">geno</span><span class="p">[,</span><span class="n">i</span><span class="p">]))]</span> <span class="o">&lt;-</span> <span class="n">f</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
      <span class="k">return</span><span class="p">(</span><span class="n">geno</span><span class="p">)</span>
    <span class="p">}</span>
    
    <span class="n">is_zero_variance</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="p">{</span>
      <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">x</span><span class="p">))</span><span class="o">==</span><span class="mi">1</span><span class="p">)</span> <span class="k">return</span><span class="p">(</span><span class="n">T</span><span class="p">)</span>
      <span class="k">else</span> <span class="k">return</span><span class="p">(</span><span class="n">F</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="c1">### Filter X matrix</span>
    <span class="n">filter_X</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">missing_rate_thresh</span><span class="p">,</span> <span class="n">maf_thresh</span><span class="p">)</span> <span class="p">{</span>
      <span class="n">rm_col</span> <span class="o">&lt;-</span> <span class="n">which</span><span class="p">(</span><span class="nb">apply</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">compute_missing</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">missing_rate_thresh</span><span class="p">)</span>
      <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">rm_col</span><span class="p">))</span> <span class="n">X</span> <span class="o">&lt;-</span> <span class="n">X</span><span class="p">[,</span> <span class="o">-</span><span class="n">rm_col</span><span class="p">]</span>
      <span class="n">rm_col</span> <span class="o">&lt;-</span> <span class="n">which</span><span class="p">(</span><span class="nb">apply</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">compute_maf</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">maf_thresh</span><span class="p">)</span>
      <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">rm_col</span><span class="p">))</span> <span class="n">X</span> <span class="o">&lt;-</span> <span class="n">X</span><span class="p">[,</span> <span class="o">-</span><span class="n">rm_col</span><span class="p">]</span>
      <span class="n">rm_col</span> <span class="o">&lt;-</span> <span class="n">which</span><span class="p">(</span><span class="nb">apply</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">is_zero_variance</span><span class="p">))</span>
      <span class="k">if</span> <span class="p">(</span><span class="n">length</span><span class="p">(</span><span class="n">rm_col</span><span class="p">))</span> <span class="n">X</span> <span class="o">&lt;-</span> <span class="n">X</span><span class="p">[,</span> <span class="o">-</span><span class="n">rm_col</span><span class="p">]</span>
      <span class="k">return</span><span class="p">(</span><span class="n">mean_impute</span><span class="p">(</span><span class="n">X</span><span class="p">))</span>
    <span class="p">}</span>
    
    <span class="c1">### Produce CV dataset</span>
    <span class="n">cv_data_gen</span> <span class="o">=</span> <span class="n">function</span><span class="p">(</span><span class="n">X</span><span class="p">,</span><span class="n">Y</span><span class="p">,</span><span class="n">times</span><span class="p">,</span><span class="n">test_prop</span><span class="p">){</span>
    <span class="c1"># Merged the X and Y for producing testing and training set for modelr cv</span>
    <span class="n">cv_df_raw</span> <span class="o">=</span> <span class="n">cbind</span><span class="p">(</span><span class="n">X</span><span class="p">,</span><span class="n">Y</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">as_tibble</span><span class="p">()</span> 
    <span class="n">cv_df</span> <span class="o">=</span> <span class="n">crossv_mc</span><span class="p">(</span><span class="n">cv_df_raw</span><span class="p">,</span> <span class="n">times</span> <span class="p">,</span><span class="n">test</span> <span class="o">=</span> <span class="n">test_prop</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span>
      <span class="n">train_X</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="n">train</span><span class="p">,</span><span class="o">~</span><span class="n">as_tibble</span><span class="p">(</span><span class="o">.</span><span class="n">x</span><span class="p">)[</span><span class="mi">1</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">X</span><span class="p">)]</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">),</span>
      <span class="n">train_Y</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="n">train</span><span class="p">,</span><span class="o">~</span><span class="n">as_tibble</span><span class="p">(</span><span class="o">.</span><span class="n">x</span><span class="p">)[(</span><span class="n">ncol</span><span class="p">(</span><span class="n">X</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">):(</span><span class="n">ncol</span><span class="p">(</span><span class="n">X</span><span class="p">)</span><span class="o">+</span><span class="n">ncol</span><span class="p">(</span><span class="n">Y</span><span class="p">))]</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">),</span>
      <span class="n">test_X</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="n">test</span><span class="p">,</span><span class="o">~</span><span class="n">as_tibble</span><span class="p">(</span><span class="o">.</span><span class="n">x</span><span class="p">)[</span><span class="mi">1</span><span class="p">:</span><span class="n">ncol</span><span class="p">(</span><span class="n">X</span><span class="p">)]</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">),</span>
      <span class="n">test_Y</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="n">test</span><span class="p">,</span><span class="o">~</span><span class="n">as_tibble</span><span class="p">(</span><span class="o">.</span><span class="n">x</span><span class="p">)[(</span><span class="n">ncol</span><span class="p">(</span><span class="n">X</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span><span class="p">):(</span><span class="n">ncol</span><span class="p">(</span><span class="n">X</span><span class="p">)</span><span class="o">+</span><span class="n">ncol</span><span class="p">(</span><span class="n">Y</span><span class="p">))]</span><span class="o">%&gt;%</span><span class="n">as_tibble</span><span class="p">)</span>
    <span class="p">)</span>  
    
    <span class="c1"># Filter Train X with maf and missing, filter test X with the same col as Train X</span>
    <span class="n">cv_df</span> <span class="o">=</span> <span class="n">cv_df</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span>
    <span class="n">train_X</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="n">train_X</span><span class="p">,</span><span class="o">~</span><span class="n">filter_X</span><span class="p">(</span><span class="o">.</span><span class="n">x</span><span class="p">,</span><span class="mf">0.1</span><span class="p">,</span><span class="mf">0.01</span><span class="p">)),</span>
    <span class="n">test_X</span> <span class="o">=</span> <span class="n">map2</span><span class="p">(</span><span class="n">test_X</span><span class="p">,</span><span class="n">train_X</span><span class="p">,</span><span class="o">~.</span><span class="n">x</span><span class="o">%&gt;%</span><span class="n">as_tibble</span><span class="p">()</span><span class="o">%&gt;%</span><span class="n">select</span><span class="p">(</span><span class="n">colnames</span><span class="p">(</span><span class="o">.</span><span class="n">y</span><span class="p">))</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">matrix</span><span class="p">())</span>
    <span class="p">)</span>
    <span class="k">return</span><span class="p">(</span><span class="n">cv_df</span><span class="p">)</span>
    <span class="p">}</span>
    
    <span class="c1"># Load Data</span>
    <span class="n">full_model</span> <span class="o">=</span> <span class="n">attach</span><span class="p">(</span><span class="s1">&#39;$[_input[2]]&#39;</span><span class="p">)</span>
    <span class="n">full_model</span> <span class="o">=</span> <span class="n">full_model</span><span class="err">$</span><span class="n">fitted1</span>
    <span class="n">X</span> <span class="o">=</span> <span class="n">full_model</span><span class="err">$</span><span class="n">X</span>
    <span class="n">Y</span> <span class="o">=</span> <span class="n">full_model</span><span class="err">$</span><span class="n">Y</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">tibble</span><span class="p">()</span>

    <span class="c1"># Create cv dataset</span>
        
    <span class="n">cv_df</span> <span class="o">=</span> <span class="n">cv_data_gen</span><span class="p">(</span><span class="n">X</span><span class="p">,</span><span class="n">Y</span><span class="p">,</span><span class="err">$</span><span class="p">[</span><span class="n">cv_times</span><span class="p">],</span><span class="err">$</span><span class="p">[</span><span class="n">testing_prop</span><span class="p">])</span>

    <span class="c1"># Actual cv</span>
    
    <span class="n">cv_df</span> <span class="o">=</span> <span class="n">cv_df</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span>
    
   
    <span class="c1">## Do susie</span>
    
      <span class="n">susie</span> <span class="o">=</span> <span class="n">pmap</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">train_X</span><span class="p">,</span><span class="n">train_Y</span><span class="p">),</span><span class="n">function</span><span class="p">(</span><span class="n">first</span><span class="p">,</span><span class="n">second</span><span class="p">)(</span>
        
        <span class="n">susie</span><span class="p">(</span><span class="n">first</span><span class="p">,</span> <span class="n">second</span><span class="p">,</span>
        <span class="n">L</span> <span class="o">=</span> <span class="err">$</span><span class="p">[</span><span class="n">causal_variables_L</span><span class="p">],</span>
        <span class="n">estimate_residual_variance</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> 
        <span class="n">estimate_prior_variance</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span>
        <span class="n">scaled_prior_variance</span> <span class="o">=</span> <span class="err">$</span><span class="p">[</span><span class="n">scaled_prior_variance</span><span class="p">])</span>
        <span class="p">)))</span>
    
    <span class="c1"># Extract data </span>
    
    <span class="n">cv_df</span> <span class="o">=</span> <span class="n">cv_df</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span>
      <span class="n">weight</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="n">susie</span><span class="p">,</span><span class="o">~</span>
      <span class="p">(</span><span class="n">coef</span><span class="p">(</span><span class="o">.</span><span class="n">x</span><span class="p">)[</span><span class="mi">2</span><span class="p">:</span><span class="n">length</span><span class="p">(</span><span class="n">coef</span><span class="p">(</span><span class="o">.</span><span class="n">x</span><span class="p">))])</span>
      <span class="p">),</span>
      <span class="n">test_fitted</span> <span class="o">=</span> <span class="n">map2</span><span class="p">(</span><span class="n">susie</span><span class="p">,</span><span class="n">test_X</span><span class="p">,</span><span class="o">~</span><span class="n">predict</span><span class="p">(</span><span class="o">.</span><span class="n">x</span><span class="p">,</span><span class="o">.</span><span class="n">y</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">as_tibble</span><span class="p">),</span>
      <span class="n">rmse</span> <span class="o">=</span> <span class="n">map2</span><span class="p">(</span><span class="n">test_Y</span><span class="p">,</span><span class="n">test_fitted</span><span class="p">,</span><span class="o">~</span><span class="n">compute_rmse</span><span class="p">(</span><span class="o">.</span><span class="n">x</span><span class="p">,</span><span class="o">.</span><span class="n">y</span><span class="p">)),</span>
      <span class="n">r2</span> <span class="o">=</span> <span class="n">map2</span><span class="p">(</span><span class="n">test_Y</span><span class="p">,</span><span class="n">test_fitted</span><span class="p">,</span><span class="o">~</span><span class="n">compute_r2</span><span class="p">(</span><span class="o">.</span><span class="n">x</span><span class="p">,</span><span class="o">.</span><span class="n">y</span><span class="p">)),</span>
      <span class="n">r2_raw</span> <span class="o">=</span> <span class="n">map2</span><span class="p">(</span><span class="n">test_Y</span><span class="p">,</span><span class="n">test_fitted</span><span class="p">,</span><span class="o">~</span><span class="n">compute_r2_raw</span><span class="p">(</span><span class="o">.</span><span class="n">x</span><span class="p">,</span><span class="o">.</span><span class="n">y</span><span class="p">)),</span>
      <span class="n">pval</span> <span class="o">=</span> <span class="n">map2</span><span class="p">(</span><span class="n">test_Y</span><span class="p">,</span><span class="n">test_fitted</span><span class="p">,</span><span class="o">~</span><span class="n">compute_pval</span><span class="p">(</span><span class="o">.</span><span class="n">x</span><span class="p">,</span><span class="o">.</span><span class="n">y</span><span class="p">))</span>
    <span class="p">)</span>
    
    <span class="c1"># Calculate metrics</span>
    
    <span class="n">mean_rmse</span> <span class="o">=</span> <span class="n">cv_df</span><span class="o">%&gt;%</span><span class="n">pull</span><span class="p">(</span><span class="n">rmse</span><span class="p">)</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">()</span><span class="o">%&gt;%</span><span class="n">t</span><span class="p">()</span><span class="o">%&gt;%</span><span class="n">as_tibble</span><span class="p">()</span><span class="o">%&gt;%</span><span class="n">colMeans</span><span class="p">()</span>
    <span class="n">mean_r2</span> <span class="o">=</span> <span class="n">cv_df</span><span class="o">%&gt;%</span><span class="n">pull</span><span class="p">(</span><span class="n">r2</span><span class="p">)</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">()</span><span class="o">%&gt;%</span><span class="n">t</span><span class="p">()</span><span class="o">%&gt;%</span><span class="n">as_tibble</span><span class="p">()</span><span class="o">%&gt;%</span><span class="n">colMeans</span><span class="p">()</span>
    <span class="n">mean_r2_raw</span> <span class="o">=</span> <span class="n">cv_df</span><span class="o">%&gt;%</span><span class="n">pull</span><span class="p">(</span><span class="n">r2_raw</span><span class="p">)</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">()</span><span class="o">%&gt;%</span><span class="n">t</span><span class="p">()</span><span class="o">%&gt;%</span><span class="n">as_tibble</span><span class="p">()</span><span class="o">%&gt;%</span><span class="n">colMeans</span><span class="p">()</span>
    <span class="n">mean_pval</span> <span class="o">=</span> <span class="n">cv_df</span><span class="o">%&gt;%</span><span class="n">pull</span><span class="p">(</span><span class="n">pval</span><span class="p">)</span><span class="o">%&gt;%</span><span class="k">as</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">()</span><span class="o">%&gt;%</span><span class="n">t</span><span class="p">()</span><span class="o">%&gt;%</span><span class="n">as_tibble</span><span class="p">()</span><span class="o">%&gt;%</span><span class="n">colMeans</span><span class="p">()</span>

  
    <span class="c1"># Save metrics</span>
    <span class="n">full_model</span><span class="err">$</span><span class="n">rmse</span> <span class="o">=</span> <span class="n">mean_rmse</span>
    <span class="n">full_model</span><span class="err">$</span><span class="n">r2</span> <span class="o">=</span> <span class="n">mean_r2</span> 
    <span class="n">full_model</span><span class="err">$</span><span class="n">r2_raw</span> <span class="o">=</span> <span class="n">mean_r2_raw</span>    
    <span class="n">full_model</span><span class="err">$</span><span class="n">pval</span> <span class="o">=</span> <span class="n">mean_pval</span>    
    <span class="n">fitted1</span> <span class="o">=</span> <span class="n">full_model</span>
    <span class="c1"># Save the CV data</span>
    <span class="n">save</span><span class="p">(</span><span class="n">cv_df</span><span class="p">,</span><span class="nb">file</span> <span class="o">=</span> <span class="s2">&quot;$[_output[1]]&quot;</span><span class="p">)</span>
    
    <span class="c1">#Output</span>
    <span class="n">save</span><span class="p">(</span><span class="n">fitted1</span><span class="p">,</span><span class="nb">file</span> <span class="o">=</span> <span class="s2">&quot;$[_output[0]]&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Save-output">Save output<a class="anchor-link" href="#Save-output">&#182;</a></h2><p>This step create the all_hsq.txt file to summarize the susie result and create a R object to host all the susieR models.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1">#Create all_hsq for susie and saved all the susie object into one RData</span>
<span class="p">[</span><span class="n">susie_4</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">output:</span><span class="n">f</span><span class="s1">&#39;{wd:a}/susie/all_hsq.txt&#39;</span><span class="p">,</span>
       <span class="n">f</span><span class="s1">&#39;{wd:a}/susie.RData&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;12h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;6G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output[0]:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span>
  <span class="n">head</span> <span class="o">-</span><span class="mi">1</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
  <span class="n">cat</span> <span class="err">$</span><span class="p">[</span><span class="n">wd</span><span class="p">:</span><span class="n">a</span><span class="p">]</span><span class="o">/</span><span class="n">susie</span><span class="o">/*.</span><span class="n">hsq</span> <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="n">hsq_full_sample</span> <span class="o">|</span> <span class="n">uniq</span> <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>

<span class="kn">R:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[1]}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[1]}.stdout&#39;</span><span class="p">,</span><span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;dplyr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;tibble&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;readr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;purrr&quot;</span><span class="p">)</span>
    <span class="c1"># Load a template</span>
    <span class="n">region</span> <span class="o">=</span> <span class="n">read_delim</span><span class="p">(</span><span class="s2">&quot;$[_output[0]]&quot;</span><span class="p">,</span><span class="n">delim</span> <span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">select</span><span class="p">(</span><span class="n">path</span> <span class="o">=</span> <span class="nb">file</span><span class="p">,</span> <span class="n">ID</span> <span class="o">=</span> <span class="n">region</span> <span class="p">)</span>
    <span class="c1"># get the path</span>
    <span class="nb">dir</span> <span class="o">=</span> <span class="s2">&quot;$[_input[0]:d]/&quot;</span>
    <span class="n">pre</span> <span class="o">=</span> <span class="s2">&quot;$[_input[0]:bnnnnn]&quot;</span>
    <span class="n">sur</span> <span class="o">=</span> <span class="s2">&quot;.susie.model.RData&quot;</span>
    <span class="n">region</span> <span class="o">=</span> <span class="n">region</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span><span class="n">path</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="n">ID</span><span class="p">,</span> <span class="o">~</span><span class="n">paste</span><span class="p">(</span><span class="n">collapse</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">c</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="n">pre</span><span class="p">,</span><span class="s2">&quot;.&quot;</span><span class="p">,</span><span class="o">.</span><span class="n">x</span><span class="p">,</span><span class="n">sur</span><span class="p">))))</span>
    <span class="c1"># Load the data</span>
    <span class="n">output</span> <span class="o">=</span> <span class="n">region</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span><span class="kp">env</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="n">path</span><span class="p">,</span><span class="o">~</span><span class="n">attach</span><span class="p">(</span><span class="o">.</span><span class="n">x</span><span class="p">)),</span>
                             <span class="n">model</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="kp">env</span><span class="p">,</span> <span class="o">~.</span><span class="n">x</span><span class="err">$</span><span class="n">fitted1</span><span class="p">))</span>
    <span class="c1"># Save the combined output</span>
    <span class="n">save</span><span class="p">(</span><span class="n">output</span><span class="p">,</span><span class="nb">file</span> <span class="o">=</span> <span class="s2">&quot;$[_output[1]]&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="c1">#Create all_hsq for susie and saved all the susie object into one RData</span>
<span class="p">[</span><span class="n">susie_cv_5</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">output:</span><span class="n">f</span><span class="s1">&#39;{wd:a}/susie/all_hsq.txt&#39;</span><span class="p">,</span>
       <span class="n">f</span><span class="s1">&#39;{wd:a}/susie_cv.RData&#39;</span>
<span class="kn">task:</span> <span class="kp">trunk_workers</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">trunk_size</span> <span class="o">=</span> <span class="n">job_size</span><span class="p">,</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;12h&#39;</span><span class="p">,</span>  <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;6G&#39;</span><span class="p">,</span> <span class="kp">tags</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{step_name}_{_output[0]:bn}&#39;</span>
<span class="kn">bash:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span>

  <span class="n">head</span> <span class="o">-</span><span class="mi">1</span> <span class="err">$</span><span class="p">[</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span><span class="n">nnn</span><span class="p">]</span><span class="o">.</span><span class="n">hsq</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>
  <span class="n">cat</span> <span class="err">$</span><span class="p">[</span><span class="n">wd</span><span class="p">:</span><span class="n">a</span><span class="p">]</span><span class="o">/</span><span class="n">susie</span><span class="o">/*.</span><span class="n">hsq</span> <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="n">hsq_full_sample</span> <span class="o">|</span> <span class="n">uniq</span> <span class="o">&gt;&gt;</span> <span class="err">$</span><span class="p">[</span><span class="n">_output</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span>

<span class="kn">R:</span> <span class="n">expand</span><span class="o">=</span> <span class="s2">&quot;$[ ]&quot;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[1]}.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output[1]}.stdout&#39;</span><span class="p">,</span><span class="n">container</span> <span class="o">=</span> <span class="n">container</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;dplyr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;tibble&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;readr&quot;</span><span class="p">)</span>
    <span class="n">library</span><span class="p">(</span><span class="s2">&quot;purrr&quot;</span><span class="p">)</span>
    <span class="c1"># Load a template</span>
    <span class="n">region</span> <span class="o">=</span> <span class="n">read_delim</span><span class="p">(</span><span class="s2">&quot;$[_output[0]]&quot;</span><span class="p">,</span><span class="n">delim</span> <span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">%&gt;%</span><span class="n">select</span><span class="p">(</span><span class="n">path</span> <span class="o">=</span> <span class="nb">file</span><span class="p">,</span> <span class="n">ID</span> <span class="o">=</span> <span class="n">region</span> <span class="p">)</span>
    <span class="c1"># get the path</span>
    <span class="nb">dir</span> <span class="o">=</span> <span class="s2">&quot;$[_input[0]:d]/&quot;</span>
    <span class="n">pre</span> <span class="o">=</span> <span class="s2">&quot;$[_input[0]:bnnnnn]&quot;</span>
    <span class="n">sur</span> <span class="o">=</span> <span class="s2">&quot;.susie.model.cv.RData&quot;</span>
    <span class="n">region</span> <span class="o">=</span> <span class="n">region</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span><span class="n">path</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="n">ID</span><span class="p">,</span> <span class="o">~</span><span class="n">paste</span><span class="p">(</span><span class="n">collapse</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">c</span><span class="p">(</span><span class="nb">dir</span><span class="p">,</span><span class="n">pre</span><span class="p">,</span><span class="s2">&quot;.&quot;</span><span class="p">,</span><span class="o">.</span><span class="n">x</span><span class="p">,</span><span class="n">sur</span><span class="p">))))</span>
    <span class="c1"># Load the data</span>
    <span class="n">output</span> <span class="o">=</span> <span class="n">region</span><span class="o">%&gt;%</span><span class="n">mutate</span><span class="p">(</span><span class="kp">env</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="n">path</span><span class="p">,</span><span class="o">~</span><span class="n">attach</span><span class="p">(</span><span class="o">.</span><span class="n">x</span><span class="p">)),</span>
                             <span class="n">model</span> <span class="o">=</span> <span class="nb">map</span><span class="p">(</span><span class="kp">env</span><span class="p">,</span> <span class="o">~.</span><span class="n">x</span><span class="err">$</span><span class="n">fitted1</span><span class="p">))</span>
    <span class="c1"># Save the combined output</span>
    <span class="n">save</span><span class="p">(</span><span class="n">output</span><span class="p">,</span><span class="nb">file</span> <span class="o">=</span> <span class="s2">&quot;$[_output[1]]&quot;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
<hr>
&copy 2020-2021 Hao Sun, Department of Neurology, Columbia University
<p><small>Exported from <a href="http://github.com/hsun3163/neuro-twas/blob/778b6fae0668fc16b895c0439b2e83bc7c7d2415/workflow/uni_susie.ipynb"><code>workflow/uni_susie.ipynb</code></a> committed by Hao Sun on Thu May 6 21:43:34 2021 <a href="http://github.com/hsun3163/neuro-twas/commit/778b6fae0668fc16b895c0439b2e83bc7c7d2415">revision 3, 778b6fa</a> <a href="https://stephenslab.github.io/ipynb-website/notes.html#Note-about-commit-ids"><span class="fa fa-question-circle"></span></a></small></p>
</div>
</div>
</body>
</html>
